---
title: "Simulating Mutations"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Simulating Mutations}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  markdown:
    wrap: 72
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

rRACES/RACES can simulate genomic mutations on the cells represented in
a `SamplesForest` according to a specified mutational signature of single
base substitutions (SBS) (see [@alexandrov2020repertoire]).
This process is performed by the class `MutationEngine` which also takes
into account the mutation rate of the simulated species and gives the
chance to dynamicaly change the SBS signature.

### Setting Up Mutation Engine

The creation of an object of the type `MutationEngine` requires to download
the reference sequence and the corresponding SBS file and to build the
corresponding context index. The function `build_mutation_engine` performs
all these steps in a single call.

The function `build_mutation_engine` can be called by specifing the name
of the set-up directory, the complete URLs of both refence genome and
SBS file, the default number of alleles per chromosome, and, optionally,
a name list of exceptions for the default number of alleles whose names
must be chromosome names.

```{r}
library(rRACES)

reference_url <- paste0("https://ftp.ensembl.org/pub/release-110/fasta/",
                        "homo_sapiens/dna/Homo_sapiens.GRCh38.dna.",
                        "chromosome.22.fa.gz")

sbs_url <- paste0("https://cancer.sanger.ac.uk/signatures/documents/2124/",
                  "COSMIC_v3.4_SBS_GRCh38.txt")

# build a mutation engine and place all the files in the directory "Test"
m_engine <- build_mutation_engine(directory = "Test",
                                  reference_url = reference_url,
                                  SBS_url = sbs_url,
                                  default_num_of_alleles = 2)

m_engine
```

The above call creates the directory `Test`, fills it with all the
data required to build a mutation engine, and build it.

```{r}
dir.exists("Test")

list.files("Test")
```

The execution of the function may takes some time, but it is meant to be performed
one for all and, as long as the user does not need to change the reference genome
or the SBS file, it is no more required. In this spirit, any call to the function
`MutationEngine` checks whether all the building process sub-tasks are really
needed and, if this is not the case, avoids to performs them.

```{r}
# build a new mutation engine and use all the data stored into the
# directory "Test"
m_engine <- build_mutation_engine("Test", reference_url,
                                  sbs_url, default_num_of_alleles = 2)

m_engine
```

The `MutationEngine` function also builds a context index of the reference
sequence. The complete context index of a sequence whose size is up to 4Gbps
takes 4 times the length of the sequence itself both in memory and on the disk.
For instance, the complete contex index of a human genome takes about 12GBytes.
In order to avoid such requirement in memory and disk space,
`MutationEngine` allows to sample the reference genome contexts and
to stores only some of the them in the context index. This is achieved by
the optional parameter `context_sampling` which specifies how many
occurences of the same context must be identified before adding one of
them to the context index. The larger the number of context sampling,
the larger the context index. On the other side, the lower the number
of context sampling, the lower the number of sites in the refernce
genome that can be affected by simulated mutations.
The `context_sampling` is set to 100 by default, but it can
be specified during the `MutationEngine` call as the last parameter.

```{r}
# set-up the data and build the context index with `context_sampling` is 50
m_engine <- build_mutation_engine("Test", reference_url,
                                  sbs_url, default_num_of_alleles = 2,
                                  context_sampling = 50)

# get the size of the context index when `context_sampling` is 100
utils:::format.object_size(file.size("Test/context_index_100.cif"), "auto")

# get the size of the context index when `context_sampling` is 50
utils:::format.object_size(file.size("Test/context_index_50.cif"), "auto")
```

```{r, echo=FALSE,  results='hide', message=FALSE}
unlink("Test", recursive = TRUE)
```

The function `MutationEngine` is quite flexible and it allows to specify
any possible reference genome and any SBS file. However, the vast majority of
rRACES users aim a standard set-up for instance involving the human genome.
Because of this, rRACES provides some predefined set-ups that can be used
when building a `MutationEngine` object.
A complete list of the supported predefined set-ups can be obtained by invoking
the function `get_mutation_engine_codes`.

```{r}
get_mutation_engine_codes()
```

Let us build a `MutationEngine` by using the "demo" configuration.

```{r, echo=FALSE,  results='hide', message=FALSE}
unlink("demo", recursive = TRUE)
```

```{r}
# building a mutation engine by using the "demo" set-up configuration
m_engine <- build_mutation_engine(setup_code = "demo")

dir.exists("demo")
```

We can figure out the size of the context index as done above.

```{r}
# get the size of the context index when `context_sampling` is 100
utils:::format.object_size(file.size("demo/context_index_100.cif"), "auto")
```

We can remove the set-up and rebuild it decreasing the
`context_sampling` parameter.

```{r}
# building a mutation engine by using the "demo" set-up configuration
# and setting context_sampling to 50
m_engine <- build_mutation_engine(setup_code = "demo", context_sampling = 50)

# get the size of the context index when `context_sampling` is 50
utils:::format.object_size(file.size("demo/context_index_50.cif"), "auto")
```


#### Mutant Genetic Characterization 

Once the mutation engine has been built, we can define mutant genotype and
declare species mutation rates.

Let us consider the _"Border Growth"_ simulation peformed in
`vignette("growth_models")`. It involves the two mutants `A` and
"B". Both of them has two possible epigenetic states, `+` and `-`, leading 
to the four the species `A+`, `A-`, `B+`, and `B-`, respectively.
Each of these species has a passenger mutation rate that must be 
declated to the mutation engine before labeling the samples forest.
Let 1e-9, 3e-8, 8e-7, and 5e-8 be the passenger SNV rates of the
species `A+`, `A-`, `B+`, and `B-`, respectively. Furthermore, 
let 0, 1e-11, 0, and 0 be the passenger CNA rates of the same 
species, respectively.

The two mutants may also be genetically characterized by some
driver mutations. 
The driver mutations associated with each of the mutants must occur 
in any cell belonging to the mutant itself. Hence, they must be 
declared to the mutant engine before the labeling.

The method `MutationEngine$add_mutant` takes care of all these 
declarations.

For the sake of example, let us assume that `A` is characterized by one
driver mutation on chromosome 22, while `B` has two driver mutations on
the same chromosome.

```{r}
# add the mutant "A" characterized by one driver SNV on chromosome 22 and 
# two CNAs: an amplification and a deletion. The mutant has two epigenetic 
# states and its species "A+" and "A-" have passenger SNV rates 1e-9 and 
# 3e-8, respectively, and passenger CNA rates 0 and 1e-11, respectively.
m_engine$add_mutant(mutant_name = "A",
                    passenger_rates = list("+" = c(SNV = 1e-9),
                                           "-" = c(SNV = 3e-8, CNA = 1e-11)),
                    driver_SNVs = c(SNV("22", 10510210, "C")),
                    driver_CNAs = c(CNA(type = "A", "22", pos_in_chr = 10303470,
                                        len = 200000),
                                    CNA("D", "22", 5010000, 200000)))

# add the mutant "B" characterized by two driver SNVs on chromosome 22 (no
# CNA) and two epigenetic states. Its species "B+" and "B-" have passenger
# SNV rates 8e-7 and 5e-8, respectively, and CNA rates 0 for both species.
m_engine$add_mutant("B", list("+" = c(SNV = 8e-7), "-" = c(SNV = 5e-8)),
                    c(SNV("22", 12028576, "G"),
                      SNV("22", 12029660, "A")))

m_engine
```

#### Mutational Exposures

The probability for a mutation to occur depends on both its genomic and 
environmental context. 

A SBS is a single base mutation substituion signature that provides for any
genomic context (i.e., a triplet of bases) the probability for a SNV to occur
on that context.

The SBSs depends on the environmental context (see [@alexandrov2020repertoire]) 
and, because of that, more than one SBSs may be active at the same time with
different probabilities. An _mutational exposure_ (or _exposure_) is a
discrete probability distribution among SBS signatures.

In order to simulate passeger SNVs, we need to specify a default exposure.
This can be achieved as it follows.

```{r}
# add a default exposure. This will be used from simulated
# time 0 up to the successive exposure change.
m_engine$add_exposure(coefficients = c(SBS13 = 0.2, SBS1 = 0.8))
```

Further exposures can also be defined by specifying an activation time 
for each of them, i.e., the new exposures will be used
from the specified time up to the next exposure change.

```{r}
# add a new exposure that will be used from simulated
# time 100 up to the successive exposure change.
m_engine$add_exposure(time = 100,
                      c(SBS5 = 0.3, SBS2 = 0.2, SBS3 = 0.5))

m_engine
```

#### Passenger CNAs and Tumor Type

The passenger CNAs applied during the simulation depend on tumor type. 
The type of tumor can be specified when building the mutation engine 
by using the `tumor_type` parameter. For instance, the passenger CNAs 
identified in breast carcinoma in the UK can be applied during the 
simulation as it follows.

```{r, eval=FALSE}
m_engine <- build_mutation_engine(setup_code = "demo",
                                  tumor_type = "BRCA-UK")
```

The function `get_admissible_CNA_tumor_types()` returns the 
complete list of tumors supported by rRACES.

```{r}
get_admissible_CNA_tumor_types()
```

### Building Phylogenetic Forests

The configurated mutation engine can be used to label each node in a 
samples forest by mutations.

Since the mutation engine has be configurated to deal with the 
simulation peformed in `vignette("growth_models")`,
we can use the samples forest saved in the file `"samples_forest.sff"`
in that article.

```{r}
samples_forest <- load_samples_forest("samples_forest.sff")

# place mutations on the sample forest assuming 1000
# pre-neoplastic mutations
phylo_forest <- m_engine$place_mutations(samples_forest, 1000)

phylo_forest
```

The phylogenetic forest stores all the mutations labeling the 
sampled cells which are represented by the forest leaves.
Users can retrieve such data by using the methods 
`Phylogenetic$get_sampled_cell_SNVs` and 
`Phylogenetic$get_sampled_cell_CNAs`.

```{r}
library(dplyr)

# select the first SNVs among all the mutations occuring in
# the genomes of the sampled cells
phylo_forest$get_sampled_cell_SNVs() %>% head()

# select the first CNAs among all the mutations occuring in
# the genomes of the sampled cells
phylo_forest$get_sampled_cell_CNAs() %>% head()

# get the sampled cells
sampled_cells <- phylo_forest$get_nodes() %>%
  filter(!is.na(.data$sample))

# show the first of them
sampled_cells %>% head()

# get the identifier of the 3rd cell in `sampled_cells`
cell_id <- sampled_cells[3, 1]

# get the SNVs of the 3rd cell in `sampled_cells`
phylo_forest$get_sampled_cell_SNVs(cell_id) %>% head()

# get the CNAs of the 3rd cell in `sampled_cells`
phylo_forest$get_sampled_cell_CNAs(cell_id) %>% head()
```

Users can also identify the cell in which a mutation emerged 
even when the cell was not sampled.

```{r}
# select one of the mutations
mutation_row <- phylo_forest$get_sampled_cell_SNVs(cell_id)[2,]

# rebuild the corresponding SNV
snv <- SNV(mutation_row["chromosome"][1, ],
           mutation_row["chr_pos"][1, ],
           mutation_row["mutated_base"][1, ],
           mutation_row["context"][1, ],)

# get the identifier of the oldest cells in which the SNV occurs
phylo_forest$get_first_occurrences(snv)
```

### Storing Phylogenetic Forests

As in the case of the samples forests, the phylogenetic forests 
can be saved by using the method `PhylogeneticForest$save` and 
load by the function `load_phylogenetic_forest`.

```{r}
# save the phylogenetic forest in the file "phylo_forest.sff"
phylo_forest$save("phylo_forest.sff")

# loading the saved forest
phylo_forest2 <- load_phylogenetic_forest("phylo_forest.sff")

phylo_forest2
```

```{r, echo=FALSE,  results='hide', message=FALSE}
unlink("samples_forest.sff")
```
