---
title: "Simulating Mutations"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Simulating Mutations}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  markdown:
    wrap: 72
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

rRACES/RACES can simulate genomic mutations on the cells represented in
a `SamplesForest` according to a specified mutational signature of single
base substitutions (SBS) (see [@alexandrov2020repertoire]).
This process is performed by the class `MutationEngine` which also takes
into account the mutation rate of the simulated species and gives the
chance to dynamically change the SBS signature.

### Setting Up Mutation Engine

The creation of an object of the type `MutationEngine` requires to download
the reference sequence and the corresponding SBS file and to build the
corresponding context index. The function `build_mutation_engine()` performs
all these steps in a single call.

This function is quite flexible and it allows to customize the mutation
engine in many ways. However, the vast majority of rRACES users aim a
standard setup for instance involving the human genome.

Because of this, rRACES provides some predefined setups.
Their list can be obtained by invoking the function `get_mutation_engine_codes()`.

```{r}
library(rRACES)

get_mutation_engine_codes()
```

Let us build a `MutationEngine` by using the `"demo"` setup.

```{r, echo=FALSE,  results='hide', message=FALSE}
unlink("demo", recursive = TRUE)
```

```{r}
# building a mutation engine by using the "demo" setup
m_engine <- build_mutation_engine(setup_code = "demo")

m_engine
```

The above call creates the directory `demo`,
downloads all the data required by the mutation
engine, and builds it.

```{r}
dir.exists("demo")

list.files("demo")
```

The execution of the function `build_mutation_engine()` may takes some time, but
it is meant to be performed one for all and, as long as the user does not need to
change the reference genome or the SBS file, it is no more required. In this
spirit, any call to the function `MutationEngine` checks whether all the building
process sub-tasks are really needed and, if this is not the case, it avoids to
performs them.

```{r}
# building a mutation engine by using the "demo" set-up configuration
m_engine <- build_mutation_engine(setup_code = "demo")
```

#### Context Index Sampling

The `MutationEngine` function also builds a context index of the reference
sequence. The complete context index of a sequence whose size is up to 4Gbps
takes 4 times the length of the sequence itself both in memory and on the disk.
For instance, the complete contex index of a human genome takes about 12GBytes.
In order to avoid such requirement in memory and disk space,
`MutationEngine` allows to sample the reference genome contexts and
to stores only some of the them in the context index. This is achieved by
the optional parameter `context_sampling` which specifies how many
occurences of the same context must be identified before adding one of
them to the context index. The larger the number of context sampling,
the larger the context index. On the other side, the lower the number
of context sampling, the lower the number of sites in the refernce
genome that can be affected by simulated mutations.
The `context_sampling` is set to 100 by default, but it can
be specified during the `MutationEngine` call as the last parameter.

```{r}
# get the size of the context index when `context_sampling` is 100
utils:::format.object_size(file.size("demo/context_index_100.cif"), "auto")
```

Let us rebuild the context index sampling one context every 50.

```{r}
# building a mutation engine by using the "demo" set-up configuration
# and setting context_sampling to 50
m_engine <- build_mutation_engine(setup_code = "demo", context_sampling = 50)
```

Above call uses the already downloaded data in the directory `"demo"` to
build the context and produces the file `"demo/context_index_50.cif"`.

```
# get the size of the context index when `context_sampling` is 50
utils:::format.object_size(file.size("demo/context_index_50.cif"), "auto")
```

#### Customizing `MutationEngine` (Advanced Topic)

The function `build_mutation_engine()` can be called by specifing the name
of the set-up directory, the path or URL of the reference sequence, the
SBS file, the driver mutation file, the passenger CNAs file, and the germline
data directory.

```{r}
reference_url <- paste0("https://ftp.ensembl.org/pub/grch37/current/",
                        "fasta/homo_sapiens/dna/Homo_sapiens.GRCh37.",
                        "dna.chromosome.22.fa.gz")

sbs_url <- paste0("https://cancer.sanger.ac.uk/signatures/documents/2123/",
                  "COSMIC_v3.4_SBS_GRCh37.txt")

drivers_url <- paste0("https://raw.githubusercontent.com/",
                      "caravagnalab/rRACES/main/inst/extdata/",
                      "driver_mutations_hg19.csv")

passenger_cnas_url <- paste0("https://raw.githubusercontent.com/",
                             "caravagnalab/rRACES/main/inst/extdata/",
                             "passenger_CNAs_hg19.csv")

germline_url <- paste0("https://www.dropbox.com/scl/fi/g9oloxkip18tr1r",
                       "m6wjve/germline_data_demo.tar.gz?rlkey=15jshul",
                       "d3bqgyfcs7fa0bzqeo&dl=1")

# build a mutation engine and place all the files in the directory "Test"
m_engine <- build_mutation_engine(directory = "Test",
                                  reference_src = reference_url,
                                  SBS_src = sbs_url,
                                  drivers_src = drivers_url,
                                  passenger_CNAs_src = passenger_cnas_url,
                                  germline_src = germline_url)

dir.exists("Test")
```

```{r, echo=FALSE,  results='hide', message=FALSE}
unlink("Test", recursive = TRUE)
```

#### Mutant Genetic Characterization

Once the mutation engine has been built, we can define mutant genotype and
declare species mutation rates.

Let us consider the _"Border Growth"_ simulation peformed in
`vignette("growth_models")`. It involves the two mutants `A` and
"B". Both of them has two possible epigenetic states, `+` and `-`, leading
to the four the species `A+`, `A-`, `B+`, and `B-`, respectively.
Each of these species has a passenger mutation rate that must be
declated to the mutation engine before labeling the samples forest.
Let 1e-9, 3e-8, 8e-7, and 5e-8 be the passenger SNV rates of the
species `A+`, `A-`, `B+`, and `B-`, respectively. Furthermore,
let 0, 1e-11, 0, and 0 be the passenger CNA rates of the same
species, respectively.

The two mutants may also be genetically characterized by some
driver mutations.
The driver mutations associated with each of the mutants must occur
in any cell belonging to the mutant itself. Hence, they must be
declared to the mutant engine before the labeling.

The method `MutationEngine$add_mutant()` takes care of all these
declarations.

For the sake of example, let us assume that `A` is characterized by one
driver mutation on chromosome 22, while `B` has three driver mutations on
the same chromosome.

The genetic specification of the mutant `A` can be declared as it follows.

```{r, echo=FALSE,  results='hide', message=FALSE}
m_engine <- build_mutation_engine(setup_code = "demo")
```

DA QUI

```{r}
# add the mutant "A" characterized by one driver SNV on the allele 1 of
# chromosome 22 and two CNAs: a deletion on the allele 1 of chromosome 22
# and an amplification on a random allele of the same chromosome. The
# mutant has two epigenetic states and its species "A+" and "A-" have
# passenger SNV rates 1e-9 and 3e-8, respectively, and passenger CNA rates
# 0 and 1e-11, respectively.
m_engine$add_mutant(mutant_name = "A",
                    passenger_rates = list("+" = c(SNV = 1e-9),
                                           "-" = c(SNV = 3e-8, CNA = 1e-11)),
                    drivers = list(SNV("22", 10510210, "C", allele = 1),
                                   CNA(type = "A", chr = "22",
                                       chr_pos = 10303470, len = 200000),
                                   CNA("D", "22", 5010000, 200000,
                                       allele = 1)))

m_engine
```

In the above code, the driver CNAs, SNVs, and indels are declared by calling
the functions `CNA()`, `SNV()`, and `indel()`, respectively. These function allow
to specify the allele in which the mutation must lay.

As far as the SNVs and indels concern, rRACES provides the users with a more
compact and, sometimes, convenient approach. The mutation engine stores a list
of known driver mutations and labels each of them by a code which can be used
during the mutant declaration. The corresponding data frame can be obtained
by using the method `MutationEngine$get_known_drivers()`.

```{r}
library(dplyr)

m_engine$get_known_drivers() %>% filter(chr == "22")
```

The code of the known driver mutations can be used in place of the full
specification as follows.

```{r}
# add the mutant "B" characterized by three driver SNVs on chromosome 22 (no
# CNA) and two epigenetic states. The first two SNVs are "NF2 R262", which
# can lay in any available allele, and "NF2 R221*", which instead must lay
# in the allele 1. The remaining SNV is specified by using the SNV function
# as done above. The species "B+" and "B-" have passenger SNV rates 8e-7
# and 5e-8, respectively, and CNA rates 0 for both species.
m_engine$add_mutant("B", list("+" = c(SNV = 8e-7), "-" = c(SNV = 5e-8)),
                    list(SNV("22", 12028576, "G"),         # the first SNV
                         "NF2 R262*",                      # the second SNV
                         list("NF2 R221*", allele = 1)))   # the third SNV

m_engine
```

A QUI

#### Mutational Exposures

The probability for a mutation to occur depends on both its genomic and
environmental context.

A SBS is a single base mutation substituion signature that provides for any
genomic context (i.e., a triplet of bases) the probability for a SNV to occur
on that context.

The SBSs depends on the environmental context (see [@alexandrov2020repertoire])
and, because of that, more than one SBSs may be active at the same time with
different probabilities. An _mutational exposure_ (or _exposure_) is a
discrete probability distribution among SBS signatures.

In order to simulate passeger SNVs, we need to specify a default exposure.
This can be achieved as it follows.

```{r}
# add a default exposure. This will be used from simulated
# time 0 up to the successive exposure change.
m_engine$add_exposure(coefficients = c(SBS13 = 0.2, SBS1 = 0.8))
```

Further exposures can also be defined by specifying an activation time
for each of them, i.e., the new exposures will be used
from the specified time up to the next exposure change.

```{r}
# add a new exposure that will be used from simulated
# time 100 up to the successive exposure change.
m_engine$add_exposure(time = 100,
                      c(SBS5 = 0.3, SBS2 = 0.2, SBS3 = 0.5))

m_engine
```

#### Passenger CNAs and Tumor Type

The passenger CNAs applied during the simulation depend on tumor type.
The type of tumor can be specified when building the mutation engine
by using the `tumor_type` parameter. For instance, if the
passenger CNAs file used to build the mutation engine contains
some of the data identified in breast carcinoma in the UK,
the specific CNA can be applied during the simulation as it follows.

```{r, eval=FALSE}
m_engine <- build_mutation_engine(setup_code = "demo",
                                  tumor_type = "BRCA-UK")
```

#### Germline Mutations

rRACES allows users to apply the germline mutations of one of
the subjects available in the germline data provided in building
the mutation engine. This feature will enable users to simulate
a specific cancer type's evolution on an individual with the
desired gender and ethnicity.

The available subjects, together with their sex and ethnicity, can
be obtained by using the method
`[MutationEngine$get_germline_subjects()]`.

```{r}
subjects <- m_engine$get_germline_subjects()

subjects
```

The column `sample` contains the names of the available subjects.
The columns `pop` and `super_pop` report the subjects' population
and super-population codes. The last column,
`gender`, includes the subject gender.

The method `MutationEngine$get_population_descritions()` clarifies
the meaning of the codes reported in the pop columns.

```{r}
m_engine$get_population_descritions()
```

The method `MutationEngine$get_active_germline()` returns the
the active germline subject.

```{r}
m_engine$get_active_germline()
```

Users can change the germline subject using the method
`MutationEngine$set_germline_subject()`.

When a subject is selected for the first time, rRACES builds
a binary representation of the subject genome, saving
it for future use. This step may take a few minutes. However,
all the successive selections of the same subject directly
load the binary file.

```{r}
m_engine$set_germline_subject(subjects[2, "sample"])

m_engine$get_active_germline()
```

### Building Phylogenetic Forests

The configurated mutation engine can be used to label each node in a
samples forest by mutations.

Since the mutation engine has be configurated to deal with the
simulation peformed in `vignette("growth_models")`,
we can use the samples forest saved in the file `"samples_forest.sff"`
in that article.

```{r}
samples_forest <- load_samples_forest("samples_forest.sff")

# place mutations on the sample forest assuming 1000
# pre-neoplastic mutations
phylo_forest <- m_engine$place_mutations(samples_forest, 1000)

phylo_forest
```

The phylogenetic forest stores all the mutations labeling the
sampled cells which are represented by the forest leaves.
Users can retrieve such data by using the methods
`PhylogeneticForest$get_sampled_cell_mutations()` and
`PhylogeneticForest$get_sampled_cell_CNAs()`.

```{r}
library(dplyr)

# select the first mutations among all the mutations occuring in
# the genomes of the sampled cells
phylo_forest$get_sampled_cell_mutations() %>% head()

# select the first CNAs among all the mutations occuring in
# the genomes of the sampled cells
phylo_forest$get_sampled_cell_CNAs() %>% head()

# get the sampled cells
sampled_cells <- phylo_forest$get_nodes() %>%
  filter(!is.na(.data$sample))

# show the first of them
sampled_cells %>% head()

# get the identifier of the 3rd cell in `sampled_cells`
cell_id <- sampled_cells[3, 1]

# get the SNVs and the indels of the 3rd cell in `sampled_cells`
phylo_forest$get_sampled_cell_mutations(cell_id) %>% head()

# get the CNAs of the 3rd cell in `sampled_cells`
phylo_forest$get_sampled_cell_CNAs(cell_id) %>% head()
```

The method `PhylogeneticForest$get_germline_mutations()` returns
the SNVs and the indels in the germline.

```{r}
# extract the germline mutation
phylo_forest$get_germline_mutations() %>% head()
```


Users can also identify the cell in which a mutation emerged
even when the cell was not sampled.

```{r}
# select one of the mutations
mutation_row <- phylo_forest$get_sampled_cell_mutations(cell_id)[2, ]

# rebuild the corresponding SNV
snv <- SNV(mutation_row["chr"][1, ],
           mutation_row["chr_pos"][1, ],
           mutation_row["alt"][1, ],
           mutation_row["ref"][1, ])

# get the identifier of the oldest cells in which the SNV occurs
phylo_forest$get_first_occurrences(snv)
```

The exposures used in placing the mutations on the cells in the
phylogenetic forest can be obtained by using the method
`PhylogeneticForest$get_exposures()`.

```{r}
# get the exposures used in placing the mutations
phylo_forest$get_exposures()
```

The details about the SBSs adopted during the evolution are
available in the mutation engine and they can be retrieved
by using the method `MutationEngine$get_SBSs()`.

```{r}
# get the SBSs used in placing the mutations
m_engine$get_SBSs()[1:6, 1:5]
```

The name of the subject whose germline corresponds to
wild-type genome in the phylogenetic forest can be
obtained by the method `Phylogenetic$get_germline_subject()`.

```{r}
phylo_forest$get_germline_subject()
```

### Infinite Sites Model

The mutation engine places mutation on the sampled cell genome according to the
[infinite sites model](https://en.wikipedia.org/wiki/Infinite_sites_model) by default.
In particular, any new mutations is placed on a locus whose context is mutation-free.

```{r}
# this function verifies whether any mutation arised in
# two unrelated cells represented in a phylogenetic forest
test_infinite_sites_model <- function(phylo_forest) {
  # extract non-germinal SNVs that appear multiple times either
  # in the same cell or different cells
  snvs <- phylo_forest$get_sampled_cell_mutations() %>%
    filter(class != "germinal", .data$type == "SNV") %>%
    count(.data$chr, .data$chr_pos, .data$ref, .data$alt) %>%
    filter(n > 1)

  # search for an SNV that independently occurred in two unrelated cells
  first_occurrences <- c()
  row <- 1
  while (length(first_occurrences) < 2 && row <= nrow(snvs)) {
    snv <- SNV(snvs[row, "chr"], snvs[row, "chr_pos"],
               ref = snvs[row, "ref"], alt = snvs[row, "alt"])

    first_occurrences <- phylo_forest$get_first_occurrences(snv)
    row <- row + 1
  }

  # if the last handled SNV independently occurred in two unrelated
  # cells at least
  if (length(first_occurrences) >= 2) {

    # print a message containing the two cells
    paste0("SNV('", snv$get_chromosome(), "'',",
           snv$get_position_in_chromosome(), ",'", snv$get_ref(),
           "','", snv$get_alt(),
           "') independently arises in cells ", first_occurrences[1],
           " and ", first_occurrences[2])
  } else {
    print("Every mutation arises exclusively in one cell")
  }
}

# test whether the infinite sites conditions hold
# in the built phylogenetic forest
test_infinite_sites_model(phylo_forest)
```

This behavior can be changed by using the mutation engine property
`MutationEngine$infinite_sites_model()`. This property is a Boolean flag
that enable/disable the infinite sites model.

```{r}
# establish whether the infinite sites model is used
m_engine$infinite_sites_model

# disable it
m_engine$infinite_sites_model <- FALSE
```

When the infinite sites model is disabled,
`MutationEngine$place_mutations()` may place two mutations
in the same locus of different alleles of the same genome
or the same mutation in the same locus of one allele
of the genomes of two cells that are not each other ancestors.

```{r}
# test whether the infinite sites model is enable
m_engine$infinite_sites_model

# place the mutations on the same samples forest
# used above
phylo_forest2 <- m_engine$place_mutations(samples_forest, 1000)

# test whether the infinite sites conditions hold
# in the new phylogenetic forest
test_infinite_sites_model(phylo_forest2)
```

### Storing Phylogenetic Forests

As in the case of the samples forests, the phylogenetic forests
can be saved by using the method `PhylogeneticForest$save()` and
load by the function `load_phylogenetic_forest()`.

```{r}
# save the phylogenetic forest in the file "phylo_forest.sff"
phylo_forest$save("phylo_forest.sff")

# loading the saved forest
loaded_phylo_forest <- load_phylogenetic_forest("phylo_forest.sff")

loaded_phylo_forest
```

```{r, echo=FALSE,  results='hide', message=FALSE}
unlink("samples_forest.sff")
```
