---
title: "Simulating Mutations"
output: rmarkdown::html_vignette
description: >
  How to simulate mutations on sampled cells
vignette: >
  %\VignetteIndexEntry{Simulating Mutations}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  markdown:
    wrap: 72
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

> *Disclaimer:* RACES/rRACES internally implements the
> probability distributions using the C++11 random number
> distribution classes. The standard does not specify their
> algorithms, and the class implementations are left free
> for the compiler. Thus, the simulation output depends on
> the compiler used to compile RACES, and because of that,
> the results reported in this article may differ from
> those obtained by the reader.

rRACES/RACES can simulate genomic mutations on the cells represented in
a `SamplesForest` according to the specified SBS and indel mutational
signatures (see [@alexandrov2020repertoire]).
This process is performed by the class `MutationEngine` which also takes
into account the mutation rate of the simulated species and gives the
chance to dynamically change the signatures.

### Setting Up Mutation Engine

The creation of an object of the type `MutationEngine` requires to download
the reference sequence and the signature files and to build the
corresponding two signature indices. The function `MutationEngine()`
performs all these steps in a single call.

This function is quite flexible and it allows to customize the mutation
engine in many ways. However, the vast majority of rRACES users aim a
standard setup for instance involving the human genome.

Because of this, rRACES provides some predefined setups.
Their list can be obtained by invoking the function `get_mutation_engine_codes()`.

```{r}
library(rRACES)

get_mutation_engine_codes()
```

Let us build a `MutationEngine` by using the `"demo"` setup.

```{r, echo=FALSE,  results='hide', message=FALSE}
unlink("demo", recursive = TRUE)
```

```{r}
# building a mutation engine by using the "demo" setup
m_engine <- MutationEngine(setup_code = "demo")

m_engine
```

The above call creates the directory `demo`,
downloads all the data required by the mutation
engine, and builds it.

```{r}
dir.exists("demo")

list.files("demo")
```

The execution of the function `MutationEngine()` may takes some time, but
it is meant to be performed one for all and, as long as the user does not need to
change the reference genome or the signature files, it is no more required. In this
spirit, any call to the function `MutationEngine` checks whether all the building
process sub-tasks are really needed and, if this is not the case, it avoids to
performs them.

```{r}
# building a mutation engine by using the "demo" set-up configuration
m_engine <- MutationEngine(setup_code = "demo")
```

#### Mutant Genetic Characterization

Once the mutation engine has been built, we can define mutant genotype and
declare species mutation rates.

Let us consider simulation peformed in Section "_Two populations with
epigenetic state_" of `vignette("sampling")`. It involves the two
mutants `A` and
"B". Both of them has two possible epigenetic states, `+` and `-`, leading
to the four the species `A+`, `A-`, `B+`, and `B-`, respectively.
Each of these species has a passenger mutation rate that must be
declated to the mutation engine before labeling the samples forest.
Let 1e-9, 3e-8, 8e-7, and 5e-8 be the passenger SNV rates for the
species `A+`, `A-`, `B+`, and `B-`, respectively. Let the indel
rates be 0 for all the species, but `A+` which has indel rate 1e-8.
Furthermore, let 0, 1e-11, 0, and 0 be the passenger CNA rates of the same
species, respectively.

The two mutants may also be genetically characterized by some
driver mutations.
The driver mutations associated with each of the mutants must occur
in any cell belonging to the mutant itself. Hence, they must be
declared to the mutant engine before the labeling.

The method `MutationEngine$add_mutant()` takes care of all these
declarations.

For the sake of example, let us assume that `A` is characterized by one
driver mutation on chromosome 22, while `B` has three driver mutations on
the same chromosome.

The genetic specification of the mutant `A` can be declared as it follows.

```{r, echo=FALSE,  results='hide', message=FALSE}
m_engine <- MutationEngine(setup_code = "demo")
```

```{r}
# add the mutant "A" characterized by one driver SNV on the allele 1 of
# chromosome 22, one indel deletion on the same chromosome, and two CNAs:
# a deletion on the allele 1 of chromosome 22 and an amplification on a
# random allele of the same chromosome. The mutant has two epigenetic
# states and its species "A+" and "A-" have passenger SNV rates 1e-9 and
# 3e-8, respectively, passenger indel rates 1e-10 and 0, respectively,
# and passenger CNA rates 0 and 1e-11, respectively.
m_engine$add_mutant(mutant_name = "A",
                    passenger_rates = list("+" = c(SNV = 1e-9, indel = 1e-8),
                                           "-" = c(SNV = 3e-8, CNA = 1e-11)),
                    drivers = list(SNV("22", 10510210, "C", allele = 1),
                                   Mutation("22", 15220157, "GTTTTTTTT", "G"),
                                   CNA(type = "A", chr = "22",
                                       chr_pos = 10303470, len = 200000),
                                   CNA("D", "22", 5010000, 200000,
                                       allele = 1)))

m_engine
```

In the above code, the driver CNAs, SNVs, and indels are declared by calling
the functions `CNA()`, `SNV()`, and `Mutation()`, respectively. These function
allow to specify the allele in which the mutation must lay.

As far as the SNVs and indels concern, rRACES provides the users with a more
compact and, sometimes, convenient approach. The mutation engine stores a list
of known driver mutations and labels each of them by a code which can be used
during the mutant declaration. The corresponding data frame can be obtained
by using the method `MutationEngine$get_known_drivers()`.

```{r}
library(dplyr)

m_engine$get_known_drivers() %>% filter(chr == "22")
```

The code of the known driver mutations can be used in place of the full
specification as follows.

```{r}
# add the mutant "B" characterized by two driver SNVs on chromosome 22 (no
# CNA) and two epigenetic states. The first SNV is "NF2 R221*" and it must 
# lay in the allele 1. The remaining SNV is specified by using the SNV function
# as done above. The species "B+" and "B-" have passenger SNV rates 8e-7
# and 5e-8, respectively, and CNA rates 0 for both species.
m_engine$add_mutant("B", list("+" = c(SNV = 8e-7), "-" = c(SNV = 5e-8)),
                    list(list("NF2 R221*", allele = 1),   # the first SNV
                         SNV("22", 12028576, "G")))       # the second SNV

m_engine
```

#### Mutational Exposures

The probability for a mutation to occur depends on both its genomic and
environmental context.

A _signature_ is a mutation probability distribution over mutation contexts or
mutation structure. _SBS (single base substitution) signatures_ provide
for any genomic context (i.e., a triplet of bases) the probability for a SNV
to occur on that context. On the contrary, _ID (indel) signatures_ associates
the probability of an indel to its length and structure
(see [@alexandrov2020repertoire]).

The signature depends on the environmental context and, because of that, more
than one signature may be active at the same time with different probabilities.
An _mutational exposure_ (or _exposure_) is a discrete probability distribution
among signatures.

In order to simulate passeger mutations of a given type, we need to specify a
default exposure for that type. This can be achieved as it follows.

```{r}
# add SNV and indel default exposures. This will be used from simulated time 0
# up to the successive exposure change.
m_engine$add_exposure(coefficients = c(SBS13 = 0.2, SBS1 = 0.8))
m_engine$add_exposure(c(ID2 = 0.6, ID13 = 0.2, ID21 = 0.2))
```

Further exposures can also be defined by specifying an activation time
for each of them, i.e., the new exposures will be used
from the specified time up to the next exposure change.

```{r}
# add a new SNV exposure that will be used from simulated
# time 100 up to the successive exposure change.
m_engine$add_exposure(time = 100, c(SBS5 = 0.3, SBS2 = 0.2, SBS3 = 0.5))

m_engine
```

The default exposures and the simultaneous changes in SNV and ID exposures
can be specified by a sigle call to the function
`MutationEngine$add_exposure()` as it follows.

```{r}
m_engine$add_exposure(time = 120, c(SBS5 = 0.3, SBS2 = 0.2, SBS3 = 0.5,
                                    ID1 = 0.8, ID9 = 0.2))

m_engine
```

#### Passenger CNAs and Tumour Type

The passenger CNAs applied during the simulation depend on tumour type.
The type of tumour can be specified when building the mutation engine
by using the `tumour_type` parameter. For instance, if the
passenger CNAs file used to build the mutation engine contains
some of the data identified in breast carcinoma in the UK,
the specific CNA can be applied during the simulation as it follows.

```{r, eval=FALSE}
m_engine <- MutationEngine(setup_code = "demo",
                                  tumour_type = "BRCA-UK")
```

#### Germline Mutations

rRACES allows users to apply the germline mutations of one of
the subjects available in the germline data provided in building
the mutation engine. This feature will enable users to simulate
a specific cancer type's evolution on an individual with the
desired gender and ethnicity.

The available subjects, together with their sex and ethnicity, can
be obtained by using the method
`[MutationEngine$get_germline_subjects()]`.

```{r}
subjects <- m_engine$get_germline_subjects()

subjects
```

The column `sample` contains the names of the available subjects.
The columns `pop` and `super_pop` report the subjects' population
and super-population codes. The last column,
`gender`, includes the subject gender.

The method `MutationEngine$get_population_descritions()` clarifies
the meaning of the codes reported in the pop columns.

```{r}
m_engine$get_population_descritions()
```

The method `MutationEngine$get_active_germline()` returns the
the active germline subject.

```{r}
m_engine$get_active_germline()
```

Users can change the germline subject using the method
`MutationEngine$set_germline_subject()`.

When a subject is selected for the first time, rRACES builds
a binary representation of the subject genome, saving
it for future use. This step may take a few minutes. However,
all the successive selections of the same subject directly
load the binary file.

```{r}
m_engine$set_germline_subject(subjects[2, "sample"])

m_engine$get_active_germline()
```

### Building Phylogenetic Forests

The configurated mutation engine can be used to label each node in a
samples forest by mutations.

Since the mutation engine has be configurated to deal with the
simulation peformed in Section "_Two populations with
epigenetic state_" of `vignette("sampling")`,
we can load its samples forest from the file `"samples_forest.sff"`
in saved that article.

```{r}
samples_forest <- load_samples_forest("samples_forest.sff")

# place mutations on the sample forest assuming 1000 pre-neoplastic SNVs and
# 500 indels
phylo_forest <- m_engine$place_mutations(samples_forest, 1000, 500)

phylo_forest
```

The phylogenetic forest stores all the mutations labeling the
sampled cells which are represented by the forest leaves.
Users can retrieve such data by using the methods
`PhylogeneticForest$get_sampled_cell_mutations()` and
`PhylogeneticForest$get_sampled_cell_CNAs()`.

```{r}
library(dplyr)

# select the first mutations among all the mutations occuring in
# the genomes of the sampled cells
phylo_forest$get_sampled_cell_mutations() %>% head()

# select the first CNAs among all the mutations occuring in
# the genomes of the sampled cells
phylo_forest$get_sampled_cell_CNAs() %>% head()

# get the sampled cells
sampled_cells <- phylo_forest$get_nodes() %>%
  filter(!is.na(.data$sample))

# show the first of them
sampled_cells %>% head()

# get the identifier of the 3rd cell in `sampled_cells`
cell_id <- sampled_cells[3, 1]

# get the SNVs and the indels of the 3rd cell in `sampled_cells`
phylo_forest$get_sampled_cell_mutations(cell_id) %>% head()

# get the CNAs of the 3rd cell in `sampled_cells`
phylo_forest$get_sampled_cell_CNAs(cell_id) %>% head()
```

The method `PhylogeneticForest$get_germline_mutations()` returns
the SNVs and the indels in the germline.

```{r}
# extract the germline mutation
phylo_forest$get_germline_mutations() %>% head()
```


Users can also identify the cell in which a mutation emerged
even when the cell was not sampled.

```{r}
# select one of the mutations
mutation_row <- phylo_forest$get_sampled_cell_mutations(cell_id)[2, ]

# rebuild the corresponding mutation
mutation <- Mutation(mutation_row["chr"][1, ],
                     mutation_row["chr_pos"][1, ],
                     mutation_row["ref"][1, ],
                     mutation_row["alt"][1, ])

# get the identifier of the oldest cells in which the mutation occurs
phylo_forest$get_first_occurrences(mutation)
```

The exposures used in placing the mutations on the cells in the
phylogenetic forest can be obtained by using the method
`PhylogeneticForest$get_exposures()`.

```{r}
# get the exposures used in placing the mutations
phylo_forest$get_exposures()
```

The details about the SNV and indel signatures adopted during the evolution are
available in the mutation engine and they can be retrieved by using the methods
`MutationEngine$get_SNV_signatures()` and `MutationEngine$get_indel_signatures()`.

```{r}
# get the SNV signatures used in placing the mutations
m_engine$get_SNV_signatures()[1:6, 1:5]

# get the indel signatures used in placing the mutations
m_engine$get_indel_signatures()[1:6, 1:5]
```

The name of the subject whose germline corresponds to
wild-type genome in the phylogenetic forest can be
obtained by the method `Phylogenetic$get_germline_subject()`.

```{r}
phylo_forest$get_germline_subject()
```


### Storing Phylogenetic Forests

As in the case of the samples forests, the phylogenetic forests
can be saved by using the method `PhylogeneticForest$save()` and
load by the function `load_phylogenetic_forest()`.

```{r}
# save the phylogenetic forest in the file "phylo_forest.sff"
phylo_forest$save("phylo_forest.sff")

# loading the saved forest
loaded_phylo_forest <- load_phylogenetic_forest("phylo_forest.sff")

loaded_phylo_forest
```

### Getting and Setting the Reference Genome Path

The phylogenetic forest object contains the reference genome FASTA file path.
The methods `PhylogeneticForest$get_reference_path()` and
`PhylogeneticForest$set_reference_path()` can be used to get the path and
set it, respectively.
```{r}
phylo_forest$get_reference_path()

phylo_forest$set_reference_path("demo/reference.fasta")

phylo_forest$get_reference_path()
```
