---
title: "Sequencing Simulation"
output: rmarkdown::html_vignette
description: >
  How to simulate sample sequencing
vignette: >
  %\VignetteIndexEntry{Sequencing Simulation}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options:
  markdown:
    wrap: 72
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r, echo=FALSE, results='hide', message=FALSE}
unlink("rRACES_SAM", recursive = TRUE)
```

> *Disclaimer:* RACES/rRACES internally implements the
> probability distributions using the C++11 random number
> distribution classes. The standard does not specify their
> algorithms, and the class implementations are left free
> for the compiler. Thus, the simulation output depends on
> the compiler used to compile RACES, and because of that,
> the results reported in this article may differ from
> those obtained by the reader.

Once a phylogenetic forest has been computed, as detailed in
`vignette("mutations")`, rRACES can simulate the sequencing of the
samples in the forest and return the observed data.

### Sequencing Simulation

Let us reload the phylogenetic forest produced in
`vignette("mutations")`.

```{r}
library(rRACES)

phylo_forest <- load_phylogenetic_forest("phylo_forest.sff")

phylo_forest
```

The loaded phylogenetic forest models the cell evolution of
4 different samples: `S_1_1`, `S_1_2`, `S_2_1`, and `S_2_2`.

We can simulate the sequencing of these samples with coverage
of 2.5X by as it follows.

```{r}
# let us simulate a 50x sequencing of the four sample
seq_results <- simulate_seq(phylo_forest, coverage = 30)

# let us load the dplyr library to filter the `simulate_seq` output
library(dplyr)

seq_results$parameters

seq_results$mutations %>% head
```

The function `simulate_seq()` returns named list of two elements:
a data frame whose rows represent the mutations observed in the
simulated reads (name "`mutations`") and the
calling parameters (name "`parameters`").

The number of columns in the data frame "`mutations`" depends on
the number of samples in the phylogenetic forest. The first four
columns describe the mutation and report the chromosome and the
position in the chromosome of the mutation (columns `chr` and
`chr_pos`, respectively), the genetic context in which the
mutation occurs (column `context`), and the new base (column
`alt_base`).
Then, there are three columns for each of the samples: they
contain the number of simulated reads affected by the
mutation, the sequencing coverage of the mutation locus,
and the ratio between these two last columns
(columns `<sample name> occurrences`, `<sample name> coverage`,
and `<sample name> VAF`, respectively).

### Plotting Sequencing Data

rRACES/RACE provides functions to plot the Variant Allele
Frequency (VAF), the B-Allele Frequency (BAF),
and the Depth Ratio (DR) of a sample.

```{r}
library(dplyr)

# filter germinal mutations
f_seq <- seq_results$mutations %>% filter(classes != "germinal")

# plot the VAF of the mutations sequenced on the sample S_2_2
# over the reference genome
plot_VAF(f_seq, sample = "S_2_2")

# plot the BAF of the mutations sequenced on the sample S_2_2
# over the reference genome
plot_BAF(f_seq, sample = "S_2_2")

# plot the DR of the mutations sequenced on the sample S_2_2
# over the reference genome
plot_DR(f_seq, sample = "S_2_2")
```

The VAF and the VAF marginal distributions can be
plot too.

```{r}
# plotting the VAF histogram
plot_VAF_histogram(f_seq, cuts = c(0.02, 1))

# plotting the VAF marginals
plot_VAF_marginals(f_seq, samples = c("S_1_1", "S_1_2", "S_2_2"))
```

The VAF and the VAF marginal plots can be labelled.

```{r}
# plotting the VAF histogram with labels
plot_VAF_histogram(f_seq, labels = f_seq["classes"], cuts = c(0.02, 1))

# plotting the VAF marginals and labelling it
plot_VAF_marginals(f_seq, samples = c("S_1_1", "S_1_2", "S_2_2"),
                   labels = f_seq["classes"])
```

### Saving the Simulated Reads

rRACES/RACE can also save the simulated reads in the SAM format
(see [@10.1093/bioinformatics/btp352]). By setting the optional
parameter `write_SAM` to `TRUE`, the function `simulate_seq()`
creates the directory `rRACES_SAM` and saves the SAM files in it.
Each file is named after one of the reference genome chromosomes
and contains the simulated reads. The reads are split into read
groups corresponding to the collected samples.

```{r, echo=FALSE, results='hide', message=FALSE}
unlink("SAM_0.7", recursive = TRUE)
unlink("rRACES_SAM", recursive = TRUE)
```

```{r}
seq_results <- simulate_seq(phylo_forest, coverage = 2.5, write_SAM = TRUE)

sam_files <- list.files("rRACES_SAM/")

ex_file <- paste("rRACES_SAM/", sam_files[1], sep = "")
for (line in readLines(ex_file, n = 10)) {
  cat(paste(line, "\n"))
}
```

Each SAM file contains the reads produced by simulating the sequencing
of *all* the samples. The command-line tools [`samtools`](http://www.htslib.org) can be used to
split the reads by sample.

```{bash, eval = FALSE}
foo@bar % samtools split  -f "%*_%\!.sam" rRACES_SAM/chr_22.sam
foo@bar % ls chr_22_*
chr_22_S_1_1.sam        chr_22_S_1_2.sam        chr_22_S_2_1.sam        chr_22_S_2_2.sam
```

The resulting files are named after the samples, each
containing the reads of only one sample.

Please refer to the [`samtools split` manual](http://www.htslib.org/doc/samtools-split.html)
for more details.

The `simulate_seq()` parameter `output_dir` set the SAM output directory. See the following
section for usage examples of this parameter.

### Sample Purity

The _sample purity_ represents the concentration of tumour cells in a sample.
More formally, it is the ratio between the number of tumour cells in the sample
and that of all the sample cells, either tumour or normal (i.e., cells
having germline and pre-neoplastic mutations).

rRACES can simulate different sample purity in sequencing
simulations.

```{r}
# simulate the sequencing of a sample in which 70% of the cells are
# tumour cells
seq_results <- simulate_seq(phylo_forest, coverage = 2.5, purity = 0.7,
                            write_SAM = TRUE, output_dir = "SAM_0.7")

sam_files <- list.files("SAM_0.7/")

ex_file <- paste("SAM_0.7/", sam_files[1], sep = "")
for (line in readLines(ex_file, n = 10)) {
  cat(paste(line, "\n"))
}
```

The default value of the parameter `purity` is 1, i.e., `simulate_seq()`
assumes all the cells in the sample to be neoplastic by default.

The function `simulate_normal_seq()` simulates the sequencing
of a sample whose cells have germline and tumoral pre-neoplastic
mutations. The pre-neoplastic mutations can be avoided
by setting the optional parameter `with_preneoplastic`.

### Simulating Sequencing Errors

To simulate sequencing errors, users need to specify
a sequencer to the function `simulate_seq()`.

At the moment, there are only two classes implementing
sequencers: the `ErrorlessIlluminaSequencer` class and
the `BasicIlluminaSequencer` classes.

#### The `ErrorlessIlluminaSequencer` Class

This class models a perfect Illumina sequencer. No sequencing
errors are produced, and all the bases have the maximum
quality.

The following code simulates the sequencing by using this
kind of sequencer.

```{r, eval=FALSE}
# build an error-less Illumina sequencer
no_error_seq <- ErrorlessIlluminaSequencer()

# let us simulate a 2.5x sequencing of the four sample
# on the error-less sequencer
seq_results <- simulate_seq(phylo_forest, sequencer = no_error_seq,
                            coverage = 2.5)
```

#### The `BasicIlluminaSequencer` Class

This class sets a sequencing error rate independent of the
base in which the error occurs, from the position of the base
in the read and from the position of the read on the genome.

```{r, eval=FALSE}
# build a basic Illumina sequencer model in which errors occur
# at rate 4e-3 per base
basic_seq <- BasicIlluminaSequencer(4e-3)

# let us simulate a 2.5x sequencing of the four sample
# on the error-less sequencer
seq_results <- simulate_seq(phylo_forest, sequencer = basic_seq,
                            coverage = 2.5, write_SAM = TRUE)
```

### Chromosome Sequencing

The function `simulate_seq()` allows users to simulate the
sequencing of a selection of the reference chromosomes by using
the parameter `chromosomes`.

```{r, eval=FALSE}
# let us simulate a 2.5x sequencing of the chromosomes 22 and
# X of the four sample
seq_results <- simulate_seq(phylo_forest, chromosomes = c("22", "X"),
                            coverage = 2.5, write_SAM = TRUE)
```

### Updating the SAM Output Directory

Users may want to simulate sequencing in multiple steps,
for instance, by splitting it by chromosome or reaching
the aimed coverage with different simulations.
However, rRACES prevents successive writing in the same
directory by default. The Boolean parameter `update_SAM`
allows multiple writing in the same directory.

```{r}
# the default SAM directory already exists
sam_files <- list.files("rRACES_SAM/")

# since the default save directory already exists, any call to
# `simulate_seq` throws an error
tryCatch(
  {
    simulate_seq(phylo_forest, coverage = 2.5, write_SAM = TRUE)
  },
  error = function(e) {
    print(e)
  }
)

# setting `update_SAM` to TRUE enables successive writing in
# the output directory
seq_results <- simulate_seq(phylo_forest, coverage = 2.5,
                            write_SAM = TRUE, update_SAM = TRUE)
```

When the output directory already contains a SAM file associated
with a chromosome, the alignments on that chromosome are saved
in a file whose name is the first available, having the format
`"chr_{chromosome name}_{natural number}.sam"`.

```{r}
list.files("rRACES_SAM/")
```

```{r, echo=FALSE, results='hide', message=FALSE}
unlink("SAM_0.7", recursive = TRUE)
unlink("rRACES_SAM", recursive = TRUE)
unlink("demo", recursive = TRUE)
unlink("phylo_forest.sff")
unlink("samples_forest.sff")
```
