<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="Represents the phylogenetic forest of the
      cells sampled during the computation. The leaves of
      this forest are the sampled cells.
      This class is analoguous to the class `SamplesForest`, 
      but each node is labelled with the mutations occuring
      for the first time on the cell represented by the node
      itself. Moreover each leaf is also associated with the
      genome mutations occurring in the corresponding cell."><title>The phylogenetic forest of cells in samples. — PhylogeneticForest • rRACES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="The phylogenetic forest of cells in samples. — PhylogeneticForest"><meta property="og:description" content="Represents the phylogenetic forest of the
      cells sampled during the computation. The leaves of
      this forest are the sampled cells.
      This class is analoguous to the class `SamplesForest`, 
      but each node is labelled with the mutations occuring
      for the first time on the cell represented by the node
      itself. Moreover each leaf is also associated with the
      genome mutations occurring in the corresponding cell."><meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/growth_models.html">Growth Models</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>The phylogenetic forest of cells in samples.</h1>
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESR/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
      <div class="d-none name"><code>PhylogeneticForest.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Represents the phylogenetic forest of the
      cells sampled during the computation. The leaves of
      this forest are the sampled cells.
      This class is analoguous to the class `SamplesForest`, 
      but each node is labelled with the mutations occuring
      for the first time on the cell represented by the node
      itself. Moreover each leaf is also associated with the
      genome mutations occurring in the corresponding cell.</p>
    </div>


    <div class="section level2">
    <h2 id="fields">Fields<a class="anchor" aria-label="anchor" href="#fields"></a></h2>
    

<dl><dt><code>get_coalescent_cells</code></dt>
<dd><p>Retrieve most recent common ancestors</p><ul><li><p><em>Parameter:</em> <code>cell_ids</code> - The list of the identifiers of the
              cells whose most recent common ancestors are aimed (optional).</p></li>
<li><p><em>Return:</em> A data frame representing, for each of the identified
        cells, the identified (column "cell_id"), whenever the
        node is not a root, the ancestor identifier (column
        "ancestor"), whenever the node was sampled, i.e., it is
        one of the forest leaves, the name of the sample
        containing the node, (column "sample"), the mutant
        (column "mutant"), the epistate (column "epistate"),
        and the birth time (column "birth_time").</p></li>
</ul></dd>


<dt><code>get_first_occurrences</code></dt>
<dd><p>Gets the identifier of the cell in which a 
       mutation occurs for the first time</p><ul><li><p><em>Parameter:</em> <code>mutation</code> - A mutation being either a 
             SNV or a CNA.</p></li>
<li><p><em>Return:</em> The identifier of the cell in which a mutation 
             occurs for the first time</p></li>
</ul></dd>


<dt><code>get_nodes</code></dt>
<dd><p>Get the forest nodes</p><ul><li><p><em>Return:</em> A data frame representing, for each node
             in the forest, the identified (column "id"),
             whenever the node is not a root, the ancestor
             identifier (column "ancestor"), whenever the node
             was sampled, i.e., it is one of the forest
             leaves, the name of the sample containing the
             node, (column "sample"), the mutant (column
             "mutant"), the epistate (column "epistate"),
             and the birth time (column "birth_time").</p></li>
</ul></dd>


<dt><code>get_sampled_cell_SNVs</code></dt>
<dd><p>Gets a the SNVs of the sampled cells</p><ul><li><p><em>Returns:</em> A data frame reporting `cell_id`, `chromosome`, 
         `chr_pos` (i.e., position in the chromosome), `allele` (in which
         the SNV occurs), `context`, `mutated_base`, and `cause` for each
         SNV in the sampled cell genomes.</p></li>
</ul></dd>


<dt><code>get_samples_info</code></dt>
<dd><p>Retrieve information about the samples</p><ul><li><p><em>Returns:</em> A data frame containing, for each sample collected
        during the simulation, the columns "name", "time", "ymin",
        "xmin", "ymax", "xmax", and  "tumoral cells". "ymin",
        "xmin", "ymax", "xmax" report the boundaries of the sampled
        rectangular region, while "tumoral cells" is the number of
        tumoral cells in the sample.</p></li>
</ul></dd>


<dt><code>get_species_info</code></dt>
<dd><p>Gets the species data</p><ul><li><p><em>Returns:</em> A data frame reporting "mutant" and "epistate"
           for each registered species.</p></li>
</ul></dd>


<dt><code>get_subforest_for</code></dt>
<dd><p>Build a subforest using as leaves some of the original samples</p><ul><li><p><em>Parameter:</em> <code>sample_names</code> - The names of the samples whose cells will be used
        as leaves of the new forest.</p></li>
<li><p><em>Returns:</em> A samples forest built on the samples mentioned in `sample_names`.</p></li>
</ul></dd>


<dt><code>save</code></dt>
<dd><p>Save a phylogenetic forest in a file</p><ul><li><p><em>Parameter:</em> <code>filename</code> - The path of the file in which the phylogenetic 
           forest must be saved.</p></li>
</ul></dd>


</dl></div>

  </main></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Giulio Caravagna, Alberto Casagrande.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

