<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulating wild-type sequencing — simulate_normal_seq • rRACES</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating wild-type sequencing — simulate_normal_seq"><meta name="description" content="This method simulates a wild-type sample sequencing in a
phylogenetic forest. Add the cells in the wild-type sample contains
the germline mutations. The forest pre-neoplastic mutations are also
added to the sample by default. However, they can be avoided by
using the parameter with_preneoplastic."><meta property="og:description" content="This method simulates a wild-type sample sequencing in a
phylogenetic forest. Add the cells in the wild-type sample contains
the germline mutations. The forest pre-neoplastic mutations are also
added to the sample by default. However, they can be avoided by
using the parameter with_preneoplastic."><meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/rRACES.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a></li>
    <li><a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a></li>
    <li><a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a></li>
    <li><a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulating wild-type sequencing</h1>
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACES/R/RcppExports.R" class="external-link"><code>R/RcppExports.R</code></a></small>
      <div class="d-none name"><code>simulate_normal_seq.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This method simulates a wild-type sample sequencing in a
phylogenetic forest. Add the cells in the wild-type sample contains
the germline mutations. The forest pre-neoplastic mutations are also
added to the sample by default. However, they can be avoided by
using the parameter <code>with_preneoplastic</code>.</p>
    </div>


    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-phylo-forest">phylo_forest<a class="anchor" aria-label="anchor" href="#arg-phylo-forest"></a></dt>
<dd><p>A phylogenetic forest.</p></dd>


<dt id="arg-sequencer">sequencer<a class="anchor" aria-label="anchor" href="#arg-sequencer"></a></dt>
<dd><p>The sequencer that performs the sequencing simulation
(default: an <code>ErrorlessIlluminaSequencer</code>).</p></dd>


<dt id="arg-reference-genome">reference_genome<a class="anchor" aria-label="anchor" href="#arg-reference-genome"></a></dt>
<dd><p>The reference genome (default: NULL to use the
mutation engine reference genome).</p></dd>


<dt id="arg-chromosomes">chromosomes<a class="anchor" aria-label="anchor" href="#arg-chromosomes"></a></dt>
<dd><p>The chromosomes that must be considered (default:
<code>NULL</code>, i.e., all the reference chromosomes).</p></dd>


<dt id="arg-coverage">coverage<a class="anchor" aria-label="anchor" href="#arg-coverage"></a></dt>
<dd><p>The sequencing coverage (default: <code>10</code>).</p></dd>


<dt id="arg-read-size">read_size<a class="anchor" aria-label="anchor" href="#arg-read-size"></a></dt>
<dd><p>The read size (default: <code>150</code>).</p></dd>


<dt id="arg-insert-size-mean">insert_size_mean<a class="anchor" aria-label="anchor" href="#arg-insert-size-mean"></a></dt>
<dd><p>The insert size mean. Use 0 for single read
sequencing and any value greater than 0 for pair read sequencing
(default: <code>0</code>).</p></dd>


<dt id="arg-insert-size-stddev">insert_size_stddev<a class="anchor" aria-label="anchor" href="#arg-insert-size-stddev"></a></dt>
<dd><p>The insert size standard deviation.
(default: <code>10</code>).</p></dd>


<dt id="arg-output-dir">output_dir<a class="anchor" aria-label="anchor" href="#arg-output-dir"></a></dt>
<dd><p>The SAM output directory (default:
<code>"rRACES_normal_SAM"</code>).</p></dd>


<dt id="arg-write-sam">write_SAM<a class="anchor" aria-label="anchor" href="#arg-write-sam"></a></dt>
<dd><p>A Boolean flag to enable/disable SAM generation
(default: <code>TRUE</code>).</p></dd>


<dt id="arg-update-sam">update_SAM<a class="anchor" aria-label="anchor" href="#arg-update-sam"></a></dt>
<dd><p>Update the output directory (default: <code>FALSE</code>).</p></dd>


<dt id="arg-with-preneoplastic">with_preneoplastic<a class="anchor" aria-label="anchor" href="#arg-with-preneoplastic"></a></dt>
<dd><p>Add the forest pre-neoplastic mutations
to the sample cells. (default: <code>TRUE</code>).</p></dd>


<dt id="arg-filename-prefix">filename_prefix<a class="anchor" aria-label="anchor" href="#arg-filename-prefix"></a></dt>
<dd><p>The prefix of the output SAM file name
(default: <code>"chr_"</code>).</p></dd>


<dt id="arg-template-name-prefix">template_name_prefix<a class="anchor" aria-label="anchor" href="#arg-template-name-prefix"></a></dt>
<dd><p>The template name prefix (default:
<code>"r"</code>).</p></dd>


<dt id="arg-include-non-sequenced-mutations">include_non_sequenced_mutations<a class="anchor" aria-label="anchor" href="#arg-include-non-sequenced-mutations"></a></dt>
<dd><p>A Boolean flag to include
in the resulting data frame also the mutations that are not
covered by any of the simulated reads, but occur to one of the
samples at least (default: <code>FALSE</code>).</p></dd>


<dt id="arg-seed">seed<a class="anchor" aria-label="anchor" href="#arg-seed"></a></dt>
<dd><p>The random seed for the internal random generator
(optional).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A named list of two elements: the sequencing output data
frame (name "<code>mutations</code>") and the calling parameters
(name "<code>parameters</code>").</p>
<p>The sequencing output data frame reports, for each of the
observed SNVs and indels, the chromosome and the position in
which it occurs (columns <code>chr</code> and <code>chr_pos</code>),
the SNV reference base, the alterate base, the causes,
and the classes of the SNV (columns <code>ref_base</code>, <code>alt_base</code>,
<code>causes</code>, and <code>classes</code>, respectively). Moreover, for each
of the sequencied samples <code>normal_sample</code>, the returned
data frame contains three columns: the number of reads in
which the corresponding SNV occurs (column
<code>normal_sample.occurrences</code>), the coverage of the SNV
locus (column <code>normal_sample.coverage</code>), and the
corresponding VAF (column <code>normal_sample.VAF</code>).</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code>BasicIlluminaSequencer</code> and
<code>ErrorlessIlluminaSequencer</code> as sequencer types, and
<code><a href="../articles/sequencing.html">vignette("sequencing")</a></code> for usage examples</p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

