<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="How to customize the mutation engine and its indices
">
<title>Customizing mutation engine • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Customizing mutation engine">
<meta property="og:description" content="How to customize the mutation engine and its indices
">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/index.html">More articles...</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Customizing mutation engine</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/custom_engine.Rmd" class="external-link"><code>vignettes/custom_engine.Rmd</code></a></small>
      <div class="d-none name"><code>custom_engine.Rmd</code></div>
    </div>

    
    
<p>rRACES/RACES can simulate genomic mutations on the cells represented
in a <code>SamplesForest</code> according to the specified SBS and indel
mutational signatures (see <code>vignette["mutations"]</code>). This
task is achieved by the mutation engine which uses two indices on the
reference genome to place the most probable mutations according to
signatures: the <em>context index</em> and the <em>repeted sequence
index</em>.</p>
<p>The function <code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> builds the mutation
engine and these two indices.</p>
<div class="section level2">
<h2 id="customizing-the-mutation-engine">Customizing the mutation engine<a class="anchor" aria-label="anchor" href="#customizing-the-mutation-engine"></a>
</h2>
<p>While the most convient way to invoke
<code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> is by using the parameter
<code>setup_code</code> (see <code>vignette["mutations"]</code>), the
are cases in which a custom set-up is required. In these cases, the
function <code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> can be called by
specifying the name of the set-up directory, the path or URL of the
reference sequence, the signature files, the driver mutation file, the
passenger CNAs file, and the germline data directory.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st">"rRACES"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">reference_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://ftp.ensembl.org/pub/grch37/release-111/"</span>,</span>
<span>                        <span class="st">"fasta/homo_sapiens/dna/Homo_sapiens.GRCh37."</span>,</span>
<span>                        <span class="st">"dna.chromosome.22.fa.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sbs_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://cancer.sanger.ac.uk/signatures/documents/2123/"</span>,</span>
<span>                  <span class="st">"COSMIC_v3.4_SBS_GRCh37.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">indel_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://cancer.sanger.ac.uk/signatures/documents/"</span>,</span>
<span>                    <span class="st">"2121/COSMIC_v3.4_ID_GRCh37.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">drivers_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/"</span>,</span>
<span>                      <span class="st">"caravagnalab/rRACES/main/inst/extdata/"</span>,</span>
<span>                      <span class="st">"driver_mutations_hg19.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">passenger_cnas_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/"</span>,</span>
<span>                             <span class="st">"caravagnalab/rRACES/main/inst/extdata/"</span>,</span>
<span>                             <span class="st">"passenger_CNAs_hg19.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">germline_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://www.dropbox.com/scl/fi/g9oloxkip18tr1r"</span>,</span>
<span>                       <span class="st">"m6wjve/germline_data_demo.tar.gz?rlkey=15jshul"</span>,</span>
<span>                       <span class="st">"d3bqgyfcs7fa0bzqeo&amp;dl=1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># build a mutation engine and place all the files in the directory "Test"</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="st">"Test"</span>,</span>
<span>                                  reference_src <span class="op">=</span> <span class="va">reference_url</span>,</span>
<span>                                  SBS_signatures_src <span class="op">=</span> <span class="va">sbs_url</span>,</span>
<span>                                  indel_signatures_src <span class="op">=</span> <span class="va">indel_url</span>,</span>
<span>                                  drivers_src <span class="op">=</span> <span class="va">drivers_url</span>,</span>
<span>                                  passenger_CNAs_src <span class="op">=</span> <span class="va">passenger_cnas_url</span>,</span>
<span>                                  germline_src <span class="op">=</span> <span class="va">germline_url</span><span class="op">)</span></span>
<span><span class="co">#&gt; Downloading reference genome...</span></span>
<span><span class="co">#&gt; Reference genome downloaded</span></span>
<span><span class="co">#&gt; Decompressing reference file...done</span></span>
<span><span class="co">#&gt; Downloading SBS file...</span></span>
<span><span class="co">#&gt; SBS file downloaded</span></span>
<span><span class="co">#&gt; Downloading indel file...</span></span>
<span><span class="co">#&gt; indel file downloaded</span></span>
<span><span class="co">#&gt; Downloading driver mutation file...</span></span>
<span><span class="co">#&gt; Driver mutation file downloaded</span></span>
<span><span class="co">#&gt; Downloading passenger CNAs file...</span></span>
<span><span class="co">#&gt; Passenger CNAs file downloaded</span></span>
<span><span class="co">#&gt; Downloading germline mutations...</span></span>
<span><span class="co">#&gt; Germline mutations downloaded</span></span>
<span><span class="co">#&gt; Building context index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█████████████████-----------------------] 40% [00m:01s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:02s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:02s] Context index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving context index [████████████████████████████████████████] 100% [00m:00s] Context index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt; Building repeated sequence index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:03s] Processing chr. 22 [█---------------------------------------] 0% [00m:05s] Processing chr. 22 [█---------------------------------------] 0% [00m:06s] Processing chr. 22 [█---------------------------------------] 0% [00m:08s] Processing chr. 22 [█---------------------------------------] 0% [00m:09s] Processing chr. 22 [█---------------------------------------] 0% [00m:13s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:13s] RS index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving RS index [███████████████████████-----------------] 56% [00m:01s] Saving RS index [████████████████████████████████████████] 100% [00m:01s] RS index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving germline [████████████████████████████████████████] 100% [00m:00s] Germline saved</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"Test"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
<div class="section level3">
<h3 id="the-context-index">The Context Index<a class="anchor" aria-label="anchor" href="#the-context-index"></a>
</h3>
<p>The SBS signatures fixes, for every possible context (i.e., a triplet
of consetutive nucleotides), the probability for a SNV to occur in such
a context (see <span class="citation">(Alexandrov et al. 2020)</span>).
Hence, the function <code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> builds a
context index of the reference sequence to place SNVs on the forest
cells. The complete context index of a sequence whose size is up to
4Gbps takes 4 times the length of the sequence itself both in memory and
on the disk because 4 bytes per nucleotide are required to store the
context position. For instance, the complete contex index of a human
genome takes about 12GBytes. In order to avoid such requirement in
memory and disk space, <code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> allows to
sample the reference genome contexts and to stores only some of the them
in the context index. This is achieved by the optional parameter
<code>context_sampling</code> which specifies how many occurences of the
same context must be identified before adding one of them to the context
index. The larger the number of context sampling, the larger the context
index. On the other side, the lower the number of context sampling, the
lower the number of sites in the refernce genome that can be affected by
simulated mutations. The <code>context_sampling</code> is set to 100 by
default, but it can be specified during the
<code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> call.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Downloading reference genome...</span></span>
<span><span class="co">#&gt; Reference genome downloaded</span></span>
<span><span class="co">#&gt; Decompressing reference file...done</span></span>
<span><span class="co">#&gt; Downloading SBS file...</span></span>
<span><span class="co">#&gt; SBS file downloaded</span></span>
<span><span class="co">#&gt; Downloading indel file...</span></span>
<span><span class="co">#&gt; indel file downloaded</span></span>
<span><span class="co">#&gt; Downloading driver mutation file...</span></span>
<span><span class="co">#&gt; Driver mutation file downloaded</span></span>
<span><span class="co">#&gt; Downloading passenger CNAs file...</span></span>
<span><span class="co">#&gt; Passenger CNAs file downloaded</span></span>
<span><span class="co">#&gt; Downloading germline mutations...</span></span>
<span><span class="co">#&gt; Germline mutations downloaded</span></span>
<span><span class="co">#&gt; Building context index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█████████████████-----------------------] 40% [00m:01s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:02s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:02s] Context index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving context index [████████████████████████████████████████] 100% [00m:00s] Context index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt; Building repeated sequence index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:04s] Processing chr. 22 [█---------------------------------------] 0% [00m:05s] Processing chr. 22 [█---------------------------------------] 0% [00m:07s] Processing chr. 22 [█---------------------------------------] 0% [00m:08s] Processing chr. 22 [█---------------------------------------] 0% [00m:10s] Processing chr. 22 [█---------------------------------------] 0% [00m:14s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:14s] RS index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving RS index [███████████████████████-----------------] 56% [00m:01s] Saving RS index [████████████████████████████████████████] 100% [00m:01s] RS index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving germline [████████████████████████████████████████] 100% [00m:00s] Germline saved</span></span>
<span></span>
<span><span class="co"># get the size of the context index when `context_sampling` is 100</span></span>
<span><span class="fu">utils</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">format.object_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.size</a></span><span class="op">(</span><span class="st">"demo/context_index_100.cif"</span><span class="op">)</span>, <span class="st">"auto"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "1.3 Mb"</span></span></code></pre></div>
<p>Let us rebuild the context index sampling one context every 50.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># building a mutation engine by using the "demo" set-up configuration</span></span>
<span><span class="co"># and setting context_sampling to 50</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span>, context_sampling <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">#&gt; Building context index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█████████████████-----------------------] 40% [00m:01s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:02s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:03s] Context index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving context index [████████████████████████████████████████] 100% [00m:00s] Context index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading RS index [█████████████████-----------------------] 41% [00m:01s] Loading RS index [████████████████████████████████--------] 79% [00m:02s] Loading RS index [████████████████████████████████████████] 100% [00m:02s] RS index loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span></code></pre></div>
<p>Above call uses the already downloaded data in the directory
<code>"demo"</code> to build the context and produces the file
<code>"demo/context_index_50.cif"</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the size of the context index when `context_sampling` is 50</span></span>
<span><span class="fu">utils</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">format.object_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.size</a></span><span class="op">(</span><span class="st">"demo/context_index_50.cif"</span><span class="op">)</span>, <span class="st">"auto"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "2.7 Mb"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-repeated-sequence-index">The Repeated Sequence Index<a class="anchor" aria-label="anchor" href="#the-repeated-sequence-index"></a>
</h3>
<p>Analoguosly to the SBS signatures, the ID (indel) signatures fixes,
for some kinds of repeated sequences, the probability for an indel to
occur in them (see <span class="citation">(Alexandrov et al.
2020)</span>). Thus, the <code>MutationEngine</code> function also
builds an index for the repeated sequences in the reference. As in the
case of the SBS signatures, the ID signatures can be memory eager. To
mitigate the memory requirement, the function
<code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> accepts as parameters the maximum
size of the considered indel (the parameter <code>max_index_size</code>)
and the maximum number of reference position stored per repetition type
(the parameter <code>max_repetition_storage</code>). The larger the two
parameters, the larger the repetition index. As far the building time
concerns, the larger <code>max_index_size</code>, the longer takes the
index construction. On the contrary, <code>max_repetition_storage</code>
does not directly affects the computation time. By default, the two
parameters are set to 50 and 500000, respectively. However, their values
can be specified during the <code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code>
call.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the size of the repeated index when the default parameter values</span></span>
<span><span class="co"># are used</span></span>
<span><span class="fu">utils</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">format.object_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.size</a></span><span class="op">(</span><span class="st">"demo/rs_index_50_500000.rsif"</span><span class="op">)</span>, <span class="st">"auto"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "212.5 Mb"</span></span></code></pre></div>
<p>Let us rebuild the context index sampling one context every 50.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># building a mutation engine by using the "demo" set-up configuration</span></span>
<span><span class="co"># and setting max_repetition_storage to 10,000,000</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span>,</span>
<span>                                  max_repetition_storage <span class="op">=</span> <span class="fl">10000000</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading context index [████████████████████████████████████████] 100% [00m:00s] Context index loaded</span></span>
<span><span class="co">#&gt; Building repeated sequence index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:03s] Processing chr. 22 [█---------------------------------------] 0% [00m:04s] Processing chr. 22 [█---------------------------------------] 0% [00m:06s] Processing chr. 22 [█---------------------------------------] 0% [00m:07s] Processing chr. 22 [█---------------------------------------] 0% [00m:09s] Processing chr. 22 [█---------------------------------------] 0% [00m:11s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:11s] RS index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving RS index [█---------------------------------------] 1% [00m:01s] Saving RS index [██████████------------------------------] 23% [00m:02s] Saving RS index [████████████████████--------------------] 48% [00m:03s] Saving RS index [█████████████████████████████-----------] 72% [00m:04s] Saving RS index [███████████████████████████████████████-] 97% [00m:05s] Saving RS index [████████████████████████████████████████] 100% [00m:05s] RS index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span></code></pre></div>
<p>Above call uses the already downloaded data in the directory
<code>"demo"</code> to build the repeated sequence index and produces
the file <code>"demo/rs_index_50_10000000.rsif"</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the size of the context index when `context_sampling` is 50</span></span>
<span><span class="fu">utils</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">format.object_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.size</a></span><span class="op">(</span><span class="st">"demo/rs_index_50_10000000.rsif"</span><span class="op">)</span>,</span>
<span>                                     <span class="st">"auto"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "732.1 Mb"</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-alexandrov2020repertoire" class="csl-entry">
Alexandrov, Ludmil B, Jaegil Kim, Nicholas J Haradhvala, Mi Ni Huang,
Alvin Wei Tian Ng, Yang Wu, Arnoud Boot, et al. 2020. <span>“The
Repertoire of Mutational Signatures in Human Cancer.”</span>
<em>Nature</em> 578 (7793): 94–101.
</div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
