<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Simulating Mutations • ProCESS</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating Mutations">
<meta name="description" content="How to simulate mutations on sampled cells
">
<meta property="og:description" content="How to simulate mutations on sampled cells
">
<meta property="og:image" content="https://caravagnalab.github.io/ProCESS/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ProCESS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/ProCESS.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a></li>
    <li><a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a></li>
    <li><a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a></li>
    <li><a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/caravagnalab/ProCESS/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulating Mutations</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/ProCESS/vignettes/mutations.Rmd" class="external-link"><code>vignettes/mutations.Rmd</code></a></small>
      <div class="d-none name"><code>mutations.Rmd</code></div>
    </div>

    
    
<blockquote>
<p><em>Disclaimer:</em> RACES/ProCESS internally implements the
probability distributions using the C++11 random number distribution
classes. The standard does not specify their algorithms, and the class
implementations are left free for the compiler. Thus, the simulation
output depends on the compiler used to compile RACES, and because of
that, the results reported in this article may differ from those
obtained by the reader.</p>
</blockquote>
<p>ProCESS/RACES can simulate genomic mutations on the cells represented
in a <code>SamplesForest</code> according to the specified SBS and indel
mutational signatures (see <span class="citation">(Alexandrov et al.
2020)</span>). This process is performed by the class
<code>MutationEngine</code> which also takes into account the mutation
rate of the simulated species and gives the chance to dynamically change
the signatures.</p>
<div class="section level3">
<h3 id="setting-up-mutation-engine">Setting Up Mutation Engine<a class="anchor" aria-label="anchor" href="#setting-up-mutation-engine"></a>
</h3>
<p>The creation of an object of the type <code>MutationEngine</code>
requires to download the reference sequence and the signature files and
to build the corresponding two signature indices. The function
<code><a href="../reference/MutationEngine.html">MutationEngine()</a></code> performs all these steps in a single
call.</p>
<p>This function is quite flexible and it allows to customize the
mutation engine in many ways. However, the vast majority of ProCESS
users aim a standard setup for instance involving the human genome.</p>
<p>Because of this, ProCESS provides some predefined setups. Their list
can be obtained by invoking the function
<code><a href="../reference/get_mutation_engine_codes.html">get_mutation_engine_codes()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://caravagnalab.github.io/ProCESS">ProCESS</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/get_mutation_engine_codes.html">get_mutation_engine_codes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;     name            description</span></span>
<span><span class="co">#&gt; 1 GRCh37  Homo sapiens (GRCh37)</span></span>
<span><span class="co">#&gt; 2 GRCh38  Homo sapiens (GRCh38)</span></span>
<span><span class="co">#&gt; 3   demo A demonstrative set-up</span></span></code></pre></div>
<p>Let us build a <code>MutationEngine</code> by using the
<code>"demo"</code> setup.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># building a mutation engine by using the "demo" setup</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MutationEngine.html">MutationEngine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Downloading reference genome...</span></span>
<span><span class="co">#&gt; Reference genome downloaded</span></span>
<span><span class="co">#&gt; Decompressing reference file...done</span></span>
<span><span class="co">#&gt; Downloading signature files...</span></span>
<span><span class="co">#&gt; Signature file downloaded</span></span>
<span><span class="co">#&gt; Downloading driver mutation file...</span></span>
<span><span class="co">#&gt; Driver mutation file downloaded</span></span>
<span><span class="co">#&gt; Downloading passenger CNAs file...</span></span>
<span><span class="co">#&gt; Passenger CNAs file downloaded</span></span>
<span><span class="co">#&gt; Downloading germline mutations...</span></span>
<span><span class="co">#&gt; Germline mutations downloaded</span></span>
<span><span class="co">#&gt; Building context index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█████████████---------------------------] 32% [00m:01s] Processing chr. 22 [███████████████████████-----------------] 57% [00m:02s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:04s] Processing chr. 22 [████████████████████████████████████----] 89% [00m:05s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:05s] Context index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving context index [████████████████████████████████████████] 100% [00m:00s] Context index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt; Building repeated sequence index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [█---------------------------------------] 0% [00m:02s] Processing chr. 22 [█---------------------------------------] 0% [00m:10s] Processing chr. 22 [█---------------------------------------] 0% [00m:14s] Processing chr. 22 [█---------------------------------------] 0% [00m:16s] Processing chr. 22 [█---------------------------------------] 0% [00m:17s] Processing chr. 22 [█---------------------------------------] 0% [00m:19s] Processing chr. 22 [█---------------------------------------] 0% [00m:26s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:26s] RS index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving RS index [█---------------------------------------] 0% [00m:01s] Saving RS index [█---------------------------------------] 0% [00m:02s] Saving RS index [█---------------------------------------] 0% [00m:03s] Saving RS index [█████████-------------------------------] 20% [00m:04s] Saving RS index [███████████████████---------------------] 47% [00m:05s] Saving RS index [████████████████████████████████████----] 88% [00m:06s] Saving RS indexdone</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:06s] RS index saved</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [███████████████████████████████████-----] 86% [00m:01s] Loading germline [████████████████████████████████████████] 100% [00m:01s] Germline loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving germline [████████████████████████████████████████] 100% [00m:00s] Germline saved</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span></code></pre></div>
<p>The above call creates the directory <code>demo</code>, downloads all
the data required by the mutation engine, and builds it.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "context_index_100.cif"     "drivers.txt"              </span></span>
<span><span class="co">#&gt;  [3] "germline_data"             "germline_data_demo.tar.gz"</span></span>
<span><span class="co">#&gt;  [5] "indel_signatures.txt"      "passenger_CNAs.txt"       </span></span>
<span><span class="co">#&gt;  [7] "reference.fasta"           "rs_index_50_500000.rsif"  </span></span>
<span><span class="co">#&gt;  [9] "SBS_signatures.txt"        "sources.csv"</span></span></code></pre></div>
<p>The execution of the function <code><a href="../reference/MutationEngine.html">MutationEngine()</a></code> may takes
some time, but it is meant to be performed one for all and, as long as
the user does not need to change the reference genome or the signature
files, it is no more required. In this spirit, any call to the function
<code>MutationEngine</code> checks whether all the building process
sub-tasks are really needed and, if this is not the case, it avoids to
performs them.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># building a mutation engine by using the "demo" set-up configuration</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MutationEngine.html">MutationEngine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Downloading signature files...</span></span>
<span><span class="co">#&gt; Signature file downloaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading context index [████████████████████████████████████████] 100% [00m:00s] Context index loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading RS index [███████████-----------------------------] 25% [00m:01s] Loading RS index [█████████████████████-------------------] 52% [00m:02s] Loading RS index [███████████████████████████████---------] 76% [00m:03s] Loading RS index [████████████████████████████████████████] 100% [00m:04s] RS index loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span></code></pre></div>
<div class="section level4">
<h4 id="mutant-genetic-characterization">Mutant Genetic Characterization<a class="anchor" aria-label="anchor" href="#mutant-genetic-characterization"></a>
</h4>
<p>Once the mutation engine has been built, we can define mutant
genotype and declare species mutation rates.</p>
<p>Let us consider simulation peformed in Section “<em>Two populations
with epigenetic state</em>” of <code><a href="../articles/sampling.html">vignette("sampling")</a></code>. It
involves the two mutants <code>A</code> and “B”. Both of them has two
possible epigenetic states, <code>+</code> and <code>-</code>, leading
to the four the species <code>A+</code>, <code>A-</code>,
<code>B+</code>, and <code>B-</code>, respectively. Each of these
species has a passenger mutation rate that must be declated to the
mutation engine before labeling the samples forest. Let 1e-9, 3e-8,
8e-7, and 5e-8 be the passenger SNV rates for the species
<code>A+</code>, <code>A-</code>, <code>B+</code>, and <code>B-</code>,
respectively. Let the indel rates be 0 for all the species, but
<code>A+</code> which has indel rate 1e-8. Furthermore, let 0, 1e-11, 0,
and 0 be the passenger CNA rates of the same species, respectively.</p>
<p>The two mutants may also be genetically characterized by some driver
mutations. The driver mutations associated with each of the mutants must
occur in any cell belonging to the mutant itself. Hence, they must be
declared to the mutant engine before the labeling.</p>
<p>The method <code>MutationEngine$add_mutant()</code> takes care of all
these declarations.</p>
<p>For the sake of example, let us assume that <code>A</code> is
characterized by one driver mutation on chromosome 22, while
<code>B</code> has three driver mutations on the same chromosome.</p>
<p>The genetic specification of the mutant <code>A</code> can be
declared as it follows.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add the mutant "A" characterized by one driver SNV on the allele 1 of</span></span>
<span><span class="co"># chromosome 22, one indel deletion on the same chromosome, two CNAs (a</span></span>
<span><span class="co"># deletion on the allele 1 of chromosome 22 and an amplification on a random</span></span>
<span><span class="co"># allele of the same chromosome) and, finally, a whole genome doubling event</span></span>
<span><span class="co"># (WGD). The mutant has two epigenetic states and its species "A+" and "A-"</span></span>
<span><span class="co"># have passenger SNV rates 1e-9 and 3e-8, respectively, passenger indel rates</span></span>
<span><span class="co"># 1e-10 and 0, respectively, and passenger CNA rates 0 and 1e-11, respectively.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>mutant_name <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>                    passenger_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">1e-9</span>, indel <span class="op">=</span> <span class="fl">1e-8</span><span class="op">)</span>,</span>
<span>                                           <span class="st">"-"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">3e-8</span>, CNA <span class="op">=</span> <span class="fl">1e-11</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    drivers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="../reference/SNV.html">SNV</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">10510210</span>, <span class="st">"C"</span>, allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="../reference/Mutation.html">Mutation</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">15220157</span>, <span class="st">"GTTTTTTTT"</span>, <span class="st">"G"</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="../reference/CNA.html">CNA</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"A"</span>, chr <span class="op">=</span> <span class="st">"22"</span>,</span>
<span>                                       chr_pos <span class="op">=</span> <span class="fl">10303470</span>, len <span class="op">=</span> <span class="fl">200000</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="../reference/CNA.html">CNA</a></span><span class="op">(</span><span class="st">"D"</span>, <span class="st">"22"</span>, <span class="fl">5010000</span>, <span class="fl">200000</span>,</span>
<span>                                       allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   <span class="va">WGD</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Retrieving "A" SNVs [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:02s] Reading 22 [████████████████████████████████████████] 100% [00m:02s] "A" SNVs retrieved</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09, indel: 1e-08}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A":</span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        Whole genome duplication</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span></code></pre></div>
<p>In the above code, the driver CNAs, SNVs, and indels are declared by
calling the functions <code><a href="../reference/CNA.html">CNA()</a></code>, <code><a href="../reference/SNV.html">SNV()</a></code>, and
<code><a href="../reference/Mutation.html">Mutation()</a></code>, respectively. These function allow to specify
the allele in which the mutation must lay. The mutant <code>A</code> is
also characterized by a <em>whole genome doubling</em>
(<code>WGD</code>): a genomic event that simultaneously duplicates all
the chromosome alleles.</p>
<p>The driver mutations are applied according to their specification
order. For instance, the following two pieces of code specify two
different genomic characterizations for the same mutant
<code>E</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span><span class="st">"E"</span>, <span class="va">passenger_rates</span>,</span>
<span>                    drivers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="../reference/SNV.html">SNV</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">10510210</span>, <span class="st">"C"</span>, allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   <span class="va">WGD</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span><span class="st">"E"</span>, <span class="va">passenger_rates</span>,</span>
<span>                    drivers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">WGD</span>,</span>
<span>                                   <span class="fu"><a href="../reference/SNV.html">SNV</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">10510210</span>, <span class="st">"C"</span>, allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The former snippet places an SNV and, afterwards, duplicates all the
alleles producing a genome in which two alleles contain the placed SNV.
Instead, the latter snippet requires a whole genome doubling event and,
then, places the SNV yielding a single occurrence of the SNV in the
final genome.</p>
<p>As far as the SNVs and indels concern, ProCESS provides the users
with a more compact and, sometimes, convenient approach. The mutation
engine stores a list of known driver mutations and labels each of them
by a code which can be used during the mutant declaration. The
corresponding data frame can be obtained by using the method
<code>MutationEngine$get_known_drivers()</code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_known_drivers</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"22"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr     from       to ref alt mutation_type tumour_type driver_gene</span></span>
<span><span class="co">#&gt; 1  22 30057302 30057302   C   T           SNV          OV         NF2</span></span>
<span><span class="co">#&gt; 2  22 30057302 30057302   C   T           SNV          OV         NF2</span></span>
<span><span class="co">#&gt;   driver_code</span></span>
<span><span class="co">#&gt; 1   NF2 R221*</span></span>
<span><span class="co">#&gt; 2   NF2 R262*</span></span></code></pre></div>
<p>The code of the known driver mutations can be used in place of the
full specification as follows.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add the mutant "B" characterized by two driver SNVs on chromosome 22 (no</span></span>
<span><span class="co"># CNA) and two epigenetic states. The first SNV is "NF2 R221*" and it must </span></span>
<span><span class="co"># lay in the allele 1. The remaining SNV is specified by using the SNV function</span></span>
<span><span class="co"># as done above. The species "B+" and "B-" have passenger SNV rates 8e-8</span></span>
<span><span class="co"># and 5e-8, respectively, and CNA rates 0 for both species.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span><span class="st">"B"</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">8e-8</span><span class="op">)</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">5e-8</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"NF2 R221*"</span>, allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,   <span class="co"># the first SNV</span></span>
<span>                         <span class="fu"><a href="../reference/SNV.html">SNV</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">12028576</span>, <span class="st">"G"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>       <span class="co"># the second SNV</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Retrieving "B" SNVs [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:02s] Reading 22 [████████████████████████████████████████] 100% [00m:02s] "B" SNVs retrieved</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09, indel: 1e-08}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt;    "B+": {SNV: 8e-08}</span></span>
<span><span class="co">#&gt;    "B-": {SNV: 5e-08}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A":</span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        Whole genome duplication</span></span>
<span><span class="co">#&gt;    "B":</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(12028576)[N&gt;G](allele: random)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="mutational-exposures">Mutational Exposures<a class="anchor" aria-label="anchor" href="#mutational-exposures"></a>
</h4>
<p>The probability for a mutation to occur depends on both its genomic
and environmental context.</p>
<p>A <em>signature</em> is a mutation probability distribution over
mutation contexts or mutation structure. <em>SBS (single base
substitution) signatures</em> provide for any genomic context (i.e., a
triplet of bases) the probability for a SNV to occur on that context. On
the contrary, <em>ID (indel) signatures</em> associates the probability
of an indel to its length and structure (see <span class="citation">(Alexandrov et al. 2020)</span>).</p>
<p>The signature depends on the environmental context and, because of
that, more than one signature may be active at the same time with
different probabilities. An <em>mutational exposure</em> (or
<em>exposure</em>) is a discrete probability distribution among
signatures.</p>
<p>In order to simulate passeger mutations of a given type, we need to
specify a default exposure for that type. This can be achieved as it
follows.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add SNV and indel default exposures. This will be used from simulated time 0</span></span>
<span><span class="co"># up to the successive exposure change.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span>coefficients <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SBS13 <span class="op">=</span> <span class="fl">0.2</span>, SBS1 <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ID2 <span class="op">=</span> <span class="fl">0.6</span>, ID13 <span class="op">=</span> <span class="fl">0.2</span>, ID21 <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Further exposures can also be defined by specifying an activation
time for each of them, i.e., the new exposures will be used from the
specified time up to the next exposure change.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add a new SNV exposure that will be used from simulated</span></span>
<span><span class="co"># time 100 up to the successive exposure change.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span>time <span class="op">=</span> <span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SBS5 <span class="op">=</span> <span class="fl">0.3</span>, SBS2 <span class="op">=</span> <span class="fl">0.2</span>, SBS3 <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09, indel: 1e-08}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt;    "B+": {SNV: 8e-08}</span></span>
<span><span class="co">#&gt;    "B-": {SNV: 5e-08}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A":</span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        Whole genome duplication</span></span>
<span><span class="co">#&gt;    "B":</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(12028576)[N&gt;G](allele: random)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, 100[: {"SBS1": 0.8, "SBS13": 0.2}</span></span>
<span><span class="co">#&gt;      [100, ∞[: {"SBS2": 0.2, "SBS3": 0.5, "SBS5": 0.3}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, ∞[: {"ID13": 0.2, "ID2": 0.6, "ID21": 0.2}</span></span></code></pre></div>
<p>The default exposures and the simultaneous changes in SNV and ID
exposures can be specified by a sigle call to the function
<code>MutationEngine$add_exposure()</code> as it follows.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span>time <span class="op">=</span> <span class="fl">120</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SBS5 <span class="op">=</span> <span class="fl">0.3</span>, SBS2 <span class="op">=</span> <span class="fl">0.2</span>, SBS3 <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                    ID1 <span class="op">=</span> <span class="fl">0.8</span>, ID9 <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09, indel: 1e-08}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt;    "B+": {SNV: 8e-08}</span></span>
<span><span class="co">#&gt;    "B-": {SNV: 5e-08}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A":</span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        Whole genome duplication</span></span>
<span><span class="co">#&gt;    "B":</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(12028576)[N&gt;G](allele: random)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, 100[: {"SBS1": 0.8, "SBS13": 0.2}</span></span>
<span><span class="co">#&gt;      [100, 120[: {"SBS2": 0.2, "SBS3": 0.5, "SBS5": 0.3}</span></span>
<span><span class="co">#&gt;      [120, ∞[: {"SBS2": 0.2, "SBS3": 0.5, "SBS5": 0.3}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, 120[: {"ID13": 0.2, "ID2": 0.6, "ID21": 0.2}</span></span>
<span><span class="co">#&gt;      [120, ∞[: {"ID1": 0.8, "ID9": 0.2}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="passenger-cnas-and-tumour-type">Passenger CNAs and Tumour Type<a class="anchor" aria-label="anchor" href="#passenger-cnas-and-tumour-type"></a>
</h4>
<p>The passenger CNAs applied during the simulation depend on tumour
type. The type of tumour can be specified when building the mutation
engine by using the parameters <code>tumour_type</code> or
<code>tumour_study</code>. For instance, if the passenger CNAs file used
to build the mutation engine contains some of the data identified in
breast carcinoma in the UK, the specific CNA can be applied during the
simulation as it follows.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/MutationEngine.html">MutationEngine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span>,</span>
<span>                           tumour_type <span class="op">=</span> <span class="st">"BRCA"</span>,</span>
<span>                           tumour_study <span class="op">=</span> <span class="st">"UK"</span><span class="op">)</span></span></code></pre></div>
<p>The complete list of available tumour types and studies in a set-up
can be obtained by using the function
<code><a href="../reference/get_available_tumours_in.html">get_available_tumours_in()</a></code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/get_available_tumours_in.html">get_available_tumours_in</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Downloading signature files...</span></span>
<span><span class="co">#&gt; Signature file downloaded</span></span>
<span><span class="co">#&gt;   type study</span></span>
<span><span class="co">#&gt; 1 BLCA    US</span></span>
<span><span class="co">#&gt; 2 BOCA    UK</span></span>
<span><span class="co">#&gt; 3 BRCA    EU</span></span>
<span><span class="co">#&gt; 4 BRCA    UK</span></span>
<span><span class="co">#&gt; 5 BRCA    US</span></span>
<span><span class="co">#&gt; 6 BTCA    SG</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="germline-mutations">Germline Mutations<a class="anchor" aria-label="anchor" href="#germline-mutations"></a>
</h4>
<p>ProCESS allows users to apply the germline mutations of one of the
subjects available in the germline data provided in building the
mutation engine. This feature will enable users to simulate a specific
cancer type’s evolution on an individual with the desired gender and
ethnicity.</p>
<p>The available subjects, together with their sex and ethnicity, can be
obtained by using the method
<code>[MutationEngine$get_germline_subjects()]</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subjects</span> <span class="op">&lt;-</span> <span class="va">m_engine</span><span class="op">$</span><span class="fu">get_germline_subjects</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">subjects</span></span>
<span><span class="co">#&gt;    sample pop super_pop gender</span></span>
<span><span class="co">#&gt; 1 NA18941 JPT       EAS female</span></span>
<span><span class="co">#&gt; 2 NA20513 TSI       EUR   male</span></span></code></pre></div>
<p>The column <code>sample</code> contains the names of the available
subjects. The columns <code>pop</code> and <code>super_pop</code> report
the subjects’ population and super-population codes. The last column,
<code>gender</code>, includes the subject gender.</p>
<p>The method <code>MutationEngine$get_population_descriptions()</code>
clarifies the meaning of the codes reported in the pop columns.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_population_descriptions</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    code          description</span></span>
<span><span class="co">#&gt; 1   CHB          Han Chinese</span></span>
<span><span class="co">#&gt; 2   JPT             Japanese</span></span>
<span><span class="co">#&gt; 3   CHS Southern Han Chinese</span></span>
<span><span class="co">#&gt; 4   CDX          Dai Chinese</span></span>
<span><span class="co">#&gt; 5   KHV      Kinh Vietnamese</span></span>
<span><span class="co">#&gt; 6   CHD       Denver Chinese</span></span>
<span><span class="co">#&gt; 7   CEU                 CEPH</span></span>
<span><span class="co">#&gt; 8   TSI               Tuscan</span></span>
<span><span class="co">#&gt; 9   GBR              British</span></span>
<span><span class="co">#&gt; 10  FIN              Finnish</span></span>
<span><span class="co">#&gt; 11  IBS              Spanish</span></span>
<span><span class="co">#&gt; 12  YRI               Yoruba</span></span>
<span><span class="co">#&gt; 13  LWK                Luhya</span></span>
<span><span class="co">#&gt; 14  GWD              Gambian</span></span>
<span><span class="co">#&gt; 15  MSL                Mende</span></span>
<span><span class="co">#&gt; 16  ESN                 Esan</span></span>
<span><span class="co">#&gt; 17  ASW  African-American SW</span></span>
<span><span class="co">#&gt; 18  ACB    African-Caribbean</span></span>
<span><span class="co">#&gt; 19  MXL     Mexican-American</span></span>
<span><span class="co">#&gt; 20  PUR         Puerto Rican</span></span>
<span><span class="co">#&gt; 21  CLM            Colombian</span></span>
<span><span class="co">#&gt; 22  PEL             Peruvian</span></span>
<span><span class="co">#&gt; 23  GIH             Gujarati</span></span>
<span><span class="co">#&gt; 24  PJL              Punjabi</span></span>
<span><span class="co">#&gt; 25  BEB              Bengali</span></span>
<span><span class="co">#&gt; 26  STU           Sri Lankan</span></span>
<span><span class="co">#&gt; 27  ITU               Indian</span></span>
<span><span class="co">#&gt;                                                      long.description</span></span>
<span><span class="co">#&gt; 1                                       Han Chinese in Beijing, China</span></span>
<span><span class="co">#&gt; 2                                            Japanese in Tokyo, Japan</span></span>
<span><span class="co">#&gt; 3                                                   Han Chinese South</span></span>
<span><span class="co">#&gt; 4                                 Chinese Dai in Xishuangbanna, China</span></span>
<span><span class="co">#&gt; 5                                   Kinh in Ho Chi Minh City, Vietnam</span></span>
<span><span class="co">#&gt; 6                          Chinese in Denver, Colorado (pilot 3 only)</span></span>
<span><span class="co">#&gt; 7  Utah residents (CEPH) with Northern and Western European ancestry </span></span>
<span><span class="co">#&gt; 8                                                  Toscani in Italia </span></span>
<span><span class="co">#&gt; 9                                    British in England and Scotland </span></span>
<span><span class="co">#&gt; 10                                                Finnish in Finland </span></span>
<span><span class="co">#&gt; 11                                      Iberian populations in Spain </span></span>
<span><span class="co">#&gt; 12                                          Yoruba in Ibadan, Nigeria</span></span>
<span><span class="co">#&gt; 13                                             Luhya in Webuye, Kenya</span></span>
<span><span class="co">#&gt; 14                           Gambian in Western Division, The Gambia </span></span>
<span><span class="co">#&gt; 15                                              Mende in Sierra Leone</span></span>
<span><span class="co">#&gt; 16                                                    Esan in Nigeria</span></span>
<span><span class="co">#&gt; 17                                  African Ancestry in Southwest US </span></span>
<span><span class="co">#&gt; 18                                      African Caribbean in Barbados</span></span>
<span><span class="co">#&gt; 19                        Mexican Ancestry in Los Angeles, California</span></span>
<span><span class="co">#&gt; 20                                        Puerto Rican in Puerto Rico</span></span>
<span><span class="co">#&gt; 21                                    Colombian in Medellin, Colombia</span></span>
<span><span class="co">#&gt; 22                                             Peruvian in Lima, Peru</span></span>
<span><span class="co">#&gt; 23                                     Gujarati Indian in Houston, TX</span></span>
<span><span class="co">#&gt; 24                                        Punjabi in Lahore, Pakistan</span></span>
<span><span class="co">#&gt; 25                                              Bengali in Bangladesh</span></span>
<span><span class="co">#&gt; 26                                         Sri Lankan Tamil in the UK</span></span>
<span><span class="co">#&gt; 27                                            Indian Telugu in the UK</span></span></code></pre></div>
<p>The method <code>MutationEngine$get_active_germline()</code> returns
the the active germline subject.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_active_germline</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    sample pop super_pop gender</span></span>
<span><span class="co">#&gt; 1 NA18941 JPT       EAS female</span></span></code></pre></div>
<p>Users can change the germline subject using the method
<code>MutationEngine$set_germline_subject()</code>.</p>
<p>When a subject is selected for the first time, ProCESS builds a
binary representation of the subject genome, saving it for future use.
This step may take a few minutes. However, all the successive selections
of the same subject directly load the binary file.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">set_germline_subject</span><span class="op">(</span><span class="va">subjects</span><span class="op">[</span><span class="fl">2</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving germline [████████████████████████████████████████] 100% [00m:00s] Germline saved</span></span>
<span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_active_germline</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    sample pop super_pop gender</span></span>
<span><span class="co">#&gt; 1 NA20513 TSI       EUR   male</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="building-phylogenetic-forests">Building Phylogenetic Forests<a class="anchor" aria-label="anchor" href="#building-phylogenetic-forests"></a>
</h3>
<p>The configurated mutation engine can be used to label each node in a
samples forest by mutations.</p>
<p>Since the mutation engine has be configurated to deal with the
simulation peformed in Section “<em>Two populations with epigenetic
state</em>” of <code><a href="../articles/sampling.html">vignette("sampling")</a></code>, we can load its
samples forest from the file <code>"samples_forest.sff"</code> in saved
that article.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples_forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_samples_forest.html">load_samples_forest</a></span><span class="op">(</span><span class="st">"samples_forest.sff"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># place mutations on the sample forest assuming 1000 pre-neoplastic SNVs and</span></span>
<span><span class="co"># 500 indels</span></span>
<span><span class="va">phylo_forest</span> <span class="op">&lt;-</span> <span class="va">m_engine</span><span class="op">$</span><span class="fu">place_mutations</span><span class="op">(</span><span class="va">samples_forest</span>, <span class="fl">1000</span>, <span class="fl">500</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Placing mutations [███████---------------------------------] 15% [00m:01s] Placing mutations [█████████████---------------------------] 30% [00m:02s] Placing mutations [███████████████████---------------------] 45% [00m:03s] Placing mutations [█████████████████████████---------------] 60% [00m:04s] Placing mutations [███████████████████████████████---------] 75% [00m:05s] Placing mutations [███████████████████████████████████-----] 87% [00m:06s] Placing mutations [██████████████████████████████████████--] 94% [00m:07s] Placing mutations [████████████████████████████████████████] 100% [00m:08s] Mutations placed</span></span>
<span></span>
<span><span class="va">phylo_forest</span></span>
<span><span class="co">#&gt; PhylogeneticForest</span></span>
<span><span class="co">#&gt;   # of trees: 1</span></span>
<span><span class="co">#&gt;   # of nodes: 18845</span></span>
<span><span class="co">#&gt;   # of leaves: 5396</span></span>
<span><span class="co">#&gt;   samples: {"S_1_1", "S_1_2", "S_2_1", "S_2_2"}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # of emerged SNVs and indels: 112932</span></span>
<span><span class="co">#&gt;   # of emerged CNAs: 33</span></span></code></pre></div>
<p>The phylogenetic forest stores all the mutations labeling the sampled
cells which are represented by the forest leaves. Users can retrieve
such data by using the methods
<code>PhylogeneticForest$get_sampled_cell_mutations()</code> and
<code>PhylogeneticForest$get_sampled_cell_CNAs()</code>.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select the first mutations among all the mutations occuring in</span></span>
<span><span class="co"># the genomes of the sampled cells</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id chr  chr_pos allele       ref alt  type cause          class</span></span>
<span><span class="co">#&gt; 1    2790  22 15220157      0 GTTTTTTTT   G indel     A         driver</span></span>
<span><span class="co">#&gt; 2    2790  22 16077875      0   AGAGAGA   A indel   ID1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 3    2790  22 16130833      0         A   T   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 4    2790  22 16161146      0         A   G   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 5    2790  22 16179449      0         C   G   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 6    2790  22 16239255      0         C   A   SNV  SBS1 pre-neoplastic</span></span>
<span></span>
<span><span class="co"># select the first CNAs among all the mutations occuring in</span></span>
<span><span class="co"># the genomes of the sampled cells</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_CNAs</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id type chr    begin      end allele src.allele  class</span></span>
<span><span class="co">#&gt; 1    2790    A  22 10303470 10503469      2          0 driver</span></span>
<span><span class="co">#&gt; 2    2790    D  22  5010000  5209999      1         NA driver</span></span>
<span><span class="co">#&gt; 3    5000    A  22 10303470 10503469      2          0 driver</span></span>
<span><span class="co">#&gt; 4    5000    D  22  5010000  5209999      1         NA driver</span></span>
<span><span class="co">#&gt; 5    8774    A  22 10303470 10503469      2          0 driver</span></span>
<span><span class="co">#&gt; 6    8774    D  22  5010000  5209999      1         NA driver</span></span>
<span></span>
<span><span class="co"># get the sampled cells</span></span>
<span><span class="va">sampled_cells</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_nodes</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show the first of them</span></span>
<span><span class="va">sampled_cells</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id ancestor mutant epistate sample birth_time</span></span>
<span><span class="co">#&gt; 1    2790     2694      A        -  S_1_1   127.8854</span></span>
<span><span class="co">#&gt; 2    5000     4671      A        -  S_1_2   152.6275</span></span>
<span><span class="co">#&gt; 3    8774     8274      A        -  S_1_1   180.3081</span></span>
<span><span class="co">#&gt; 4    9651     8069      A        -  S_1_1   185.6496</span></span>
<span><span class="co">#&gt; 5   11533     7757      A        -  S_1_1   195.7661</span></span>
<span><span class="co">#&gt; 6   12905    12464      A        -  S_1_1   202.1142</span></span>
<span></span>
<span><span class="co"># get the identifier of the 3rd cell in `sampled_cells`</span></span>
<span><span class="va">cell_id</span> <span class="op">&lt;-</span> <span class="va">sampled_cells</span><span class="op">[</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># get the SNVs and the indels of the 3rd cell in `sampled_cells`</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id chr  chr_pos allele       ref alt  type cause          class</span></span>
<span><span class="co">#&gt; 1    8774  22 15220157      0 GTTTTTTTT   G indel     A         driver</span></span>
<span><span class="co">#&gt; 2    8774  22 16077875      0   AGAGAGA   A indel   ID1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 3    8774  22 16130833      0         A   T   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 4    8774  22 16161146      0         A   G   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 5    8774  22 16179449      0         C   G   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 6    8774  22 16239255      0         C   A   SNV  SBS1 pre-neoplastic</span></span>
<span></span>
<span><span class="co"># get the CNAs of the 3rd cell in `sampled_cells`</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_CNAs</span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id type chr    begin      end allele src.allele  class</span></span>
<span><span class="co">#&gt; 1    8774    A  22 10303470 10503469      2          0 driver</span></span>
<span><span class="co">#&gt; 2    8774    D  22  5010000  5209999      1         NA driver</span></span></code></pre></div>
<p>The method <code>PhylogeneticForest$get_samples_info()</code> is
analogous to <code>DescendantForest$get_samples_info()</code>: it
returns a dataframe containing information about the forest samples.
However, the former method adds two columns to the result of the latter:
“<code>DNA_quantity</code>” and “<code>equivalent_normal_cells</code>”.
The former column reports the overall quantity of tumoral DNA in the
sample, i.e., the sum of the lengths of all the alleles in the
corresponding sample. This quantity is a natural number, nevertheless,
it is stored as a real number as it usually usually exceeds the largest
natural number that can be natively represented by R. The column
“<code>equivalent_normal_cells</code>” instead contains the number of
normal cells that contain as much DNA as the sample tumoral cells.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_samples_info</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    name id xmin ymin xmax ymax tumour_cells tumour_cells_in_bbox     time</span></span>
<span><span class="co">#&gt; 1 S_1_1  0  480  480  530  530         2588                 2588 365.9279</span></span>
<span><span class="co">#&gt; 2 S_1_2  1  500  500  550  550         1632                 1632 365.9279</span></span>
<span><span class="co">#&gt; 3 S_2_1  2  370  498  394  522          572                  572 421.2143</span></span>
<span><span class="co">#&gt; 4 S_2_2  3  370  373  394  397          604                  604 421.2143</span></span>
<span><span class="co">#&gt;   DNA_quantity equivalent_normal_cells</span></span>
<span><span class="co">#&gt; 1 531246233181                5177.378</span></span>
<span><span class="co">#&gt; 2 335394547347                3268.662</span></span>
<span><span class="co">#&gt; 3 112299538631                1094.440</span></span>
<span><span class="co">#&gt; 4 123951831456                1208.000</span></span></code></pre></div>
<p>The method <code>PhylogeneticForest$get_germline_mutations()</code>
returns the SNVs and the indels in the germline.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the germline mutation</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_germline_mutations</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr  chr_pos allele ref   alt  type cause    class</span></span>
<span><span class="co">#&gt; 1  22 16051493      0   G     A   SNV       germinal</span></span>
<span><span class="co">#&gt; 2  22 16052167      0   A AAAAC indel       germinal</span></span>
<span><span class="co">#&gt; 3  22 16053659      0   A     C   SNV       germinal</span></span>
<span><span class="co">#&gt; 4  22 16054740      0   A     G   SNV       germinal</span></span>
<span><span class="co">#&gt; 5  22 16055942      0   C     T   SNV       germinal</span></span>
<span><span class="co">#&gt; 6  22 16058070      0   A     G   SNV       germinal</span></span></code></pre></div>
<p>Users can also identify the cell in which a mutation emerged even
when the cell was not sampled.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select one of the mutations</span></span>
<span><span class="va">mutation_row</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># rebuild the corresponding mutation</span></span>
<span><span class="va">mutation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Mutation.html">Mutation</a></span><span class="op">(</span><span class="va">mutation_row</span><span class="op">[</span><span class="st">"chr"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>,</span>
<span>                     <span class="va">mutation_row</span><span class="op">[</span><span class="st">"chr_pos"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>,</span>
<span>                     <span class="va">mutation_row</span><span class="op">[</span><span class="st">"ref"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>,</span>
<span>                     <span class="va">mutation_row</span><span class="op">[</span><span class="st">"alt"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the identifier of the oldest cells in which the mutation occurs</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_first_occurrences</span><span class="op">(</span><span class="va">mutation</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>The exposures used in placing the mutations on the cells in the
phylogenetic forest can be obtained by using the method
<code>PhylogeneticForest$get_exposures()</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the exposures used in placing the mutations</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_exposures</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    time signature exposure  type</span></span>
<span><span class="co">#&gt; 1     0      SBS1      0.8   SNV</span></span>
<span><span class="co">#&gt; 2     0     SBS13      0.2   SNV</span></span>
<span><span class="co">#&gt; 3   100      SBS2      0.2   SNV</span></span>
<span><span class="co">#&gt; 4   100      SBS3      0.5   SNV</span></span>
<span><span class="co">#&gt; 5   100      SBS5      0.3   SNV</span></span>
<span><span class="co">#&gt; 6   120      SBS2      0.2   SNV</span></span>
<span><span class="co">#&gt; 7   120      SBS3      0.5   SNV</span></span>
<span><span class="co">#&gt; 8   120      SBS5      0.3   SNV</span></span>
<span><span class="co">#&gt; 9     0      ID13      0.2 indel</span></span>
<span><span class="co">#&gt; 10    0       ID2      0.6 indel</span></span>
<span><span class="co">#&gt; 11    0      ID21      0.2 indel</span></span>
<span><span class="co">#&gt; 12  120       ID1      0.8 indel</span></span>
<span><span class="co">#&gt; 13  120       ID9      0.2 indel</span></span></code></pre></div>
<p>The method
<code>MutationEngine$get_bulk_allelic_fragmentation()</code> returns a
dataframe reporting the allelic type per genome fragment.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the name of the first sample</span></span>
<span><span class="va">sample_name</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_samples_info</span><span class="op">(</span><span class="op">)</span><span class="op">[[</span><span class="st">"name"</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># print the bulk allelic fragmentation</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_bulk_allelic_fragmentation</span><span class="op">(</span><span class="va">sample_name</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr   begin      end major minor       ratio</span></span>
<span><span class="co">#&gt; 1  22       1  5009999     2     1 0.001545595</span></span>
<span><span class="co">#&gt; 2  22       1  5009999     2     2 0.998454405</span></span>
<span><span class="co">#&gt; 3  22 5010000  5209999     1     0 0.001545595</span></span>
<span><span class="co">#&gt; 4  22 5010000  5209999     2     0 0.998454405</span></span>
<span><span class="co">#&gt; 5  22 5210000 10303469     2     1 0.001545595</span></span>
<span><span class="co">#&gt; 6  22 5210000 10303469     2     2 0.998454405</span></span></code></pre></div>
<p>Instead, the method
<code>MutationEngine$get_cell_allelic_fragmentation()</code> returns the
allelic fragmentation per cell.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># print the cell allelic fragmentation</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_cell_allelic_fragmentation</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id chr    begin      end major minor</span></span>
<span><span class="co">#&gt; 1    2790  22        1  5009999     2     2</span></span>
<span><span class="co">#&gt; 2    2790  22  5010000  5209999     2     0</span></span>
<span><span class="co">#&gt; 3    2790  22  5210000 10303469     2     2</span></span>
<span><span class="co">#&gt; 4    2790  22 10303470 10503469     4     2</span></span>
<span><span class="co">#&gt; 5    2790  22 10503470 51304566     2     2</span></span>
<span><span class="co">#&gt; 6    5000  22        1  5009999     2     2</span></span></code></pre></div>
<p>The details about the SNV and indel signatures adopted during the
evolution are available in the mutation engine and they can be retrieved
by using the methods <code>MutationEngine$get_SNV_signatures()</code>
and <code>MutationEngine$get_indel_signatures()</code>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the SNV signatures used in placing the mutations</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_SNV_signatures</span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      Type       SBS1       SBS2       SBS3       SBS4</span></span>
<span><span class="co">#&gt; 1 A[C&gt;A]A 0.01041667 0.01041667 0.01041667 0.01041667</span></span>
<span><span class="co">#&gt; 2 A[C&gt;A]C 0.01041667 0.01041667 0.01041667 0.01041667</span></span>
<span><span class="co">#&gt; 3 A[C&gt;A]G 0.01041667 0.01041667 0.01041667 0.01041667</span></span>
<span><span class="co">#&gt; 4 A[C&gt;A]T 0.01041667 0.01041667 0.01041667 0.01041667</span></span>
<span><span class="co">#&gt; 5 A[C&gt;G]A 0.01041667 0.01041667 0.01041667 0.01041667</span></span>
<span><span class="co">#&gt; 6 A[C&gt;G]C 0.01041667 0.01041667 0.01041667 0.01041667</span></span>
<span></span>
<span><span class="co"># get the indel signatures used in placing the mutations</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_indel_signatures</span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Type        ID1        ID2        ID3        ID4</span></span>
<span><span class="co">#&gt; 1 1:Del:C:0 0.01204819 0.01204819 0.01204819 0.01204819</span></span>
<span><span class="co">#&gt; 2 1:Del:C:1 0.01204819 0.01204819 0.01204819 0.01204819</span></span>
<span><span class="co">#&gt; 3 1:Del:C:2 0.01204819 0.01204819 0.01204819 0.01204819</span></span>
<span><span class="co">#&gt; 4 1:Del:C:3 0.01204819 0.01204819 0.01204819 0.01204819</span></span>
<span><span class="co">#&gt; 5 1:Del:C:4 0.01204819 0.01204819 0.01204819 0.01204819</span></span>
<span><span class="co">#&gt; 6 1:Del:C:5 0.01204819 0.01204819 0.01204819 0.01204819</span></span></code></pre></div>
<p>Finally, the data of the subject whose germline corresponds to
wild-type genome in the phylogenetic forest can be obtained by the
method <code>Phylogenetic$get_germline_subject()</code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_germline_subject</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    sample pop super_pop gender</span></span>
<span><span class="co">#&gt; 1 NA20513 TSI       EUR   male</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="storing-phylogenetic-forests">Storing Phylogenetic Forests<a class="anchor" aria-label="anchor" href="#storing-phylogenetic-forests"></a>
</h3>
<p>As in the case of the samples forests, the phylogenetic forests can
be saved by using the method <code>PhylogeneticForest$save()</code> and
load by the function <code><a href="../reference/load_phylogenetic_forest.html">load_phylogenetic_forest()</a></code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save the phylogenetic forest in the file "phylo_forest.sff"</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">save</span><span class="op">(</span><span class="st">"phylo_forest.sff"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving forest [█---------------------------------------] 0% [00m:01s] Saving forest [█---------------------------------------] 0% [00m:02s] Saving forest [█---------------------------------------] 0% [00m:03s] Saving forest [█---------------------------------------] 0% [00m:04s] Saving forest [█---------------------------------------] 0% [00m:05s] Saving forest [█---------------------------------------] 0% [00m:06s] Saving forest [█---------------------------------------] 0% [00m:07s] Saving forest [█---------------------------------------] 0% [00m:08s] Saving forest [█---------------------------------------] 0% [00m:09s] Saving forest [█---------------------------------------] 0% [00m:10s] Saving forest [█---------------------------------------] 0% [00m:11s] Saving forest [█---------------------------------------] 0% [00m:12s] Saving forest [█---------------------------------------] 0% [00m:13s] Saving forest [█---------------------------------------] 0% [00m:14s] Saving forest [█---------------------------------------] 0% [00m:15s] Saving forest [█████-----------------------------------] 11% [00m:16s] Saving forest [████████--------------------------------] 18% [00m:17s] Saving forest [██████████------------------------------] 24% [00m:18s] Saving forest [████████████----------------------------] 28% [00m:19s] Saving forest [███████████████-------------------------] 35% [00m:20s] Saving forest [█████████████████-----------------------] 41% [00m:21s] Saving forest [████████████████████--------------------] 48% [00m:22s] Saving forest [███████████████████████-----------------] 55% [00m:23s] Saving forest [█████████████████████████---------------] 62% [00m:24s] Saving forest [████████████████████████████------------] 69% [00m:25s] Saving forest [███████████████████████████████---------] 76% [00m:26s] Saving forest [██████████████████████████████████------] 83% [00m:27s] Saving forest [█████████████████████████████████████---] 90% [00m:28s] Saving forest [███████████████████████████████████████-] 97% [00m:30s] Saving forest [████████████████████████████████████████] 100% [00m:30s] Forest saved</span></span>
<span></span>
<span><span class="co"># loading the saved forest</span></span>
<span><span class="va">loaded_phylo_forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_phylogenetic_forest.html">load_phylogenetic_forest</a></span><span class="op">(</span><span class="st">"phylo_forest.sff"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading forest [███-------------------------------------] 7% [00m:01s] Loading forest [█████-----------------------------------] 11% [00m:02s] Loading forest [███████---------------------------------] 15% [00m:03s] Loading forest [████████--------------------------------] 19% [00m:04s] Loading forest [██████████------------------------------] 23% [00m:05s] Loading forest [███████████-----------------------------] 27% [00m:06s] Loading forest [█████████████---------------------------] 30% [00m:08s] Loading forest [██████████████--------------------------] 34% [00m:09s] Loading forest [████████████████------------------------] 38% [00m:10s] Loading forest [█████████████████-----------------------] 42% [00m:11s] Loading forest [███████████████████---------------------] 46% [00m:12s] Loading forest [█████████████████████-------------------] 50% [00m:13s] Loading forest [██████████████████████------------------] 54% [00m:14s] Loading forest [████████████████████████----------------] 58% [00m:15s] Loading forest [█████████████████████████---------------] 62% [00m:16s] Loading forest [███████████████████████████-------------] 66% [00m:17s] Loading forest [█████████████████████████████-----------] 70% [00m:18s] Loading forest [██████████████████████████████----------] 74% [00m:19s] Loading forest [████████████████████████████████--------] 78% [00m:20s] Loading forest [█████████████████████████████████-------] 82% [00m:21s] Loading forest [███████████████████████████████████-----] 86% [00m:22s] Loading forest [█████████████████████████████████████---] 90% [00m:23s] Loading forest [██████████████████████████████████████--] 94% [00m:24s] Loading forest [████████████████████████████████████████] 98% [00m:25s] Loading forest [████████████████████████████████████████] 100% [00m:26s] Forest loaded</span></span>
<span></span>
<span><span class="va">loaded_phylo_forest</span></span>
<span><span class="co">#&gt; PhylogeneticForest</span></span>
<span><span class="co">#&gt;   # of trees: 1</span></span>
<span><span class="co">#&gt;   # of nodes: 18845</span></span>
<span><span class="co">#&gt;   # of leaves: 5396</span></span>
<span><span class="co">#&gt;   samples: {"S_1_1", "S_1_2", "S_2_1", "S_2_2"}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # of emerged SNVs and indels: 112932</span></span>
<span><span class="co">#&gt;   # of emerged CNAs: 33</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="getting-and-setting-the-reference-genome-path">Getting and Setting the Reference Genome Path<a class="anchor" aria-label="anchor" href="#getting-and-setting-the-reference-genome-path"></a>
</h3>
<p>The phylogenetic forest object contains the reference genome FASTA
file path. The methods
<code>PhylogeneticForest$get_reference_path()</code> and
<code>PhylogeneticForest$set_reference_path()</code> can be used to get
the path and set it, respectively.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_reference_path</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Users/alberto/Library/CloudStorage/Dropbox/Lavoro/Code/ProCESS/ProCESS/vignettes/demo/reference.fasta"</span></span>
<span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">set_reference_path</span><span class="op">(</span><span class="st">"demo/reference.fasta"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_reference_path</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "/Users/alberto/Library/CloudStorage/Dropbox/Lavoro/Code/ProCESS/ProCESS/vignettes/demo/reference.fasta"</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-alexandrov2020repertoire" class="csl-entry">
Alexandrov, Ludmil B, Jaegil Kim, Nicholas J Haradhvala, Mi Ni Huang,
Alvin Wei Tian Ng, Yang Wu, Arnoud Boot, et al. 2020. <span>“The
Repertoire of Mutational Signatures in Human Cancer.”</span>
<em>Nature</em> 578 (7793): 94–101.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
