<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rRACES">
<title>Simulating Mutations • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulating Mutations">
<meta property="og:description" content="rRACES">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/growth_models.html">Growth Models</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
    <a class="dropdown-item" href="../articles/SCOUT.html">SCOUT</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Simulating Mutations</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/mutations.Rmd" class="external-link"><code>vignettes/mutations.Rmd</code></a></small>
      <div class="d-none name"><code>mutations.Rmd</code></div>
    </div>

    
    
<p>rRACES/RACES can simulate genomic mutations on the cells represented
in a <code>SamplesForest</code> according to a specified mutational
signature of single base substitutions (SBS) (see <span class="citation">(Alexandrov et al. 2020)</span>). This process is
performed by the class <code>MutationEngine</code> which also takes into
account the mutation rate of the simulated species and gives the chance
to dynamically change the SBS signature.</p>
<div class="section level3">
<h3 id="setting-up-mutation-engine">Setting Up Mutation Engine<a class="anchor" aria-label="anchor" href="#setting-up-mutation-engine"></a>
</h3>
<p>The creation of an object of the type <code>MutationEngine</code>
requires to download the reference sequence and the corresponding SBS
file and to build the corresponding context index. The function
<code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> performs all these steps in a
single call.</p>
<p>This function is quite flexible and it allows to customize the
mutation engine in many ways. However, the vast majority of rRACES users
aim a standard setup for instance involving the human genome.</p>
<p>Because of this, rRACES provides some predefined setups. Their list
can be obtained by invoking the function
<code><a href="../reference/get_mutation_engine_codes.html">get_mutation_engine_codes()</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rRACES</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/get_mutation_engine_codes.html">get_mutation_engine_codes</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;     name           description</span></span>
<span><span class="co">#&gt; 1 GRCh37 Homo sapiens (GRCh37)</span></span>
<span><span class="co">#&gt; 2 GRCh38 Homo sapiens (GRCh38)</span></span>
<span><span class="co">#&gt; 3   demo  A demostative set-up</span></span></code></pre></div>
<p>Let us build a <code>MutationEngine</code> by using the
<code>"demo"</code> setup.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># building a mutation engine by using the "demo" setup</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Downloading reference genome...</span></span>
<span><span class="co">#&gt; Reference genome downloaded</span></span>
<span><span class="co">#&gt; Decompressing reference file...done</span></span>
<span><span class="co">#&gt; Downloading SBS file...</span></span>
<span><span class="co">#&gt; SBS file downloaded</span></span>
<span><span class="co">#&gt; Downloading indel file...</span></span>
<span><span class="co">#&gt; indel file downloaded</span></span>
<span><span class="co">#&gt; Downloading driver mutation file...</span></span>
<span><span class="co">#&gt; Driver mutation file downloaded</span></span>
<span><span class="co">#&gt; Downloading passenger CNAs file...</span></span>
<span><span class="co">#&gt; Passenger CNAs file downloaded</span></span>
<span><span class="co">#&gt; Downloading germline mutations...</span></span>
<span><span class="co">#&gt; Germline mutations downloaded</span></span>
<span><span class="co">#&gt; Building context index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█████████████████-----------------------] 40% [00m:01s] Processing chr. 22 [█████████████████████████████████-------] 81% [00m:02s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:02s] Context index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving context index [████████████████████████████████████████] 100% [00m:00s] Context index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt; Building repeated sequence index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:03s] Processing chr. 22 [█---------------------------------------] 0% [00m:04s] Processing chr. 22 [█---------------------------------------] 0% [00m:05s] Processing chr. 22 [█---------------------------------------] 0% [00m:07s] Processing chr. 22 [█---------------------------------------] 0% [00m:08s] Processing chr. 22 [█---------------------------------------] 0% [00m:11s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:12s] RS index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving RS index [███████████████████████-----------------] 57% [00m:01s] Saving RS index [████████████████████████████████████████] 100% [00m:01s] RS index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving germline [████████████████████████████████████████] 100% [00m:00s] Germline saved</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span></code></pre></div>
<p>The above call creates the directory <code>demo</code>, downloads all
the data required by the mutation engine, and builds it.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "context_index_100.cif"     "drivers.txt"              </span></span>
<span><span class="co">#&gt;  [3] "germline_data"             "germline_data_demo.tar.gz"</span></span>
<span><span class="co">#&gt;  [5] "indel_signatures.txt"      "passenger_CNAs.txt"       </span></span>
<span><span class="co">#&gt;  [7] "reference.fasta"           "rs_index_50_500000.rsif"  </span></span>
<span><span class="co">#&gt;  [9] "SBS_signatures.txt"        "sources.csv"</span></span></code></pre></div>
<p>The execution of the function <code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code>
may takes some time, but it is meant to be performed one for all and, as
long as the user does not need to change the reference genome or the SBS
file, it is no more required. In this spirit, any call to the function
<code>MutationEngine</code> checks whether all the building process
sub-tasks are really needed and, if this is not the case, it avoids to
performs them.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># building a mutation engine by using the "demo" set-up configuration</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading context index [████████████████████████████████████████] 100% [00m:00s] Context index loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading RS index [████████████████████--------------------] 48% [00m:01s] Loading RS index [█████████████████████████████████████---] 92% [00m:02s] Loading RS index [████████████████████████████████████████] 100% [00m:02s] RS index loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span></code></pre></div>
<div class="section level4">
<h4 id="context-index-sampling">Context Index Sampling<a class="anchor" aria-label="anchor" href="#context-index-sampling"></a>
</h4>
<p>The <code>MutationEngine</code> function also builds a context index
of the reference sequence. The complete context index of a sequence
whose size is up to 4Gbps takes 4 times the length of the sequence
itself both in memory and on the disk. For instance, the complete contex
index of a human genome takes about 12GBytes. In order to avoid such
requirement in memory and disk space, <code>MutationEngine</code> allows
to sample the reference genome contexts and to stores only some of the
them in the context index. This is achieved by the optional parameter
<code>context_sampling</code> which specifies how many occurences of the
same context must be identified before adding one of them to the context
index. The larger the number of context sampling, the larger the context
index. On the other side, the lower the number of context sampling, the
lower the number of sites in the refernce genome that can be affected by
simulated mutations. The <code>context_sampling</code> is set to 100 by
default, but it can be specified during the <code>MutationEngine</code>
call as the last parameter.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the size of the context index when `context_sampling` is 100</span></span>
<span><span class="fu">utils</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">format.object_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.size</a></span><span class="op">(</span><span class="st">"demo/context_index_100.cif"</span><span class="op">)</span>, <span class="st">"auto"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "1.3 Mb"</span></span></code></pre></div>
<p>Let us rebuild the context index sampling one context every 50.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># building a mutation engine by using the "demo" set-up configuration</span></span>
<span><span class="co"># and setting context_sampling to 50</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span>, context_sampling <span class="op">=</span> <span class="fl">50</span><span class="op">)</span></span>
<span><span class="co">#&gt; Building context index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█████████████████-----------------------] 40% [00m:01s] Processing chr. 22 [█████████████████████████████████-------] 81% [00m:02s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:02s] Context index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving context index [████████████████████████████████████████] 100% [00m:00s] Context index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading RS index [███████████████████---------------------] 47% [00m:01s] Loading RS index [████████████████████████████████████----] 89% [00m:02s] Loading RS index [████████████████████████████████████████] 100% [00m:02s] RS index loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span></code></pre></div>
<p>Above call uses the already downloaded data in the directory
<code>"demo"</code> to build the context and produces the file
<code>"demo/context_index_50.cif"</code>.</p>
<pre><code><span><span class="co"># get the size of the context index when `context_sampling` is 50</span></span>
<span><span class="fu">utils</span><span class="fu">:::</span><span class="fu"><a href="https://rdrr.io/r/utils/object.size.html" class="external-link">format.object_size</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.info.html" class="external-link">file.size</a></span><span class="op">(</span><span class="st">"demo/context_index_50.cif"</span><span class="op">)</span>, <span class="st">"auto"</span><span class="op">)</span></span></code></pre>
</div>
<div class="section level4">
<h4 id="customizing-mutationengine-advanced-topic">Customizing <code>MutationEngine</code> (Advanced Topic)<a class="anchor" aria-label="anchor" href="#customizing-mutationengine-advanced-topic"></a>
</h4>
<p>The function <code><a href="../reference/build_mutation_engine.html">build_mutation_engine()</a></code> can be called by
specifing the name of the set-up directory, the path or URL of the
reference sequence, the SBS file, the driver mutation file, the
passenger CNAs file, and the germline data directory.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">reference_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://ftp.ensembl.org/pub/grch37/release-111/"</span>,</span>
<span>                        <span class="st">"fasta/homo_sapiens/dna/Homo_sapiens.GRCh37."</span>,</span>
<span>                        <span class="st">"dna.chromosome.22.fa.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sbs_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://cancer.sanger.ac.uk/signatures/documents/2123/"</span>,</span>
<span>                  <span class="st">"COSMIC_v3.4_SBS_GRCh37.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">indel_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://cancer.sanger.ac.uk/signatures/documents/"</span>,</span>
<span>                    <span class="st">"2121/COSMIC_v3.4_ID_GRCh37.txt"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">drivers_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/"</span>,</span>
<span>                      <span class="st">"caravagnalab/rRACES/main/inst/extdata/"</span>,</span>
<span>                      <span class="st">"driver_mutations_hg19.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">passenger_cnas_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/"</span>,</span>
<span>                             <span class="st">"caravagnalab/rRACES/main/inst/extdata/"</span>,</span>
<span>                             <span class="st">"passenger_CNAs_hg19.csv"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">germline_url</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"https://www.dropbox.com/scl/fi/g9oloxkip18tr1r"</span>,</span>
<span>                       <span class="st">"m6wjve/germline_data_demo.tar.gz?rlkey=15jshul"</span>,</span>
<span>                       <span class="st">"d3bqgyfcs7fa0bzqeo&amp;dl=1"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># build a mutation engine and place all the files in the directory "Test"</span></span>
<span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>directory <span class="op">=</span> <span class="st">"Test"</span>,</span>
<span>                                  reference_src <span class="op">=</span> <span class="va">reference_url</span>,</span>
<span>                                  SBS_signatures_src <span class="op">=</span> <span class="va">sbs_url</span>,</span>
<span>                                  indel_signatures_src <span class="op">=</span> <span class="va">indel_url</span>,</span>
<span>                                  drivers_src <span class="op">=</span> <span class="va">drivers_url</span>,</span>
<span>                                  passenger_CNAs_src <span class="op">=</span> <span class="va">passenger_cnas_url</span>,</span>
<span>                                  germline_src <span class="op">=</span> <span class="va">germline_url</span><span class="op">)</span></span>
<span><span class="co">#&gt; Downloading reference genome...</span></span>
<span><span class="co">#&gt; Reference genome downloaded</span></span>
<span><span class="co">#&gt; Decompressing reference file...done</span></span>
<span><span class="co">#&gt; Downloading SBS file...</span></span>
<span><span class="co">#&gt; SBS file downloaded</span></span>
<span><span class="co">#&gt; Downloading indel file...</span></span>
<span><span class="co">#&gt; indel file downloaded</span></span>
<span><span class="co">#&gt; Downloading driver mutation file...</span></span>
<span><span class="co">#&gt; Driver mutation file downloaded</span></span>
<span><span class="co">#&gt; Downloading passenger CNAs file...</span></span>
<span><span class="co">#&gt; Passenger CNAs file downloaded</span></span>
<span><span class="co">#&gt; Downloading germline mutations...</span></span>
<span><span class="co">#&gt; Germline mutations downloaded</span></span>
<span><span class="co">#&gt; Building context index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█████████████████-----------------------] 40% [00m:01s] Processing chr. 22 [█████████████████████████████████-------] 81% [00m:02s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:02s] Context index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving context index [████████████████████████████████████████] 100% [00m:00s] Context index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt; Building repeated sequence index...</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:00s] Processing chr. 22 [█---------------------------------------] 0% [00m:03s] Processing chr. 22 [█---------------------------------------] 0% [00m:04s] Processing chr. 22 [█---------------------------------------] 0% [00m:05s] Processing chr. 22 [█---------------------------------------] 0% [00m:07s] Processing chr. 22 [█---------------------------------------] 0% [00m:08s] Processing chr. 22 [█---------------------------------------] 0% [00m:11s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:12s] RS index built</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving RS index [█████████-------------------------------] 20% [00m:01s] Saving RS index [███████████████████████████████████-----] 85% [00m:02s] Saving RS index [████████████████████████████████████████] 100% [00m:02s] RS index saved</span></span>
<span><span class="co">#&gt; done</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving germline [████████████████████████████████████████] 100% [00m:00s] Germline saved</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.exists</a></span><span class="op">(</span><span class="st">"Test"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="mutant-genetic-characterization">Mutant Genetic Characterization<a class="anchor" aria-label="anchor" href="#mutant-genetic-characterization"></a>
</h4>
<p>Once the mutation engine has been built, we can define mutant
genotype and declare species mutation rates.</p>
<p>Let us consider the <em>“Border Growth”</em> simulation peformed in
<code><a href="../articles/growth_models.html">vignette("growth_models")</a></code>. It involves the two mutants
<code>A</code> and “B”. Both of them has two possible epigenetic states,
<code>+</code> and <code>-</code>, leading to the four the species
<code>A+</code>, <code>A-</code>, <code>B+</code>, and <code>B-</code>,
respectively. Each of these species has a passenger mutation rate that
must be declated to the mutation engine before labeling the samples
forest. Let 1e-9, 3e-8, 8e-7, and 5e-8 be the passenger SNV rates of the
species <code>A+</code>, <code>A-</code>, <code>B+</code>, and
<code>B-</code>, respectively. Furthermore, let 0, 1e-11, 0, and 0 be
the passenger CNA rates of the same species, respectively.</p>
<p>The two mutants may also be genetically characterized by some driver
mutations. The driver mutations associated with each of the mutants must
occur in any cell belonging to the mutant itself. Hence, they must be
declared to the mutant engine before the labeling.</p>
<p>The method <code>MutationEngine$add_mutant()</code> takes care of all
these declarations.</p>
<p>For the sake of example, let us assume that <code>A</code> is
characterized by one driver mutation on chromosome 22, while
<code>B</code> has three driver mutations on the same chromosome.</p>
<p>The genetic specification of the mutant <code>A</code> can be
declared as it follows.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add the mutant "A" characterized by one driver SNV on the allele 1 of</span></span>
<span><span class="co"># chromosome 22, one indel deletion on the same chromosome, and two CNAs:</span></span>
<span><span class="co"># a deletion on the allele 1 of chromosome 22 and an amplification on a</span></span>
<span><span class="co"># random allele of the same chromosome. The mutant has two epigenetic</span></span>
<span><span class="co"># states and its species "A+" and "A-" have passenger SNV rates 1e-9 and</span></span>
<span><span class="co"># 3e-8, respectively, passenger indel rates 1e-10 and 0, respectively,</span></span>
<span><span class="co"># and passenger CNA rates 0 and 1e-11, respectively.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>mutant_name <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>                    passenger_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">1e-9</span>, indel <span class="op">=</span> <span class="fl">1e-10</span><span class="op">)</span>,</span>
<span>                                           <span class="st">"-"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">3e-8</span>, CNA <span class="op">=</span> <span class="fl">1e-11</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    drivers <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="../reference/SNV.html">SNV</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">10510210</span>, <span class="st">"C"</span>, allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="../reference/Mutation.html">Mutation</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">15220157</span>, <span class="st">"GTTTTTTTT"</span>, <span class="st">"G"</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="../reference/CNA.html">CNA</a></span><span class="op">(</span>type <span class="op">=</span> <span class="st">"A"</span>, chr <span class="op">=</span> <span class="st">"22"</span>,</span>
<span>                                       chr_pos <span class="op">=</span> <span class="fl">10303470</span>, len <span class="op">=</span> <span class="fl">200000</span><span class="op">)</span>,</span>
<span>                                   <span class="fu"><a href="../reference/CNA.html">CNA</a></span><span class="op">(</span><span class="st">"D"</span>, <span class="st">"22"</span>, <span class="fl">5010000</span>, <span class="fl">200000</span>,</span>
<span>                                       allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Retrieving "A" SNVs [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [████████████████████████████████████████] 100% [00m:01s] "A" SNVs retrieved</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A" SNVs&amp;indels: </span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;    "A" CNAs: </span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span></code></pre></div>
<p>In the above code, the driver CNAs, SNVs, and indels are declared by
calling the functions <code><a href="../reference/CNA.html">CNA()</a></code>, <code><a href="../reference/SNV.html">SNV()</a></code>, and
<code><a href="../reference/Mutation.html">Mutation()</a></code>, respectively. These function allow to specify
the allele in which the mutation must lay.</p>
<p>As far as the SNVs and indels concern, rRACES provides the users with
a more compact and, sometimes, convenient approach. The mutation engine
stores a list of known driver mutations and labels each of them by a
code which can be used during the mutant declaration. The corresponding
data frame can be obtained by using the method
<code>MutationEngine$get_known_drivers()</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_known_drivers</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="st">"22"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr     from       to ref alt mutation_type tumor_type driver_gene</span></span>
<span><span class="co">#&gt; 1  22 30057302 30057302   C   T           SNV         OV         NF2</span></span>
<span><span class="co">#&gt; 2  22 30057302 30057302   C   T           SNV         OV         NF2</span></span>
<span><span class="co">#&gt;   driver_code</span></span>
<span><span class="co">#&gt; 1   NF2 R221*</span></span>
<span><span class="co">#&gt; 2   NF2 R262*</span></span></code></pre></div>
<p>The code of the known driver mutations can be used in place of the
full specification as follows.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add the mutant "B" characterized by three driver SNVs on chromosome 22 (no</span></span>
<span><span class="co"># CNA) and two epigenetic states. The first two SNVs are "NF2 R262", which</span></span>
<span><span class="co"># can lay in any available allele, and "NF2 R221*", which instead must lay</span></span>
<span><span class="co"># in the allele 1. The remaining SNV is specified by using the SNV function</span></span>
<span><span class="co"># as done above. The species "B+" and "B-" have passenger SNV rates 8e-7</span></span>
<span><span class="co"># and 5e-8, respectively, and CNA rates 0 for both species.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span><span class="st">"B"</span>, <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">8e-7</span><span class="op">)</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SNV <span class="op">=</span> <span class="fl">5e-8</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="../reference/SNV.html">SNV</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="fl">12028576</span>, <span class="st">"G"</span><span class="op">)</span>,         <span class="co"># the first SNV</span></span>
<span>                         <span class="st">"NF2 R262*"</span>,                      <span class="co"># the second SNV</span></span>
<span>                         <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">"NF2 R221*"</span>, allele <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>   <span class="co"># the third SNV</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Retrieving "B" SNVs [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [████████████████████████████████████████] 100% [00m:01s] "B" SNVs retrieved</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt;    "B+": {SNV: 8e-07}</span></span>
<span><span class="co">#&gt;    "B-": {SNV: 5e-08}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A" SNVs&amp;indels: </span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;    "A" CNAs: </span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt;    "B" SNVs&amp;indels: </span></span>
<span><span class="co">#&gt;        chr22(12028576)[N&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: random)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="mutational-exposures">Mutational Exposures<a class="anchor" aria-label="anchor" href="#mutational-exposures"></a>
</h4>
<p>The probability for a mutation to occur depends on both its genomic
and environmental context.</p>
<p>A <em>signature</em> is a mutation probability distribution over
mutation contexts or mutation structure. <em>SBS (single base
substitution) signatures</em> provide for any genomic context (i.e., a
triplet of bases) the probability for a SNV to occur on that context. On
the contrary, <em>ID (indel) signatures</em> associates the probability
of an indel to its length and structure (see <span class="citation">(Alexandrov et al. 2020)</span>).</p>
<p>The signature depends on the environmental context and, because of
that, more than one signature may be active at the same time with
different probabilities. An <em>mutational exposure</em> (or
<em>exposure</em>) is a discrete probability distribution among
signatures.</p>
<p>In order to simulate passeger mutations of a given type, we need to
specify a default exposure for that type. This can be achieved as it
follows.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add SNV and indel default exposures. This will be used from simulated time 0</span></span>
<span><span class="co"># up to the successive exposure change.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span>coefficients <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SBS13 <span class="op">=</span> <span class="fl">0.2</span>, SBS1 <span class="op">=</span> <span class="fl">0.8</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ID2 <span class="op">=</span> <span class="fl">0.6</span>, ID13 <span class="op">=</span> <span class="fl">0.2</span>, ID21 <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Further exposures can also be defined by specifying an activation
time for each of them, i.e., the new exposures will be used from the
specified time up to the next exposure change.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># add a new SNV exposure that will be used from simulated</span></span>
<span><span class="co"># time 100 up to the successive exposure change.</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span>time <span class="op">=</span> <span class="fl">100</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SBS5 <span class="op">=</span> <span class="fl">0.3</span>, SBS2 <span class="op">=</span> <span class="fl">0.2</span>, SBS3 <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt;    "B+": {SNV: 8e-07}</span></span>
<span><span class="co">#&gt;    "B-": {SNV: 5e-08}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A" SNVs&amp;indels: </span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;    "A" CNAs: </span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt;    "B" SNVs&amp;indels: </span></span>
<span><span class="co">#&gt;        chr22(12028576)[N&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: random)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, 100[: {"SBS1": 0.8, "SBS13": 0.2}</span></span>
<span><span class="co">#&gt;      [100, ∞[: {"SBS2": 0.2, "SBS3": 0.5, "SBS5": 0.3}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, ∞[: {"ID13": 0.2, "ID2": 0.6, "ID21": 0.2}</span></span></code></pre></div>
<p>The default exposures and the simultaneous changes in SNV and ID
exposures can be specified by a sigle call to the function
<code>MutationEngine$add_exposure()</code> as it follows.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">add_exposure</span><span class="op">(</span>time <span class="op">=</span> <span class="fl">120</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>SBS5 <span class="op">=</span> <span class="fl">0.3</span>, SBS2 <span class="op">=</span> <span class="fl">0.2</span>, SBS3 <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                    ID1 <span class="op">=</span> <span class="fl">0.8</span>, ID9 <span class="op">=</span> <span class="fl">0.2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">m_engine</span></span>
<span><span class="co">#&gt; MutationEngine</span></span>
<span><span class="co">#&gt;  Passenger rates</span></span>
<span><span class="co">#&gt;    "A+": {SNV: 1e-09}</span></span>
<span><span class="co">#&gt;    "A-": {SNV: 3e-08, CNA: 1e-11}</span></span>
<span><span class="co">#&gt;    "B+": {SNV: 8e-07}</span></span>
<span><span class="co">#&gt;    "B-": {SNV: 5e-08}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Driver mutations</span></span>
<span><span class="co">#&gt;    "A" SNVs&amp;indels: </span></span>
<span><span class="co">#&gt;        chr22(10510210)[N&gt;C](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(15220157)[GTTTTTTTT&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;    "A" CNAs: </span></span>
<span><span class="co">#&gt;        CNA("D",chr22(5010000), len: 200000, allele: 1)</span></span>
<span><span class="co">#&gt;        CNA("A",chr22(10303470), len: 200000)</span></span>
<span><span class="co">#&gt;    "B" SNVs&amp;indels: </span></span>
<span><span class="co">#&gt;        chr22(12028576)[N&gt;G](allele: random)</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: 1)</span></span>
<span><span class="co">#&gt;        chr22(30057302)[C&gt;T](allele: random)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Timed Exposure</span></span>
<span><span class="co">#&gt;    SBS Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, 100[: {"SBS1": 0.8, "SBS13": 0.2}</span></span>
<span><span class="co">#&gt;      [100, 120[: {"SBS2": 0.2, "SBS3": 0.5, "SBS5": 0.3}</span></span>
<span><span class="co">#&gt;      [120, ∞[: {"SBS2": 0.2, "SBS3": 0.5, "SBS5": 0.3}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;    indel Timed Exposures</span></span>
<span><span class="co">#&gt;      [0, 120[: {"ID13": 0.2, "ID2": 0.6, "ID21": 0.2}</span></span>
<span><span class="co">#&gt;      [120, ∞[: {"ID1": 0.8, "ID9": 0.2}</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="passenger-cnas-and-tumor-type">Passenger CNAs and Tumor Type<a class="anchor" aria-label="anchor" href="#passenger-cnas-and-tumor-type"></a>
</h4>
<p>The passenger CNAs applied during the simulation depend on tumor
type. The type of tumor can be specified when building the mutation
engine by using the <code>tumor_type</code> parameter. For instance, if
the passenger CNAs file used to build the mutation engine contains some
of the data identified in breast carcinoma in the UK, the specific CNA
can be applied during the simulation as it follows.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/build_mutation_engine.html">build_mutation_engine</a></span><span class="op">(</span>setup_code <span class="op">=</span> <span class="st">"demo"</span>,</span>
<span>                                  tumor_type <span class="op">=</span> <span class="st">"BRCA-UK"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="germline-mutations">Germline Mutations<a class="anchor" aria-label="anchor" href="#germline-mutations"></a>
</h4>
<p>rRACES allows users to apply the germline mutations of one of the
subjects available in the germline data provided in building the
mutation engine. This feature will enable users to simulate a specific
cancer type’s evolution on an individual with the desired gender and
ethnicity.</p>
<p>The available subjects, together with their sex and ethnicity, can be
obtained by using the method
<code>[MutationEngine$get_germline_subjects()]</code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">subjects</span> <span class="op">&lt;-</span> <span class="va">m_engine</span><span class="op">$</span><span class="fu">get_germline_subjects</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">subjects</span></span>
<span><span class="co">#&gt;    sample pop super_pop gender</span></span>
<span><span class="co">#&gt; 1 NA18941 JPT       EAS female</span></span>
<span><span class="co">#&gt; 2 NA20513 TSI       EUR   male</span></span></code></pre></div>
<p>The column <code>sample</code> contains the names of the available
subjects. The columns <code>pop</code> and <code>super_pop</code> report
the subjects’ population and super-population codes. The last column,
<code>gender</code>, includes the subject gender.</p>
<p>The method <code>MutationEngine$get_population_descritions()</code>
clarifies the meaning of the codes reported in the pop columns.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_population_descritions</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    code          description</span></span>
<span><span class="co">#&gt; 1   CHB          Han Chinese</span></span>
<span><span class="co">#&gt; 2   JPT             Japanese</span></span>
<span><span class="co">#&gt; 3   CHS Southern Han Chinese</span></span>
<span><span class="co">#&gt; 4   CDX          Dai Chinese</span></span>
<span><span class="co">#&gt; 5   KHV      Kinh Vietnamese</span></span>
<span><span class="co">#&gt; 6   CHD       Denver Chinese</span></span>
<span><span class="co">#&gt; 7   CEU                 CEPH</span></span>
<span><span class="co">#&gt; 8   TSI               Tuscan</span></span>
<span><span class="co">#&gt; 9   GBR              British</span></span>
<span><span class="co">#&gt; 10  FIN              Finnish</span></span>
<span><span class="co">#&gt; 11  IBS              Spanish</span></span>
<span><span class="co">#&gt; 12  YRI               Yoruba</span></span>
<span><span class="co">#&gt; 13  LWK                Luhya</span></span>
<span><span class="co">#&gt; 14  GWD              Gambian</span></span>
<span><span class="co">#&gt; 15  MSL                Mende</span></span>
<span><span class="co">#&gt; 16  ESN                 Esan</span></span>
<span><span class="co">#&gt; 17  ASW  African-American SW</span></span>
<span><span class="co">#&gt; 18  ACB    African-Caribbean</span></span>
<span><span class="co">#&gt; 19  MXL     Mexican-American</span></span>
<span><span class="co">#&gt; 20  PUR         Puerto Rican</span></span>
<span><span class="co">#&gt; 21  CLM            Colombian</span></span>
<span><span class="co">#&gt; 22  PEL             Peruvian</span></span>
<span><span class="co">#&gt; 23  GIH             Gujarati</span></span>
<span><span class="co">#&gt; 24  PJL              Punjabi</span></span>
<span><span class="co">#&gt; 25  BEB              Bengali</span></span>
<span><span class="co">#&gt; 26  STU           Sri Lankan</span></span>
<span><span class="co">#&gt; 27  ITU               Indian</span></span>
<span><span class="co">#&gt;                                                      long.description</span></span>
<span><span class="co">#&gt; 1                                       Han Chinese in Beijing, China</span></span>
<span><span class="co">#&gt; 2                                            Japanese in Tokyo, Japan</span></span>
<span><span class="co">#&gt; 3                                                   Han Chinese South</span></span>
<span><span class="co">#&gt; 4                                 Chinese Dai in Xishuangbanna, China</span></span>
<span><span class="co">#&gt; 5                                   Kinh in Ho Chi Minh City, Vietnam</span></span>
<span><span class="co">#&gt; 6                          Chinese in Denver, Colorado (pilot 3 only)</span></span>
<span><span class="co">#&gt; 7  Utah residents (CEPH) with Northern and Western European ancestry </span></span>
<span><span class="co">#&gt; 8                                                  Toscani in Italia </span></span>
<span><span class="co">#&gt; 9                                    British in England and Scotland </span></span>
<span><span class="co">#&gt; 10                                                Finnish in Finland </span></span>
<span><span class="co">#&gt; 11                                      Iberian populations in Spain </span></span>
<span><span class="co">#&gt; 12                                          Yoruba in Ibadan, Nigeria</span></span>
<span><span class="co">#&gt; 13                                             Luhya in Webuye, Kenya</span></span>
<span><span class="co">#&gt; 14                           Gambian in Western Division, The Gambia </span></span>
<span><span class="co">#&gt; 15                                              Mende in Sierra Leone</span></span>
<span><span class="co">#&gt; 16                                                    Esan in Nigeria</span></span>
<span><span class="co">#&gt; 17                                  African Ancestry in Southwest US </span></span>
<span><span class="co">#&gt; 18                                      African Caribbean in Barbados</span></span>
<span><span class="co">#&gt; 19                        Mexican Ancestry in Los Angeles, California</span></span>
<span><span class="co">#&gt; 20                                        Puerto Rican in Puerto Rico</span></span>
<span><span class="co">#&gt; 21                                    Colombian in Medellin, Colombia</span></span>
<span><span class="co">#&gt; 22                                             Peruvian in Lima, Peru</span></span>
<span><span class="co">#&gt; 23                                     Gujarati Indian in Houston, TX</span></span>
<span><span class="co">#&gt; 24                                        Punjabi in Lahore, Pakistan</span></span>
<span><span class="co">#&gt; 25                                              Bengali in Bangladesh</span></span>
<span><span class="co">#&gt; 26                                         Sri Lankan Tamil in the UK</span></span>
<span><span class="co">#&gt; 27                                            Indian Telugu in the UK</span></span></code></pre></div>
<p>The method <code>MutationEngine$get_active_germline()</code> returns
the the active germline subject.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_active_germline</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    sample pop super_pop gender</span></span>
<span><span class="co">#&gt; 1 NA18941 JPT       EAS female</span></span></code></pre></div>
<p>Users can change the germline subject using the method
<code>MutationEngine$set_germline_subject()</code>.</p>
<p>When a subject is selected for the first time, rRACES builds a binary
representation of the subject genome, saving it for future use. This
step may take a few minutes. However, all the successive selections of
the same subject directly load the binary file.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">m_engine</span><span class="op">$</span><span class="fu">set_germline_subject</span><span class="op">(</span><span class="va">subjects</span><span class="op">[</span><span class="fl">2</span>, <span class="st">"sample"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading germline [████████████████████████████████████████] 100% [00m:00s] Germline loaded</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving germline [████████████████████████████████████████] 100% [00m:00s] Germline saved</span></span>
<span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_active_germline</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    sample pop super_pop gender</span></span>
<span><span class="co">#&gt; 1 NA20513 TSI       EUR   male</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="building-phylogenetic-forests">Building Phylogenetic Forests<a class="anchor" aria-label="anchor" href="#building-phylogenetic-forests"></a>
</h3>
<p>The configurated mutation engine can be used to label each node in a
samples forest by mutations.</p>
<p>Since the mutation engine has be configurated to deal with the
simulation peformed in <code><a href="../articles/growth_models.html">vignette("growth_models")</a></code>, we can
use the samples forest saved in the file
<code>"samples_forest.sff"</code> in that article.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samples_forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_samples_forest.html">load_samples_forest</a></span><span class="op">(</span><span class="st">"samples_forest.sff"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># place mutations on the sample forest assuming 1000 pre-neoplastic SNVs and</span></span>
<span><span class="co"># 500 indels</span></span>
<span><span class="va">phylo_forest</span> <span class="op">&lt;-</span> <span class="va">m_engine</span><span class="op">$</span><span class="fu">place_mutations</span><span class="op">(</span><span class="va">samples_forest</span>, <span class="fl">1000</span>, <span class="fl">500</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Placing mutations [█████████████████-----------------------] 41% [00m:01s] Placing mutations [███████████████████████-----------------] 55% [00m:02s] Placing mutations [███████████████████████████████---------] 77% [00m:03s] Placing mutations [████████████████████████████████████████] 100% [00m:03s] Mutations placed</span></span>
<span></span>
<span><span class="va">phylo_forest</span></span>
<span><span class="co">#&gt; PhylogeneticForest</span></span>
<span><span class="co">#&gt;   # of trees: 1</span></span>
<span><span class="co">#&gt;   # of nodes: 4759</span></span>
<span><span class="co">#&gt;   # of leaves: 949</span></span>
<span><span class="co">#&gt;   samples: {"S_1_1", "S_1_2", "S_2_1", "S_2_2"}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # of emerged SNVs and indels: 69115</span></span>
<span><span class="co">#&gt;   # of emerged CNAs: 2</span></span></code></pre></div>
<p>The phylogenetic forest stores all the mutations labeling the sampled
cells which are represented by the forest leaves. Users can retrieve
such data by using the methods
<code>PhylogeneticForest$get_sampled_cell_mutations()</code> and
<code>PhylogeneticForest$get_sampled_cell_CNAs()</code>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># select the first mutations among all the mutations occuring in</span></span>
<span><span class="co"># the genomes of the sampled cells</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id chr  chr_pos allele       ref    alt  type cause          class</span></span>
<span><span class="co">#&gt; 1   11546  22 15220157      0 GTTTTTTTT      G indel     A         driver</span></span>
<span><span class="co">#&gt; 2   11546  22 16085881      0         G      A   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 3   11546  22 16212399      0         C CAAAAA indel   ID1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 4   11546  22 16291927      0         C      T   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 5   11546  22 16370220      0         C      T   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 6   11546  22 16413396      0         T TAAAAA indel   ID1 pre-neoplastic</span></span>
<span></span>
<span><span class="co"># select the first CNAs among all the mutations occuring in</span></span>
<span><span class="co"># the genomes of the sampled cells</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_CNAs</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id type chr    begin      end allele src.allele  class</span></span>
<span><span class="co">#&gt; 1   11546    D  22  5010000  5209999      1         NA driver</span></span>
<span><span class="co">#&gt; 2   11546    A  22 10303470 10503469      2          0 driver</span></span>
<span><span class="co">#&gt; 3   18706    D  22  5010000  5209999      1         NA driver</span></span>
<span><span class="co">#&gt; 4   18706    A  22 10303470 10503469      2          0 driver</span></span>
<span><span class="co">#&gt; 5   21863    D  22  5010000  5209999      1         NA driver</span></span>
<span><span class="co">#&gt; 6   21863    A  22 10303470 10503469      2          0 driver</span></span>
<span></span>
<span><span class="co"># get the sampled cells</span></span>
<span><span class="va">sampled_cells</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_nodes</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># show the first of them</span></span>
<span><span class="va">sampled_cells</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id ancestor mutant epistate sample birth_time</span></span>
<span><span class="co">#&gt; 1   11546    10564      A        -  S_1_2   183.2361</span></span>
<span><span class="co">#&gt; 2   18706     2158      A        -  S_1_1   212.3915</span></span>
<span><span class="co">#&gt; 3   21863    14238      A        -  S_1_2   222.5558</span></span>
<span><span class="co">#&gt; 4   23259    16573      A        -  S_1_1   226.7531</span></span>
<span><span class="co">#&gt; 5   34036    25714      A        -  S_1_1   253.3768</span></span>
<span><span class="co">#&gt; 6   34216    17258      A        -  S_1_1   253.7922</span></span>
<span></span>
<span><span class="co"># get the identifier of the 3rd cell in `sampled_cells`</span></span>
<span><span class="va">cell_id</span> <span class="op">&lt;-</span> <span class="va">sampled_cells</span><span class="op">[</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># get the SNVs and the indels of the 3rd cell in `sampled_cells`</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id chr  chr_pos allele       ref    alt  type cause          class</span></span>
<span><span class="co">#&gt; 1   21863  22 15220157      0 GTTTTTTTT      G indel     A         driver</span></span>
<span><span class="co">#&gt; 2   21863  22 16085881      0         G      A   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 3   21863  22 16212399      0         C CAAAAA indel   ID1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 4   21863  22 16291927      0         C      T   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 5   21863  22 16370220      0         C      T   SNV  SBS1 pre-neoplastic</span></span>
<span><span class="co">#&gt; 6   21863  22 16413396      0         T TAAAAA indel   ID1 pre-neoplastic</span></span>
<span></span>
<span><span class="co"># get the CNAs of the 3rd cell in `sampled_cells`</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_CNAs</span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id type chr    begin      end allele src.allele  class</span></span>
<span><span class="co">#&gt; 1   21863    D  22  5010000  5209999      1         NA driver</span></span>
<span><span class="co">#&gt; 2   21863    A  22 10303470 10503469      2          0 driver</span></span></code></pre></div>
<p>The method <code>PhylogeneticForest$get_germline_mutations()</code>
returns the SNVs and the indels in the germline.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># extract the germline mutation</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_germline_mutations</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr  chr_pos allele ref   alt  type cause    class</span></span>
<span><span class="co">#&gt; 1  22 16051493      0   G     A   SNV       germinal</span></span>
<span><span class="co">#&gt; 2  22 16052167      0   A AAAAC indel       germinal</span></span>
<span><span class="co">#&gt; 3  22 16053659      0   A     C   SNV       germinal</span></span>
<span><span class="co">#&gt; 4  22 16054740      0   A     G   SNV       germinal</span></span>
<span><span class="co">#&gt; 5  22 16055942      0   C     T   SNV       germinal</span></span>
<span><span class="co">#&gt; 6  22 16058070      0   A     G   SNV       germinal</span></span></code></pre></div>
<p>Users can also identify the cell in which a mutation emerged even
when the cell was not sampled.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># select one of the mutations</span></span>
<span><span class="va">mutation_row</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="va">cell_id</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co"># rebuild the corresponding mutation</span></span>
<span><span class="va">mutation</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/Mutation.html">Mutation</a></span><span class="op">(</span><span class="va">mutation_row</span><span class="op">[</span><span class="st">"chr"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>,</span>
<span>                     <span class="va">mutation_row</span><span class="op">[</span><span class="st">"chr_pos"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>,</span>
<span>                     <span class="va">mutation_row</span><span class="op">[</span><span class="st">"ref"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>,</span>
<span>                     <span class="va">mutation_row</span><span class="op">[</span><span class="st">"alt"</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the identifier of the oldest cells in which the mutation occurs</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_first_occurrences</span><span class="op">(</span><span class="va">mutation</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>The exposures used in placing the mutations on the cells in the
phylogenetic forest can be obtained by using the method
<code>PhylogeneticForest$get_exposures()</code>.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the exposures used in placing the mutations</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_exposures</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    time signature exposure  type</span></span>
<span><span class="co">#&gt; 1     0      SBS1      0.8   SNV</span></span>
<span><span class="co">#&gt; 2     0     SBS13      0.2   SNV</span></span>
<span><span class="co">#&gt; 3   100      SBS2      0.2   SNV</span></span>
<span><span class="co">#&gt; 4   100      SBS3      0.5   SNV</span></span>
<span><span class="co">#&gt; 5   100      SBS5      0.3   SNV</span></span>
<span><span class="co">#&gt; 6   120      SBS2      0.2   SNV</span></span>
<span><span class="co">#&gt; 7   120      SBS3      0.5   SNV</span></span>
<span><span class="co">#&gt; 8   120      SBS5      0.3   SNV</span></span>
<span><span class="co">#&gt; 9     0      ID13      0.2 indel</span></span>
<span><span class="co">#&gt; 10    0       ID2      0.6 indel</span></span>
<span><span class="co">#&gt; 11    0      ID21      0.2 indel</span></span>
<span><span class="co">#&gt; 12  120       ID1      0.8 indel</span></span>
<span><span class="co">#&gt; 13  120       ID9      0.2 indel</span></span></code></pre></div>
<p>The details about the SNV and indel signatures adopted during the
evolution are available in the mutation engine and they can be retrieved
by using the methods <code>MutationEngine$get_SNV_signatures()</code>
and <code>MutationEngine$get_indel_signatures()</code>.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the SNV signatures used in placing the mutations</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_SNV_signatures</span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;      Type        SBS1        SBS2       SBS3        SBS4</span></span>
<span><span class="co">#&gt; 1 A[C&gt;A]A 0.000886157 5.80000e-07 0.02080832 0.042196498</span></span>
<span><span class="co">#&gt; 2 A[C&gt;A]C 0.002280405 1.48004e-04 0.01650660 0.033297236</span></span>
<span><span class="co">#&gt; 3 A[C&gt;A]G 0.000177031 5.23000e-05 0.00175070 0.015598705</span></span>
<span><span class="co">#&gt; 4 A[C&gt;A]T 0.001280227 9.78000e-05 0.01220488 0.029497552</span></span>
<span><span class="co">#&gt; 5 A[C&gt;G]A 0.001860330 2.23000e-16 0.01970788 0.006889428</span></span>
<span><span class="co">#&gt; 6 A[C&gt;G]C 0.001220217 1.33004e-04 0.01170468 0.002839764</span></span>
<span></span>
<span><span class="co"># get the indel signatures used in placing the mutations</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="fu">get_indel_signatures</span><span class="op">(</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">6</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;        Type          ID1         ID2        ID3         ID4</span></span>
<span><span class="co">#&gt; 1 1:Del:C:0 1.598890e-04 0.004824116 0.12472711 0.007249717</span></span>
<span><span class="co">#&gt; 2 1:Del:C:1 7.735230e-04 0.000022100 0.20887617 0.002734869</span></span>
<span><span class="co">#&gt; 3 1:Del:C:2 3.310000e-18 0.000003110 0.17632422 0.002041063</span></span>
<span><span class="co">#&gt; 4 1:Del:C:3 1.907613e-03 0.002472076 0.06404276 0.001112283</span></span>
<span><span class="co">#&gt; 5 1:Del:C:4 7.059900e-04 0.003856976 0.04398981 0.001075684</span></span>
<span><span class="co">#&gt; 6 1:Del:C:5 3.370115e-03 0.014615424 0.02241749 0.002246178</span></span></code></pre></div>
<p>The name of the subject whose germline corresponds to wild-type
genome in the phylogenetic forest can be obtained by the method
<code>Phylogenetic$get_germline_subject()</code>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_germline_subject</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "NA20513"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="infinite-sites-model">Infinite Sites Model<a class="anchor" aria-label="anchor" href="#infinite-sites-model"></a>
</h3>
<p>The mutation engine places mutation on the sampled cell genome
according to the <a href="https://en.wikipedia.org/wiki/Infinite_sites_model" class="external-link">infinite sites
model</a> by default. In particular, any new mutations is placed on a
locus whose context is mutation-free.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># this function verifies whether any mutation arised in</span></span>
<span><span class="co"># two unrelated cells represented in a phylogenetic forest</span></span>
<span><span class="va">test_infinite_sites_model</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">phylo_forest</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="co"># extract non-germinal SNVs that appear multiple times either</span></span>
<span>  <span class="co"># in the same cell or different cells</span></span>
<span>  <span class="va">snvs</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">class</span> <span class="op">!=</span> <span class="st">"germinal"</span>, <span class="va">.data</span><span class="op">$</span><span class="va">type</span> <span class="op">==</span> <span class="st">"SNV"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">.data</span><span class="op">$</span><span class="va">chr</span>, <span class="va">.data</span><span class="op">$</span><span class="va">chr_pos</span>, <span class="va">.data</span><span class="op">$</span><span class="va">ref</span>, <span class="va">.data</span><span class="op">$</span><span class="va">alt</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">n</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># search for an SNV that independently occurred in two unrelated cells</span></span>
<span>  <span class="va">first_occurrences</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="va">row</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>  <span class="kw">while</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">first_occurrences</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">2</span> <span class="op">&amp;&amp;</span> <span class="va">row</span> <span class="op">&lt;=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">snvs</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">snv</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SNV.html">SNV</a></span><span class="op">(</span><span class="va">snvs</span><span class="op">[</span><span class="va">row</span>, <span class="st">"chr"</span><span class="op">]</span>, <span class="va">snvs</span><span class="op">[</span><span class="va">row</span>, <span class="st">"chr_pos"</span><span class="op">]</span>,</span>
<span>               ref <span class="op">=</span> <span class="va">snvs</span><span class="op">[</span><span class="va">row</span>, <span class="st">"ref"</span><span class="op">]</span>, alt <span class="op">=</span> <span class="va">snvs</span><span class="op">[</span><span class="va">row</span>, <span class="st">"alt"</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">first_occurrences</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_first_occurrences</span><span class="op">(</span><span class="va">snv</span><span class="op">)</span></span>
<span>    <span class="va">row</span> <span class="op">&lt;-</span> <span class="va">row</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="co"># if the last handled SNV independently occurred in two unrelated</span></span>
<span>  <span class="co"># cells at least</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">first_occurrences</span><span class="op">)</span> <span class="op">&gt;=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>    <span class="co"># print a message containing the two cells</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"SNV('"</span>, <span class="va">snv</span><span class="op">$</span><span class="fu">get_chromosome</span><span class="op">(</span><span class="op">)</span>, <span class="st">"'',"</span>,</span>
<span>           <span class="va">snv</span><span class="op">$</span><span class="fu">get_position_in_chromosome</span><span class="op">(</span><span class="op">)</span>, <span class="st">",'"</span>, <span class="va">snv</span><span class="op">$</span><span class="fu">get_ref</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="st">"','"</span>, <span class="va">snv</span><span class="op">$</span><span class="fu">get_alt</span><span class="op">(</span><span class="op">)</span>,</span>
<span>           <span class="st">"') independently arises in cells "</span>, <span class="va">first_occurrences</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>           <span class="st">" and "</span>, <span class="va">first_occurrences</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="st">"Every mutation arises exclusively in one cell"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co"># test whether the infinite sites conditions hold in the built forest</span></span>
<span><span class="fu">test_infinite_sites_model</span><span class="op">(</span><span class="va">phylo_forest</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Every mutation arises exclusively in one cell"</span></span></code></pre></div>
<p>This behavior can be changed by using the mutation engine property
<code>MutationEngine$infinite_sites_model()</code>. This property is a
Boolean flag that enable/disable the infinite sites model.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># establish whether the infinite sites model is used</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="va">infinite_sites_model</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span></span>
<span><span class="co"># disable it</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="va">infinite_sites_model</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span></code></pre></div>
<p>When the infinite sites model is disabled,
<code>MutationEngine$place_mutations()</code> may place two mutations in
the same locus of different alleles of the same genome or the same
mutation in the same locus of one allele of the genomes of two cells
that are not each other ancestors.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># test whether the infinite sites model is enable</span></span>
<span><span class="va">m_engine</span><span class="op">$</span><span class="va">infinite_sites_model</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span></span>
<span><span class="co"># place the mutations on the same samples forest used above</span></span>
<span><span class="va">phylo_forest2</span> <span class="op">&lt;-</span> <span class="va">m_engine</span><span class="op">$</span><span class="fu">place_mutations</span><span class="op">(</span><span class="va">samples_forest</span>, <span class="fl">1000</span>, <span class="fl">500</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Placing mutations [█████████████████-----------------------] 42% [00m:01s] Placing mutations [███████████████████████-----------------] 56% [00m:02s] Placing mutations [██████████████████████████████████------] 83% [00m:03s] Placing mutations [████████████████████████████████████████] 100% [00m:03s] Mutations placed</span></span>
<span></span>
<span><span class="co"># test whether the infinite sites conditions hold in the new forest</span></span>
<span><span class="fu">test_infinite_sites_model</span><span class="op">(</span><span class="va">phylo_forest2</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "SNV('22'',16057447,'G','A') independently arises in cells 12793 and 242025"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="storing-phylogenetic-forests">Storing Phylogenetic Forests<a class="anchor" aria-label="anchor" href="#storing-phylogenetic-forests"></a>
</h3>
<p>As in the case of the samples forests, the phylogenetic forests can
be saved by using the method <code>PhylogeneticForest$save()</code> and
load by the function <code><a href="../reference/load_phylogenetic_forest.html">load_phylogenetic_forest()</a></code>.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># save the phylogenetic forest in the file "phylo_forest.sff"</span></span>
<span><span class="va">phylo_forest</span><span class="op">$</span><span class="fu">save</span><span class="op">(</span><span class="st">"phylo_forest.sff"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Saving forest [████████████████████████████████████████] 100% [00m:00s] Forest saved</span></span>
<span></span>
<span><span class="co"># loading the saved forest</span></span>
<span><span class="va">loaded_phylo_forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_phylogenetic_forest.html">load_phylogenetic_forest</a></span><span class="op">(</span><span class="st">"phylo_forest.sff"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading forest [██████████████████----------------------] 43% [00m:01s] Loading forest [███████████████████████████████████-----] 85% [00m:02s] Loading forest [████████████████████████████████████████] 100% [00m:02s] Forest loaded</span></span>
<span></span>
<span><span class="va">loaded_phylo_forest</span></span>
<span><span class="co">#&gt; PhylogeneticForest</span></span>
<span><span class="co">#&gt;   # of trees: 1</span></span>
<span><span class="co">#&gt;   # of nodes: 4759</span></span>
<span><span class="co">#&gt;   # of leaves: 949</span></span>
<span><span class="co">#&gt;   samples: {"S_1_1", "S_1_2", "S_2_1", "S_2_2"}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # of emerged SNVs and indels: 69115</span></span>
<span><span class="co">#&gt;   # of emerged CNAs: 2</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-alexandrov2020repertoire" class="csl-entry">
Alexandrov, Ludmil B, Jaegil Kim, Nicholas J Haradhvala, Mi Ni Huang,
Alvin Wei Tian Ng, Yang Wu, Arnoud Boot, et al. 2020. <span>“The
Repertoire of Mutational Signatures in Human Cancer.”</span>
<em>Nature</em> 578 (7793): 94–101.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
