<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rRACES">
<title>rRACES • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="rRACES">
<meta property="og:description" content="rRACES">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/growth_models.html">Growth Models</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
    <a class="dropdown-item" href="../articles/SCOUT.html">SCOUT</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>rRACES</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/rRACES.Rmd" class="external-link"><code>vignettes/rRACES.Rmd</code></a></small>
      <div class="d-none name"><code>rRACES.Rmd</code></div>
    </div>

    
    
<p>rRACES wraps the <a href="https://github.com/albertocasagrande/RACES" class="external-link">RACES</a> simulation
and data-generation engine by using the <a href="https://www.rcpp.org/" class="external-link">R/C++ Rcpp interface</a>, making it easy to
“program” stochastic tumour evolution.</p>
<div class="section level2">
<h2 id="pipeline-overview">Pipeline overview<a class="anchor" aria-label="anchor" href="#pipeline-overview"></a>
</h2>
<p><img src="img/banner.png" alt="High-level pipeline" width="850"><br></p>
<p>The high-level overview of pipeline of rRACES is a three-steps
process, where:</p>
<ol style="list-style-type: decimal">
<li><p>a discrete tissue is first simulated, with distinct clones
defined by stochastic rates of growth and death. These compete to
colonise the tissue in a standard stochastic birth-death process, which
in rRACES is empowered by a time-varying structure that allows weasily
modelling of complex evolutionary forces that vary over time (e.g.,
therapy-induced negative selection). The tissue can be sampled at
multiple time-points and in multiple spatially-separated
positions.</p></li>
<li><p>from the sampled cells, the simulated phylogenetic history of the
tumour can be extracted. The distance among the involved cells will
represent the lineage history of the simulated clones, and mutational
processes can be used to attach genomes to each simulated cells. rRACES
supports standard germline and somatic mutations, including
single-nucleotide variants (SNVs), insertion-deletions (indels) and copy
number alterations (CNAs). For SNVs and indels time-varying mutational
signatures can also be included.</p></li>
<li><p>from the simulated phylogenetic history of the tumour it is
possible to generate tumour-matched-normal synthetic data for
benchmarking bioinformatics tools, or to infer parameters against real
sequencing data. rRACES output include common VCF formats, as well as
low-level SAM outputs that can be used for aligment and downstream
analysis. Every simulated sample can be assigned custom coverage and
purity.</p></li>
</ol>
<div class="section level3">
<h3 id="main-features">Main features<a class="anchor" aria-label="anchor" href="#main-features"></a>
</h3>
<p>The rRACES engine has a number of features, the most relevant
are:</p>
<ol style="list-style-type: decimal">
<li>cells are associated with species that proliferate in a stochastic
fashion on a 2D lattice;</li>
<li>every species is defined by <em>(i)</em> an abstract “mutant” and
<em>(ii)</em> a epi-state; the epi-state is binary (0/1, on-off). The
combination of the mutant and the epi-state determine the evolutionary
parameters of each species, which can can change over time mimicking
treatment-related evolutionary pressures, for instance.</li>
<li>Species evolution is stochastic and follows a no-back mutations
model, where at every cell division the mutant status heritable, while
the epistate can be stochastically reversible;</li>
<li>at the molecular level, the genome of tumour cells are mutated by
simple mutations (SNVs and indels) and more complex CNAs. Mutations - in
the broad sense - can be either passengers or drivers ones, and can be
linked to realistic mutational processes. All their rates of
accumulation can change over time, mimicking the emergence of
time-varying mutational forces (e.g., to model treatment). Moreover,
realistic germline samples can be simulated by interfacing with the UK
biobank database.</li>
<li>arbitrary tissue sampling schema can be simulated, including
multi-region and longitudinal datasets with any number of samples and
time-points;</li>
<li>realistic read-counts based bulk sequencing data can be generated,
both at the level of whole-genome, whole-exome and targeted panels, as
well as at the level of pre-processed outputs (VCFs) or low-level
sequencing outputs (SAM);</li>
<li>full access to the evolutionary process and its output is available,
allowing easy testing of complex clonal architecture identification
methods, mutation callers, etc.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="detailed-pipeline">Detailed pipeline<a class="anchor" aria-label="anchor" href="#detailed-pipeline"></a>
</h2>
<p><img src="img/short.png" alt="High-level pipeline" width="850"><br></p>
<p>A rRACES simulation follows the steps (some required, some optional)
that we disucssd below. The state of the simulation can be saved (and
resumed) at several time-points.</p>
<div class="section level3">
<h3 id="tissue-simulation">Tissue simulation<a class="anchor" aria-label="anchor" href="#tissue-simulation"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Initialise: a squared tissue is initialised, a species is defined
and a single cell placed onto it;</li>
<li>Evolve tissue: all the cells on the tissue are left to grow
stochastically;</li>
<li>Sample cells: cells can be sampled from the tissue to mimic a
measurement;</li>
<li>Updates/add species: the parameters of each species can be modified,
and cells from new species can be defined and added to the tissue
(subclones).</li>
</ol>
<p>Steps 2-4 represent an iterative interface for “programmable tumour
dynamics” that gives the user the flexibility to code a custom
evolutionary process.</p>
</div>
<div class="section level3">
<h3 id="mutations-generation">Mutations generation<a class="anchor" aria-label="anchor" href="#mutations-generation"></a>
</h3>
<ol style="list-style-type: decimal">
<li>Phylogeny: from sampled cells, a phylogeny is built that reflects
the evolutionary history of the simulated tumour.</li>
<li>Mutational processes: mutational processes can be mapped onto the
temporal-evolution of the process.</li>
<li>Mutational engine: an engine to simulate mutations is built using a
real reference genome.</li>
<li>Map mutations: mutations are stochatically attached to the simulated
phylogeny.</li>
</ol>
</div>
<div class="section level3">
<h3 id="sequencing-data-generation">Sequencing data generation<a class="anchor" aria-label="anchor" href="#sequencing-data-generation"></a>
</h3>
<ol style="list-style-type: decimal">
<li>VCF: given mutations on the phylogenetic tree, custom coverage and
sample purity, a VCF can be simulated to mimic variant-calling with
Beta-Binomial noise.</li>
<li>SAM: from the same mutations it is possible to generate reads for a
tumour matched-normal assay, and fed that data to a custom
biorinformatics pipeline.</li>
</ol>
</div>
<div class="section level3">
<h3 id="cheatsheet">Cheatsheet<a class="anchor" aria-label="anchor" href="#cheatsheet"></a>
</h3>
<p><a href="https://github.com/caravagnalab/rRACES/blob/06e28114d5db423fac2a4c42d1b5a04bf44091d0/vignettes/img/cheatsheet.pdf?raw=true" class="external-link"><img src="https://github.com/caravagnalab/rRACES/blob/434f364bcfdff4bbfa612ac67b6557b9d872d71b/img/vignettes/cheatsheet.png?raw=true" height="252" alt="thumbnail of rRACES cheatsheet"></a></p>
<!-- The best way to learn to use rRACES is to go thorough three  vignettes that define the standard workflow: -->
<!-- -   [*tissue simulation:*](tissue_simulation.html) simulate the spatio-temporal stochastic process of tumour evolution to define a tissue; -->
<!-- -   [*sampling:*](sampling.html) determine the sampling strategy for the simulated tissue; -->
<!-- -   [*growth models:*](growth_models.html) select one the supported growth models;  -->
<!-- -   [*mutation simulation:*](mutations.html) simulate mutations on samples forests;  -->
<!-- -   *molecular data genereation*: generating realistic sequencing data based on sampling and simulation. -->
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Giulio Caravagna, Alberto Casagrande.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
