<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Partitioning samples (FACS) • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Partitioning samples (FACS)">
<meta name="description" content="How to partitioning samples, sorting cells, and simulate FACS
">
<meta property="og:description" content="How to partitioning samples, sorting cells, and simulate FACS
">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rRACES.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a></li>
    <li><a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a></li>
    <li><a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a></li>
    <li><a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Partitioning samples (FACS)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACES/vignettes/sample_partition.Rmd" class="external-link"><code>vignettes/sample_partition.Rmd</code></a></small>
      <div class="d-none name"><code>sample_partition.Rmd</code></div>
    </div>

    
    
<blockquote>
<p><em>Note:</em> This article presents advanced topics on sequencing
simulation. Refer to <code><a href="../articles/sequencing.html">vignette("sequencing")</a></code> for an
introduction on the subject.</p>
</blockquote>
<blockquote>
<p><em>Disclaimer:</em> RACES/rRACES internally implements the
probability distributions using the C++11 random number distribution
classes. The standard does not specify their algorithms, and the class
implementations are left free for the compiler. Thus, the simulation
output depends on the compiler used to compile RACES, and because of
that, the results reported in this article may differ from those
obtained by the reader.</p>
</blockquote>
<p>The sequencing simulation supports sample partition analoguous to <a href="https://www.sinobiological.com/category/fcm-facs-facs" class="external-link">FACS</a>.
Users can provide a labelling function for the sampled cells by using
the <code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> parameter <code>cell_labelling</code>.
This function must take in input an object of the class
<code>SampledCell</code> and return a string representing the cell
label. The function <code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> applies the labelling
function to all the sampled cells and partition them according to the
labelling function outputs. The resulting sub-samples are named
according the format
<code>&lt;sample name&gt;_&lt;cell label&gt;</code>.</p>
<p>Let us assume to have built the phylogenetic forest
<code>phylo_forest</code> as detailed in
<code><a href="../articles/mutations.html">vignette("mutations")</a></code>.</p>
<p>The sampled cells can be grouped by epigenetic state as it
follows.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the labelling function parameter has type `SampledCell`</span></span>
<span><span class="va">epi_labelling</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cell</span><span class="op">)</span> <span class="op">{</span></span>
<span></span>
<span>  <span class="co"># epigenetic state is "+", "-", "". However, "+" and "-"</span></span>
<span>  <span class="co"># are not supported in R dataframe column names, so, we</span></span>
<span>  <span class="co"># can replace them by "P" and "N", respectively</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">cell</span><span class="op">$</span><span class="va">epistate</span> <span class="op">==</span> <span class="st">"+"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">"P"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">cell</span><span class="op">$</span><span class="va">epistate</span> <span class="op">==</span> <span class="st">"-"</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">"N"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                            cell_labelling <span class="op">=</span> <span class="va">epi_labelling</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:02s] Reading 22 [█---------------------------------------] 0% [00m:02s] Processing chr. 22 [█---------------------------------------] 0% [00m:08s] Processing chr. 22 [████████--------------------------------] 19% [00m:09s] Processing chr. 22 [████████████████------------------------] 39% [00m:10s] Processing chr. 22 [███████████████████---------------------] 47% [00m:11s] Processing chr. 22 [███████████████████---------------------] 47% [00m:12s] Processing chr. 22 [████████████████████--------------------] 48% [00m:13s] Processing chr. 22 [█████████████████████████---------------] 60% [00m:14s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:15s] Processing chr. 22 [███████████████████████████████---------] 77% [00m:16s] Processing chr. 22 [███████████████████████████████---------] 77% [00m:17s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:18s] Processing chr. 22 [█████████████████████████████████-------] 82% [00m:19s] Processing chr. 22 [███████████████████████████████████-----] 86% [00m:20s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:21s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:22s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:23s] Processing chr. 22 [████████████████████████████████████----] 89% [00m:24s] Processing chr. 22 [████████████████████████████████████----] 89% [00m:25s] Processing chr. 22 [█████████████████████████████████████---] 91% [00m:26s] Processing chr. 22 [███████████████████████████████████████-] 96% [00m:27s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:28s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:29s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:36s] Reads simulated</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span></span>
<span><span class="va">seq_results</span><span class="op">$</span><span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr  chr_pos ref   alt causes  classes S_1_1_N.occurrences S_1_1_N.coverage</span></span>
<span><span class="co">#&gt; 1  22 16051493   G     A        germinal                   1                1</span></span>
<span><span class="co">#&gt; 2  22 16052080   G     A        germinal                   0                0</span></span>
<span><span class="co">#&gt; 3  22 16052167   A AAAAC        germinal                   1                1</span></span>
<span><span class="co">#&gt; 4  22 16053659   A     C        germinal                   0                0</span></span>
<span><span class="co">#&gt; 5  22 16053863   G     A        germinal                   0                0</span></span>
<span><span class="co">#&gt; 6  22 16054740   A     G        germinal                   1                1</span></span>
<span><span class="co">#&gt;   S_1_1_N.VAF S_1_1_P.occurrences S_1_1_P.coverage S_1_1_P.VAF</span></span>
<span><span class="co">#&gt; 1           1                   1                1           1</span></span>
<span><span class="co">#&gt; 2         NaN                   0                0         NaN</span></span>
<span><span class="co">#&gt; 3           1                   0                0         NaN</span></span>
<span><span class="co">#&gt; 4         NaN                   1                1           1</span></span>
<span><span class="co">#&gt; 5         NaN                   0                0         NaN</span></span>
<span><span class="co">#&gt; 6           1                   0                0         NaN</span></span>
<span><span class="co">#&gt;   S_1_2_N.occurrences S_1_2_N.coverage S_1_2_N.VAF S_1_2_P.occurrences</span></span>
<span><span class="co">#&gt; 1                   0                0         NaN                   0</span></span>
<span><span class="co">#&gt; 2                   1                1           1                   0</span></span>
<span><span class="co">#&gt; 3                   2                2           1                   0</span></span>
<span><span class="co">#&gt; 4                   2                2           1                   1</span></span>
<span><span class="co">#&gt; 5                   0                0         NaN                   2</span></span>
<span><span class="co">#&gt; 6                   2                2           1                   0</span></span>
<span><span class="co">#&gt;   S_1_2_P.coverage S_1_2_P.VAF S_2_1_N.occurrences S_2_1_N.coverage S_2_1_N.VAF</span></span>
<span><span class="co">#&gt; 1                0         NaN                   0                0         NaN</span></span>
<span><span class="co">#&gt; 2                0         NaN                   0                0         NaN</span></span>
<span><span class="co">#&gt; 3                0         NaN                   0                0         NaN</span></span>
<span><span class="co">#&gt; 4                1         1.0                   0                0         NaN</span></span>
<span><span class="co">#&gt; 5                4         0.5                   0                0         NaN</span></span>
<span><span class="co">#&gt; 6                0         NaN                   1                1           1</span></span>
<span><span class="co">#&gt;   S_2_1_P.occurrences S_2_1_P.coverage S_2_1_P.VAF S_2_2_N.occurrences</span></span>
<span><span class="co">#&gt; 1                   0                0         NaN                   0</span></span>
<span><span class="co">#&gt; 2                   0                1           0                   0</span></span>
<span><span class="co">#&gt; 3                   1                1           1                   0</span></span>
<span><span class="co">#&gt; 4                   0                0         NaN                   0</span></span>
<span><span class="co">#&gt; 5                   0                0         NaN                   0</span></span>
<span><span class="co">#&gt; 6                   1                1           1                   0</span></span>
<span><span class="co">#&gt;   S_2_2_N.coverage S_2_2_N.VAF S_2_2_P.occurrences S_2_2_P.coverage S_2_2_P.VAF</span></span>
<span><span class="co">#&gt; 1                0         NaN                   1                1         1.0</span></span>
<span><span class="co">#&gt; 2                2           0                   1                2         0.5</span></span>
<span><span class="co">#&gt; 3                0         NaN                   4                4         1.0</span></span>
<span><span class="co">#&gt; 4                0         NaN                   0                0         NaN</span></span>
<span><span class="co">#&gt; 5                0         NaN                   0                1         0.0</span></span>
<span><span class="co">#&gt; 6                0         NaN                   1                1         1.0</span></span>
<span><span class="co">#&gt;   normal_sample.occurrences normal_sample.coverage normal_sample.VAF</span></span>
<span><span class="co">#&gt; 1                         0                      0               NaN</span></span>
<span><span class="co">#&gt; 2                         0                      1                 0</span></span>
<span><span class="co">#&gt; 3                         0                      0               NaN</span></span>
<span><span class="co">#&gt; 4                         0                      0               NaN</span></span>
<span><span class="co">#&gt; 5                         0                      0               NaN</span></span>
<span><span class="co">#&gt; 6                         0                      0               NaN</span></span></code></pre></div>
<p>The same approach can be easily applied to group sampled cells by
mutant name…</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a mutant-name-based labelling function</span></span>
<span><span class="va">mutant_labelling</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cell</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">cell</span><span class="op">$</span><span class="va">mutant</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                            cell_labelling <span class="op">=</span> <span class="va">mutant_labelling</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:02s] Reading 22 [█---------------------------------------] 0% [00m:02s] Processing chr. 22 [█---------------------------------------] 0% [00m:08s] Processing chr. 22 [█████████-------------------------------] 20% [00m:09s] Processing chr. 22 [█████████████████-----------------------] 41% [00m:10s] Processing chr. 22 [██████████████████████------------------] 52% [00m:11s] Processing chr. 22 [███████████████████████████-------------] 66% [00m:12s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:13s] Processing chr. 22 [█████████████████████████████████-------] 82% [00m:14s] Processing chr. 22 [███████████████████████████████████-----] 86% [00m:15s] Processing chr. 22 [█████████████████████████████████████---] 90% [00m:16s] Processing chr. 22 [███████████████████████████████████████-] 95% [00m:17s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:18s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:19s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:20s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:24s] Reads simulated</span></span>
<span></span>
<span><span class="va">seq_results</span><span class="op">$</span><span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr  chr_pos ref   alt causes  classes S_1_1_A.occurrences S_1_1_A.coverage</span></span>
<span><span class="co">#&gt; 1  22 16051493   G     A        germinal                   0                2</span></span>
<span><span class="co">#&gt; 2  22 16052080   G     A        germinal                   0                0</span></span>
<span><span class="co">#&gt; 3  22 16052167   A AAAAC        germinal                   0                0</span></span>
<span><span class="co">#&gt; 4  22 16053659   A     C        germinal                   0                0</span></span>
<span><span class="co">#&gt; 5  22 16054740   A     G        germinal                   1                1</span></span>
<span><span class="co">#&gt; 6  22 16055070   G     A        germinal                   1                1</span></span>
<span><span class="co">#&gt;   S_1_1_A.VAF S_1_2_A.occurrences S_1_2_A.coverage S_1_2_A.VAF</span></span>
<span><span class="co">#&gt; 1           0                   0                0         NaN</span></span>
<span><span class="co">#&gt; 2         NaN                   0                1           0</span></span>
<span><span class="co">#&gt; 3         NaN                   1                1           1</span></span>
<span><span class="co">#&gt; 4         NaN                   2                2           1</span></span>
<span><span class="co">#&gt; 5           1                   1                1           1</span></span>
<span><span class="co">#&gt; 6           1                   0                0         NaN</span></span>
<span><span class="co">#&gt;   S_2_1_A.occurrences S_2_1_A.coverage S_2_1_A.VAF S_2_2_B.occurrences</span></span>
<span><span class="co">#&gt; 1                   1                1           1                   0</span></span>
<span><span class="co">#&gt; 2                   0                1           0                   0</span></span>
<span><span class="co">#&gt; 3                   0                0         NaN                   1</span></span>
<span><span class="co">#&gt; 4                   1                1           1                   0</span></span>
<span><span class="co">#&gt; 5                   0                0         NaN                   2</span></span>
<span><span class="co">#&gt; 6                   0                2           0                   0</span></span>
<span><span class="co">#&gt;   S_2_2_B.coverage S_2_2_B.VAF normal_sample.occurrences normal_sample.coverage</span></span>
<span><span class="co">#&gt; 1                0         NaN                         0                      0</span></span>
<span><span class="co">#&gt; 2                0         NaN                         2                      2</span></span>
<span><span class="co">#&gt; 3                1           1                         0                      0</span></span>
<span><span class="co">#&gt; 4                0         NaN                         1                      1</span></span>
<span><span class="co">#&gt; 5                2           1                         0                      0</span></span>
<span><span class="co">#&gt; 6                0         NaN                         1                      1</span></span>
<span><span class="co">#&gt;   normal_sample.VAF</span></span>
<span><span class="co">#&gt; 1               NaN</span></span>
<span><span class="co">#&gt; 2                 1</span></span>
<span><span class="co">#&gt; 3               NaN</span></span>
<span><span class="co">#&gt; 4                 1</span></span>
<span><span class="co">#&gt; 5               NaN</span></span>
<span><span class="co">#&gt; 6                 1</span></span></code></pre></div>
<p>…, birth time…</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># a birth-time-based labelling function</span></span>
<span><span class="va">birth_time_labelling</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cell</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">cell</span><span class="op">$</span><span class="va">birth_time</span> <span class="op">&gt;</span> <span class="fl">421</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">"YOUNG"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">cell</span><span class="op">$</span><span class="va">birth_time</span> <span class="op">&gt;</span> <span class="fl">321</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">"MIDDLE_AGED"</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">"OLD"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                            cell_labelling <span class="op">=</span> <span class="va">birth_time_labelling</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:02s] Reading 22 [█---------------------------------------] 0% [00m:02s] Processing chr. 22 [█---------------------------------------] 0% [00m:09s] Processing chr. 22 [██--------------------------------------] 3% [00m:10s] Processing chr. 22 [███-------------------------------------] 7% [00m:11s] Processing chr. 22 [███████---------------------------------] 16% [00m:12s] Processing chr. 22 [█████████████---------------------------] 32% [00m:13s] Processing chr. 22 [██████████████████----------------------] 44% [00m:14s] Processing chr. 22 [████████████████████--------------------] 49% [00m:15s] Processing chr. 22 [█████████████████████-------------------] 51% [00m:16s] Processing chr. 22 [██████████████████████------------------] 52% [00m:18s] Processing chr. 22 [██████████████████████████--------------] 63% [00m:19s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:20s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:21s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:22s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:23s] Processing chr. 22 [█████████████████████████████████-------] 81% [00m:24s] Processing chr. 22 [███████████████████████████████████-----] 86% [00m:25s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:26s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:27s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:28s] Processing chr. 22 [█████████████████████████████████████---] 91% [00m:29s] Processing chr. 22 [███████████████████████████████████████-] 96% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:31s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:32s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:33s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:34s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:35s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:43s] Reads simulated</span></span>
<span></span>
<span><span class="va">seq_results</span><span class="op">$</span><span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr  chr_pos ref   alt causes  classes S_1_1_MIDDLE_AGED.occurrences</span></span>
<span><span class="co">#&gt; 1  22 16051493   G     A        germinal                             0</span></span>
<span><span class="co">#&gt; 2  22 16052080   G     A        germinal                             1</span></span>
<span><span class="co">#&gt; 3  22 16052167   A AAAAC        germinal                             0</span></span>
<span><span class="co">#&gt; 4  22 16053659   A     C        germinal                             2</span></span>
<span><span class="co">#&gt; 5  22 16053863   G     A        germinal                             0</span></span>
<span><span class="co">#&gt; 6  22 16054740   A     G        germinal                             0</span></span>
<span><span class="co">#&gt;   S_1_1_MIDDLE_AGED.coverage S_1_1_MIDDLE_AGED.VAF S_1_1_OLD.occurrences</span></span>
<span><span class="co">#&gt; 1                          1                   0.0                     0</span></span>
<span><span class="co">#&gt; 2                          2                   0.5                     0</span></span>
<span><span class="co">#&gt; 3                          0                   NaN                     1</span></span>
<span><span class="co">#&gt; 4                          2                   1.0                     1</span></span>
<span><span class="co">#&gt; 5                          0                   NaN                     0</span></span>
<span><span class="co">#&gt; 6                          0                   NaN                     0</span></span>
<span><span class="co">#&gt;   S_1_1_OLD.coverage S_1_1_OLD.VAF S_1_2_MIDDLE_AGED.occurrences</span></span>
<span><span class="co">#&gt; 1                  0           NaN                             0</span></span>
<span><span class="co">#&gt; 2                  0           NaN                             0</span></span>
<span><span class="co">#&gt; 3                  1             1                             1</span></span>
<span><span class="co">#&gt; 4                  1             1                             0</span></span>
<span><span class="co">#&gt; 5                  0           NaN                             0</span></span>
<span><span class="co">#&gt; 6                  0           NaN                             1</span></span>
<span><span class="co">#&gt;   S_1_2_MIDDLE_AGED.coverage S_1_2_MIDDLE_AGED.VAF S_1_2_OLD.occurrences</span></span>
<span><span class="co">#&gt; 1                          1                     0                     2</span></span>
<span><span class="co">#&gt; 2                          0                   NaN                     2</span></span>
<span><span class="co">#&gt; 3                          1                     1                     3</span></span>
<span><span class="co">#&gt; 4                          0                   NaN                     0</span></span>
<span><span class="co">#&gt; 5                          0                   NaN                     0</span></span>
<span><span class="co">#&gt; 6                          1                     1                     0</span></span>
<span><span class="co">#&gt;   S_1_2_OLD.coverage S_1_2_OLD.VAF S_2_1_MIDDLE_AGED.occurrences</span></span>
<span><span class="co">#&gt; 1                  2             1                             0</span></span>
<span><span class="co">#&gt; 2                  2             1                             1</span></span>
<span><span class="co">#&gt; 3                  3             1                             0</span></span>
<span><span class="co">#&gt; 4                  0           NaN                             0</span></span>
<span><span class="co">#&gt; 5                  0           NaN                             1</span></span>
<span><span class="co">#&gt; 6                  0           NaN                             1</span></span>
<span><span class="co">#&gt;   S_2_1_MIDDLE_AGED.coverage S_2_1_MIDDLE_AGED.VAF S_2_1_OLD.occurrences</span></span>
<span><span class="co">#&gt; 1                          0                   NaN                     0</span></span>
<span><span class="co">#&gt; 2                          1                   1.0                     0</span></span>
<span><span class="co">#&gt; 3                          0                   NaN                     1</span></span>
<span><span class="co">#&gt; 4                          0                   NaN                     2</span></span>
<span><span class="co">#&gt; 5                          2                   0.5                     0</span></span>
<span><span class="co">#&gt; 6                          1                   1.0                     1</span></span>
<span><span class="co">#&gt;   S_2_1_OLD.coverage S_2_1_OLD.VAF S_2_1_YOUNG.occurrences S_2_1_YOUNG.coverage</span></span>
<span><span class="co">#&gt; 1                  0           NaN                       1                    1</span></span>
<span><span class="co">#&gt; 2                  1             0                       0                    0</span></span>
<span><span class="co">#&gt; 3                  1             1                       0                    0</span></span>
<span><span class="co">#&gt; 4                  2             1                       0                    0</span></span>
<span><span class="co">#&gt; 5                  1             0                       0                    1</span></span>
<span><span class="co">#&gt; 6                  1             1                       1                    1</span></span>
<span><span class="co">#&gt;   S_2_1_YOUNG.VAF S_2_2_MIDDLE_AGED.occurrences S_2_2_MIDDLE_AGED.coverage</span></span>
<span><span class="co">#&gt; 1               1                             0                          0</span></span>
<span><span class="co">#&gt; 2             NaN                             0                          0</span></span>
<span><span class="co">#&gt; 3             NaN                             0                          0</span></span>
<span><span class="co">#&gt; 4             NaN                             0                          0</span></span>
<span><span class="co">#&gt; 5               0                             0                          0</span></span>
<span><span class="co">#&gt; 6               1                             0                          0</span></span>
<span><span class="co">#&gt;   S_2_2_MIDDLE_AGED.VAF S_2_2_YOUNG.occurrences S_2_2_YOUNG.coverage</span></span>
<span><span class="co">#&gt; 1                   NaN                       0                    1</span></span>
<span><span class="co">#&gt; 2                   NaN                       0                    1</span></span>
<span><span class="co">#&gt; 3                   NaN                       0                    0</span></span>
<span><span class="co">#&gt; 4                   NaN                       2                    2</span></span>
<span><span class="co">#&gt; 5                   NaN                       1                    1</span></span>
<span><span class="co">#&gt; 6                   NaN                       0                    0</span></span>
<span><span class="co">#&gt;   S_2_2_YOUNG.VAF normal_sample.occurrences normal_sample.coverage</span></span>
<span><span class="co">#&gt; 1               0                         0                      1</span></span>
<span><span class="co">#&gt; 2               0                         0                      1</span></span>
<span><span class="co">#&gt; 3             NaN                         1                      1</span></span>
<span><span class="co">#&gt; 4               1                         0                      0</span></span>
<span><span class="co">#&gt; 5               1                         0                      0</span></span>
<span><span class="co">#&gt; 6             NaN                         1                      1</span></span>
<span><span class="co">#&gt;   normal_sample.VAF</span></span>
<span><span class="co">#&gt; 1                 0</span></span>
<span><span class="co">#&gt; 2                 0</span></span>
<span><span class="co">#&gt; 3                 1</span></span>
<span><span class="co">#&gt; 4               NaN</span></span>
<span><span class="co">#&gt; 5               NaN</span></span>
<span><span class="co">#&gt; 6                 1</span></span></code></pre></div>
<p>…, mutations…</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># collect all the sample passenger indels</span></span>
<span><span class="va">passenger_indels</span> <span class="op">&lt;-</span> <span class="va">phylo_forest</span><span class="op">$</span><span class="fu">get_sampled_cell_mutations</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">class</span> <span class="op">==</span> <span class="st">"passenger"</span>, <span class="va">type</span> <span class="op">==</span> <span class="st">"indel"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get one of the passenger indels</span></span>
<span><span class="va">p_indel</span> <span class="op">&lt;-</span> <span class="va">passenger_indels</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">passenger_indels</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span><span class="va">p_indel</span></span>
<span><span class="co">#&gt;     cell_id chr  chr_pos allele ref                   alt  type cause     class</span></span>
<span><span class="co">#&gt; 485   28139  22 31512507      1   C CTAAATAAATAAATAAATAAA indel   ID2 passenger</span></span>
<span></span>
<span><span class="co"># a mutation-based labelling function that discriminates sampled cells</span></span>
<span><span class="co"># containing indel `p_indel` from the other sampled cells</span></span>
<span><span class="va">mutations_labelling</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">cell</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">has_indel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cell</span><span class="op">$</span><span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>                      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">chr</span> <span class="op">==</span> <span class="va">p_indel</span><span class="op">[[</span><span class="st">"chr"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                             <span class="va">chr_pos</span> <span class="op">==</span> <span class="va">p_indel</span><span class="op">[[</span><span class="st">"chr_pos"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                             <span class="va">ref</span> <span class="op">==</span> <span class="va">p_indel</span><span class="op">[[</span><span class="st">"ref"</span><span class="op">]</span><span class="op">]</span>,</span>
<span>                             <span class="va">alt</span> <span class="op">==</span> <span class="va">p_indel</span><span class="op">[[</span><span class="st">"alt"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span></span>
<span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="va">has_indel</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">"HAS_MUTATION"</span><span class="op">)</span>;</span>
<span>  <span class="op">}</span></span>
<span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="st">"MISSES_MUTATION"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                            cell_labelling <span class="op">=</span> <span class="va">mutations_labelling</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:02s] Reading 22 [█---------------------------------------] 0% [00m:02s] Processing chr. 22 [█---------------------------------------] 0% [00m:08s] Processing chr. 22 [████████--------------------------------] 19% [00m:09s] Processing chr. 22 [████████████████------------------------] 38% [00m:10s] Processing chr. 22 [███████████████████---------------------] 45% [00m:11s] Processing chr. 22 [███████████████████---------------------] 46% [00m:12s] Processing chr. 22 [███████████████████---------------------] 47% [00m:13s] Processing chr. 22 [█████████████████████-------------------] 52% [00m:14s] Processing chr. 22 [███████████████████████████-------------] 65% [00m:15s] Processing chr. 22 [███████████████████████████████---------] 77% [00m:16s] Processing chr. 22 [█████████████████████████████████-------] 80% [00m:17s] Processing chr. 22 [███████████████████████████████████-----] 85% [00m:18s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:19s] Processing chr. 22 [██████████████████████████████████████--] 93% [00m:20s] Processing chr. 22 [████████████████████████████████████████] 98% [00m:21s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:22s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:23s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:28s] Reads simulated</span></span>
<span></span>
<span><span class="va">seq_results</span><span class="op">$</span><span class="va">mutations</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   chr  chr_pos ref   alt causes  classes S_1_1_HAS_MUTATION.occurrences</span></span>
<span><span class="co">#&gt; 1  22 16051493   G     A        germinal                              0</span></span>
<span><span class="co">#&gt; 2  22 16052080   G     A        germinal                              0</span></span>
<span><span class="co">#&gt; 3  22 16052167   A AAAAC        germinal                              0</span></span>
<span><span class="co">#&gt; 4  22 16053659   A     C        germinal                              0</span></span>
<span><span class="co">#&gt; 5  22 16053863   G     A        germinal                              0</span></span>
<span><span class="co">#&gt; 6  22 16054740   A     G        germinal                              2</span></span>
<span><span class="co">#&gt;   S_1_1_HAS_MUTATION.coverage S_1_1_HAS_MUTATION.VAF</span></span>
<span><span class="co">#&gt; 1                           2                      0</span></span>
<span><span class="co">#&gt; 2                           0                    NaN</span></span>
<span><span class="co">#&gt; 3                           0                    NaN</span></span>
<span><span class="co">#&gt; 4                           0                    NaN</span></span>
<span><span class="co">#&gt; 5                           0                    NaN</span></span>
<span><span class="co">#&gt; 6                           2                      1</span></span>
<span><span class="co">#&gt;   S_1_1_MISSES_MUTATION.occurrences S_1_1_MISSES_MUTATION.coverage</span></span>
<span><span class="co">#&gt; 1                                 0                              0</span></span>
<span><span class="co">#&gt; 2                                 0                              1</span></span>
<span><span class="co">#&gt; 3                                 3                              3</span></span>
<span><span class="co">#&gt; 4                                 0                              0</span></span>
<span><span class="co">#&gt; 5                                 0                              0</span></span>
<span><span class="co">#&gt; 6                                 0                              0</span></span>
<span><span class="co">#&gt;   S_1_1_MISSES_MUTATION.VAF S_1_2_MISSES_MUTATION.occurrences</span></span>
<span><span class="co">#&gt; 1                       NaN                                 0</span></span>
<span><span class="co">#&gt; 2                         0                                 0</span></span>
<span><span class="co">#&gt; 3                         1                                 0</span></span>
<span><span class="co">#&gt; 4                       NaN                                 1</span></span>
<span><span class="co">#&gt; 5                       NaN                                 0</span></span>
<span><span class="co">#&gt; 6                       NaN                                 0</span></span>
<span><span class="co">#&gt;   S_1_2_MISSES_MUTATION.coverage S_1_2_MISSES_MUTATION.VAF</span></span>
<span><span class="co">#&gt; 1                              0                       NaN</span></span>
<span><span class="co">#&gt; 2                              0                       NaN</span></span>
<span><span class="co">#&gt; 3                              0                       NaN</span></span>
<span><span class="co">#&gt; 4                              1                         1</span></span>
<span><span class="co">#&gt; 5                              1                         0</span></span>
<span><span class="co">#&gt; 6                              0                       NaN</span></span>
<span><span class="co">#&gt;   S_2_1_MISSES_MUTATION.occurrences S_2_1_MISSES_MUTATION.coverage</span></span>
<span><span class="co">#&gt; 1                                 1                              1</span></span>
<span><span class="co">#&gt; 2                                 1                              1</span></span>
<span><span class="co">#&gt; 3                                 0                              0</span></span>
<span><span class="co">#&gt; 4                                 0                              0</span></span>
<span><span class="co">#&gt; 5                                 0                              1</span></span>
<span><span class="co">#&gt; 6                                 0                              0</span></span>
<span><span class="co">#&gt;   S_2_1_MISSES_MUTATION.VAF S_2_2_HAS_MUTATION.occurrences</span></span>
<span><span class="co">#&gt; 1                         1                              0</span></span>
<span><span class="co">#&gt; 2                         1                              0</span></span>
<span><span class="co">#&gt; 3                       NaN                              0</span></span>
<span><span class="co">#&gt; 4                       NaN                              0</span></span>
<span><span class="co">#&gt; 5                         0                              1</span></span>
<span><span class="co">#&gt; 6                       NaN                              1</span></span>
<span><span class="co">#&gt;   S_2_2_HAS_MUTATION.coverage S_2_2_HAS_MUTATION.VAF normal_sample.occurrences</span></span>
<span><span class="co">#&gt; 1                           0                    NaN                         0</span></span>
<span><span class="co">#&gt; 2                           0                    NaN                         0</span></span>
<span><span class="co">#&gt; 3                           0                    NaN                         1</span></span>
<span><span class="co">#&gt; 4                           0                    NaN                         0</span></span>
<span><span class="co">#&gt; 5                           2                    0.5                         0</span></span>
<span><span class="co">#&gt; 6                           1                    1.0                         0</span></span>
<span><span class="co">#&gt;   normal_sample.coverage normal_sample.VAF</span></span>
<span><span class="co">#&gt; 1                      0               NaN</span></span>
<span><span class="co">#&gt; 2                      0               NaN</span></span>
<span><span class="co">#&gt; 3                      1                 1</span></span>
<span><span class="co">#&gt; 4                      0               NaN</span></span>
<span><span class="co">#&gt; 5                      0               NaN</span></span>
<span><span class="co">#&gt; 6                      0               NaN</span></span></code></pre></div>
<p>… or combination of these properties.</p>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
