<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rRACES">
<title>SCOUT â€¢ rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="SCOUT">
<meta property="og:description" content="rRACES">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/growth_models.html">Growth Models</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
    <a class="dropdown-item" href="../articles/SCOUT.html">SCOUT</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>SCOUT</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/SCOUT.Rmd" class="external-link"><code>vignettes/SCOUT.Rmd</code></a></small>
      <div class="d-none name"><code>SCOUT.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rRACES</span><span class="op">)</span></span></code></pre></div>
<p>The Simulated Cohort of Universal Tumours (SCOUT) is a cohort
generated with rRACES, containing prototypical tumours with matched
sequencing data. As in other fields of Machine Learning, the intent of
SCOUT is to become a standard to benchmark algorithms for:</p>
<ul>
<li>mutation calling;</li>
<li>copy number alteration calling;</li>
<li>mutational signatures deconvolution;</li>
<li>tumour subclonal deconvolution.</li>
</ul>
<p>The main features of SCOUT are:</p>
<ul>
<li>a cohort that contains realistic tumours inspired from
adenocarcinomas (colorectal and lung), leukemias (chronic lymphocitic
and acute myeloid), breast cancer and glioblastoma;</li>
<li>for each simulated tumour, both multi-region and/or longitudinal
data, with different configurations of sample purity and sequencing
coverage;</li>
<li>for some tumours, the simulated effect of treatment following
typical patterns for each tumour type, including resistance caused by
pre-existing or de novo tumour clones;</li>
<li>data available in the format of simulated read counts format, or as
raw seqeuncing reads (FASTq) to be streamlined in any bioinformatic
pipeline, with raw data in the format of whole-genome or whole-exome
sequencing.</li>
</ul>
<p>With a total of 30 TB of data, SCOUT samples can be used to benchmark
pretty much most standard computational methods commonly used in the
cancer genomics field.</p>
<div class="section level2">
<h2 id="summary-statistics">Summary statistics<a class="anchor" aria-label="anchor" href="#summary-statistics"></a>
</h2>
<p>Once the cohort is generated, we could report total number of
mutations, CNas, clones etc etc.</p>
</div>
<div class="section level2">
<h2 id="simulated-tumours">Simulated tumours<a class="anchor" aria-label="anchor" href="#simulated-tumours"></a>
</h2>
<table style="width:99%;" class="table">
<colgroup>
<col width="4%">
<col width="4%">
<col width="9%">
<col width="81%">
</colgroup>
<thead><tr class="header">
<th>Tumour</th>
<th>
<p>Type</p>
<p>(clones)</p>
</th>
<th>Samples (timepoints)</th>
<th>Synopsis</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="#spn01">SPN01</a></td>
<td>P (n=4)</td>
<td>b=3 (t=1)</td>
<td>Micro-satellite stable (MSS) colorectal cancer composed of a single
clonal population, sampled at a single time-point with two multi-region
biopsies.</td>
</tr>
<tr class="even">
<td><a href="#spn02">SPN02</a></td>
<td>P (n=3)</td>
<td>b=2 (t=1)</td>
<td>Micro-satellite instable (MSI) colorectal cancer composed of a main
clonal population, sampled at a single time-point with two multi-region
biopsies.</td>
</tr>
<tr class="odd">
<td><a href="#spn03">SPN03</a></td>
<td>P (n=3)</td>
<td>b=4 (t=4)</td>
<td>Chronic lympocitic laeukemia in a watch-and-wait (i.e., untreated)
scenario, with multiple subclones sampled longitudinally over four
time-points.</td>
</tr>
<tr class="even">
<td><a href="#spn04">SPN04</a></td>
<td>P/R (n=3)</td>
<td>b=2 (t=2)</td>
<td>Acute myeloyd laeukemia with a dominant clone, relapsing
platinum-based chemotherapy with, sampled longitudinally before
treatment and at relapse.</td>
</tr>
<tr class="odd">
<td><a href="#spn05">SPN05</a></td>
<td>P (n=5)</td>
<td>b=3 (t=1)</td>
<td>Mismatch deficient breast cancer, with many subclones, sampled at
one time-point with with three multi-region biopsies.</td>
</tr>
<tr class="even">
<td><a href="#spn06">SPN06</a></td>
<td>P/R (n=6)</td>
<td>b=5 (t=3)</td>
<td>Smoking-associated lung adenocardinoma with many subclones, treated
with two lines of chemotherapy, both time relapsing with a de-novo
clone, sampled at three time-points with five multi-region
biopsies.</td>
</tr>
<tr class="odd">
<td><a href="#spn07">SPN07</a></td>
<td>P/R (n=6)</td>
<td>b=5 (t=2)</td>
<td>Glioblastoma with many subclones, treated with a line of
chemotherapy, relapsing with a hypermutant clone, sampled at three
time-points with five multi-region biopsies.</td>
</tr>
</tbody>
</table>
<p><strong>Legend:</strong> Primary (P), Relapse (R), Number of
simulated clones (n), Number of sampled time-points (t), Total number of
biopsies (b).</p>
<div class="section level3">
<h3 id="spn01">SPN01<a class="anchor" aria-label="anchor" href="#spn01"></a>
</h3>
<p><img src="img/SCOUT/SPN01.png" alt="SPN01" width="850"><br></p>
<p>SPN01 is a prototype micro-satellite stable (MSS) colorectal cancer
composed of 4 clones, with one dominant clone at sampling. The dominant
clone harbours a full inactivation of the APC gene by point mutation and
concurrent loss of heterozygosity (LOH) of the wildtype allele, together
with two activating mutations in the KRAS and PIK3CA oncogenes.</p>
<p>SPN01 is profiled by multi-region with 3 samples collected at the
same time-point.</p>
</div>
<div class="section level3">
<h3 id="spn02">SPN02<a class="anchor" aria-label="anchor" href="#spn02"></a>
</h3>
<p><img src="img/SCOUT/SPN02.png" alt="SPN02" width="850"><br></p>
<p>SPN02 is a prototype micro-satellite instable (MSI) colorectal cancer
composed of 3 clones, with one dominant clone at sampling. The dominant
clone harbours a mutant BRAF, PIK3CA and MSH6. The MSH6 mutation makes
the tumour hypermutant, enabling a specific mutational signature.</p>
<p>SPN02 is profiled by multi-region with 2 samples collected at the
same time-point.</p>
</div>
<div class="section level3">
<h3 id="spn03">SPN03<a class="anchor" aria-label="anchor" href="#spn03"></a>
</h3>
<p><img src="img/SCOUT/SPN03.png" alt="SPN03" width="850"><br></p>
<p>SPN03 is a prototype chronic lymphocitic leukemia composed of 3
clones, with eventually one largely dominant but one still expanding.
The largest clone harbours activating mutations in NOTCH1 and KRAS, and
the last expanding subclone is not linked to any specific genetic
mutation or copy number driver event.</p>
<p>SPN03 is profiled in a watch-and-wait scenario (an active
survaillance scenario in which the patient is observed untill white
blood counts exceed a predefined threshould), acquiring one sample at
each of 4 time-points, with distinct clonal compositions reflecting the
ongoing dynamics.</p>
</div>
<div class="section level3">
<h3 id="spn04">SPN04<a class="anchor" aria-label="anchor" href="#spn04"></a>
</h3>
<p><img src="img/SCOUT/SPN04.png" alt="SPN04" width="850"><br></p>
<p>SPN02 is a prototype acute myeloyd leukemia composed of 3 clones,
with one becoming dominant at every sampling. The dominant clone
harbours a mutant IDH1, a copy number gain of KRAS and a mutant
NRAS.</p>
<p>SPN01 is profiled before and after platinum-based chemotherapy,
relapsing with the same clone presented before treatment. Chemotherapy
generates a mutational signature specific during its exposure.</p>
</div>
<div class="section level3">
<h3 id="spn05">SPN05<a class="anchor" aria-label="anchor" href="#spn05"></a>
</h3>
<p><img src="img/SCOUT/SPN05.png" alt="SPN05" width="850"><br></p>
<p>SPN05 is a prototype mismatch-repair deficient breast cancer composed
of 5 clones, with three detectable at sampling. The ancestral clone
harbours a mutant TP53, and a full BRCA2 inactivation by joint mutation
and LOH (also linked to a specific mutational signature). The two
expanding subclones are both associated with copy numbers: one has
ongoing TP53 LOH, the other CDKN2A LOH.</p>
<p>SPN05 is profiled at a single time-point with three biopsies,
reflecting the distinct tumour composition in terms of clones
distribution.</p>
</div>
<div class="section level3">
<h3 id="spn06">SPN06<a class="anchor" aria-label="anchor" href="#spn06"></a>
</h3>
<p><img src="img/SCOUT/SPN06.png" alt="SPN06" width="850"><br></p>
<p>SPN06 is a prototype smoking-associated lung adenocarcinoma, with 6
clones that compete over time during two lines of treatment. The main
clonal population is driven by mutant TP53 and STK11, but manifests two
subclones: one associated with an EGFR amplification, and one by a KEAP1
mutation. Upon chemo-immunotherapy (which leaves a mutational
signature), a KRAS-driven subclone drives relapse, whereas the two other
subclones are eradicated by treatment. Upon a second line of
chemotherapy (which does not leave a mutational signature), resistance
is driven by an aquired amplification of the mutant KRAS allele.</p>
<p>SPN06 is profiled at a three time-points, before treatment and at
relapse of each treatment. The two diagnostic biopsies are polyclonal,
whereas relapse biopsies are monoclonal.</p>
</div>
<div class="section level3">
<h3 id="spn07">SPN07<a class="anchor" aria-label="anchor" href="#spn07"></a>
</h3>
<p><img src="img/SCOUT/SPN07.png" alt="SPN07" width="850"><br></p>
<p>SPN06 is a prototype glioblastoma with 6 clones that compete over
time during a single line of treatment. The main clonal population is
driven by inactivated PTEN (mutation and LOH), but manifests two
subclones: one associated with an ATRX1 mutation, and one by a NF1
mutation. Upon chemotherapy by timolozomide (which causes an MSH6
mutation and microsatellite instability), resistance is achieved by a
hypermutant subclone. Within the relapse clone, an extra subclone
emerges due to a TP53 mutation.</p>
<p>SPN07 is profiled at a two time-points, before treatment and at
relapse of each treatment. The three diagnostic and two relapse biopsies
are polyclonal.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Giulio Caravagna, Alberto Casagrande.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
