<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rRACES">
<title>Sequencing Simulation • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Sequencing Simulation">
<meta property="og:description" content="rRACES">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/growth_models.html">Growth Models</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
    <a class="dropdown-item" href="../articles/SCOUT.html">SCOUT</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Sequencing Simulation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/sequencing.Rmd" class="external-link"><code>vignettes/sequencing.Rmd</code></a></small>
      <div class="d-none name"><code>sequencing.Rmd</code></div>
    </div>

    
    
<p>Once a phylogenetic forest has been computed, as detailed in
<code><a href="../articles/mutations.html">vignette("mutations")</a></code>, rRACES can simulate the sequencing
of the samples in the forest and return the observed data.</p>
<div class="section level3">
<h3 id="sequencing-simulation">Sequencing Simulation<a class="anchor" aria-label="anchor" href="#sequencing-simulation"></a>
</h3>
<p>Let us reload the phylogenetic forest produced in
<code><a href="../articles/mutations.html">vignette("mutations")</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rRACES</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylo_forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_phylogenetic_forest.html">load_phylogenetic_forest</a></span><span class="op">(</span><span class="st">"phylo_forest.sff"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Loading forest [████████████----------------------------] 29% [00m:01s] Loading forest [████████████████████████████------------] 69% [00m:02s] Loading forest [████████████████████████████████████████] 100% [00m:02s] Forest loaded</span></span>
<span></span>
<span><span class="va">phylo_forest</span></span>
<span><span class="co">#&gt; PhylogeneticForest</span></span>
<span><span class="co">#&gt;   # of trees: 1</span></span>
<span><span class="co">#&gt;   # of nodes: 4759</span></span>
<span><span class="co">#&gt;   # of leaves: 949</span></span>
<span><span class="co">#&gt;   samples: {"S_1_1", "S_1_2", "S_2_1", "S_2_2"}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # of emerged SNVs and indels: 69115</span></span>
<span><span class="co">#&gt;   # of emerged CNAs: 2</span></span></code></pre></div>
<p>The loaded phylogenetic forest models the cell evolution of 4
different samples: <code>S_1_1</code>, <code>S_1_2</code>,
<code>S_2_1</code>, and <code>S_2_2</code>.</p>
<p>We can simulate the sequencing of these samples with coverage of 2.5X
by as it follows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># let us simulate a 2.5x sequencing of the four sample</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [█---------------------------------------] 1% [00m:02s] Processing chr. 22 [███-------------------------------------] 7% [00m:03s] Processing chr. 22 [██████----------------------------------] 14% [00m:04s] Processing chr. 22 [█████████-------------------------------] 20% [00m:05s] Processing chr. 22 [███████████-----------------------------] 26% [00m:06s] Processing chr. 22 [█████████████---------------------------] 32% [00m:07s] Processing chr. 22 [████████████████------------------------] 39% [00m:08s] Processing chr. 22 [███████████████████---------------------] 45% [00m:09s] Processing chr. 22 [█████████████████████-------------------] 51% [00m:10s] Processing chr. 22 [███████████████████████-----------------] 56% [00m:11s] Processing chr. 22 [█████████████████████████---------------] 61% [00m:12s] Processing chr. 22 [███████████████████████████-------------] 66% [00m:13s] Processing chr. 22 [█████████████████████████████-----------] 71% [00m:14s] Processing chr. 22 [███████████████████████████████---------] 77% [00m:15s] Processing chr. 22 [█████████████████████████████████-------] 82% [00m:16s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:17s] Processing chr. 22 [██████████████████████████████████████--] 93% [00m:18s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:19s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:20s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:21s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:22s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:23s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:25s] Read simulated</span></span>
<span></span>
<span><span class="co"># let us load the dplyr library to filter the `simulate_seq` output</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">#&gt;   chr  chr_pos ref   alt causes   classes S_1_1.occurrences S_1_1.coverage</span></span>
<span><span class="co">#&gt; 1  22 16051493   G     A   &lt;NA&gt;  germinal                 1              1</span></span>
<span><span class="co">#&gt; 2  22 16052080   G     A   &lt;NA&gt;  germinal                 1              2</span></span>
<span><span class="co">#&gt; 3  22 16052167   A AAAAC   &lt;NA&gt;  germinal                 4              4</span></span>
<span><span class="co">#&gt; 4  22 16053659   A     C   &lt;NA&gt;  germinal                 3              3</span></span>
<span><span class="co">#&gt; 5  22 16053743   G     A   SBS2 passenger                 0              2</span></span>
<span><span class="co">#&gt; 6  22 16053863   G     A   &lt;NA&gt;  germinal                 1              1</span></span>
<span><span class="co">#&gt;   S_1_1.VAF S_1_2.occurrences S_1_2.coverage S_1_2.VAF S_2_1.occurrences</span></span>
<span><span class="co">#&gt; 1       1.0                 3              6      0.50                 4</span></span>
<span><span class="co">#&gt; 2       0.5                 3              4      0.75                 3</span></span>
<span><span class="co">#&gt; 3       1.0                 5              5      1.00                 1</span></span>
<span><span class="co">#&gt; 4       1.0                 2              2      1.00                 4</span></span>
<span><span class="co">#&gt; 5       0.0                 0              1      0.00                 1</span></span>
<span><span class="co">#&gt; 6       1.0                 0              1      0.00                 1</span></span>
<span><span class="co">#&gt;   S_2_1.coverage S_2_1.VAF S_2_2.occurrences S_2_2.coverage S_2_2.VAF</span></span>
<span><span class="co">#&gt; 1              5 0.8000000                 1              4      0.25</span></span>
<span><span class="co">#&gt; 2              4 0.7500000                 1              2      0.50</span></span>
<span><span class="co">#&gt; 3              1 1.0000000                 3              3      1.00</span></span>
<span><span class="co">#&gt; 4              4 1.0000000                 1              1      1.00</span></span>
<span><span class="co">#&gt; 5              3 0.3333333                 0              1      0.00</span></span>
<span><span class="co">#&gt; 6              2 0.5000000                 0              3      0.00</span></span>
<span><span class="co">#&gt;   normal_sample.occurrences normal_sample.coverage normal_sample.VAF</span></span>
<span><span class="co">#&gt; 1                         0                      2                 0</span></span>
<span><span class="co">#&gt; 2                         0                      0               NaN</span></span>
<span><span class="co">#&gt; 3                         1                      1                 1</span></span>
<span><span class="co">#&gt; 4                         0                      0               NaN</span></span>
<span><span class="co">#&gt; 5                         0                      3                 0</span></span>
<span><span class="co">#&gt; 6                         0                      3                 0</span></span></code></pre></div>
<p>The function <code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> returns a data frame whose
rows represent the mutations observed in the simulated reads (currently,
only SNVs are supported).</p>
<p>The number of columns in the data frame depends on the number of
samples in the phylogenetic forest. The first four columns describe the
mutation and report the chromosome and the position in the chromosome of
the mutation (columns <code>chr</code> and <code>chr_pos</code>,
respectively), the genetic context in which the mutation occurs (column
<code>context</code>), and the new base (column <code>alt_base</code>).
Then, there are three columns for each of the samples: they contain the
number of simulated reads affected by the mutation, the sequencing
coverage of the mutation locus, and the ratio between these two last
columns (columns <code>&lt;sample name&gt; occurrences</code>,
<code>&lt;sample name&gt; coverage</code>, and
<code>&lt;sample name&gt; VAF</code>, respectively).</p>
</div>
<div class="section level3">
<h3 id="cell-partition-by-epigenetic-state-epi-facs">Cell Partition by Epigenetic State (Epi-FACS)<a class="anchor" aria-label="anchor" href="#cell-partition-by-epigenetic-state-epi-facs"></a>
</h3>
<p>The sequencing simulation supports cell partition by epigenetic state
(epi-FACS) using the <code>simualate_seq()</code> parameter
<code>epi_FACS</code>. At the beginning of the sequencing simulation
process, each sample in the phylogenetic forest is split into two
distinct samples: <code>&lt;sample name&gt;_P</code>, which contains all
the cells in the original sample having epigenetic state <code>+</code>,
and <code>&lt;sample name&gt;_N</code>, that consists in the cells
having epigenetic state <code>-</code>. The result of the simulation
reports the data for the split samples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>, epi_FACS <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [█---------------------------------------] 1% [00m:02s] Processing chr. 22 [███-------------------------------------] 7% [00m:03s] Processing chr. 22 [██████----------------------------------] 14% [00m:04s] Processing chr. 22 [█████████-------------------------------] 20% [00m:05s] Processing chr. 22 [███████████-----------------------------] 25% [00m:06s] Processing chr. 22 [███████████-----------------------------] 26% [00m:07s] Processing chr. 22 [███████████-----------------------------] 26% [00m:08s] Processing chr. 22 [███████████-----------------------------] 26% [00m:09s] Processing chr. 22 [███████████-----------------------------] 27% [00m:10s] Processing chr. 22 [██████████████--------------------------] 33% [00m:11s] Processing chr. 22 [████████████████------------------------] 39% [00m:12s] Processing chr. 22 [███████████████████---------------------] 45% [00m:13s] Processing chr. 22 [█████████████████████-------------------] 51% [00m:14s] Processing chr. 22 [█████████████████████-------------------] 52% [00m:15s] Processing chr. 22 [█████████████████████-------------------] 52% [00m:16s] Processing chr. 22 [██████████████████████------------------] 52% [00m:17s] Processing chr. 22 [██████████████████████------------------] 52% [00m:18s] Processing chr. 22 [███████████████████████-----------------] 56% [00m:19s] Processing chr. 22 [█████████████████████████---------------] 61% [00m:20s] Processing chr. 22 [███████████████████████████-------------] 66% [00m:21s] Processing chr. 22 [█████████████████████████████-----------] 71% [00m:22s] Processing chr. 22 [██████████████████████████████----------] 74% [00m:23s] Processing chr. 22 [███████████████████████████████---------] 75% [00m:24s] Processing chr. 22 [███████████████████████████████---------] 75% [00m:25s] Processing chr. 22 [███████████████████████████████---------] 76% [00m:26s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:27s] Processing chr. 22 [██████████████████████████████████------] 83% [00m:28s] Processing chr. 22 [████████████████████████████████████----] 89% [00m:29s] Processing chr. 22 [██████████████████████████████████████--] 94% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 98% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 98% [00m:31s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:32s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:33s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:34s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:35s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:36s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:37s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:38s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:42s] Read simulated</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">#&gt;   chr  chr_pos ref   alt causes  classes S_1_1_N.occurrences S_1_1_N.coverage</span></span>
<span><span class="co">#&gt; 1  22 16051493   G     A   &lt;NA&gt; germinal                   1                2</span></span>
<span><span class="co">#&gt; 2  22 16052080   G     A   &lt;NA&gt; germinal                   0                1</span></span>
<span><span class="co">#&gt; 3  22 16052167   A AAAAC   &lt;NA&gt; germinal                   2                2</span></span>
<span><span class="co">#&gt; 4  22 16053659   A     C   &lt;NA&gt; germinal                   2                2</span></span>
<span><span class="co">#&gt; 5  22 16053863   G     A   &lt;NA&gt; germinal                   1                1</span></span>
<span><span class="co">#&gt; 6  22 16054740   A     G   &lt;NA&gt; germinal                   1                1</span></span>
<span><span class="co">#&gt;   S_1_1_N.VAF S_1_1_P.occurrences S_1_1_P.coverage S_1_1_P.VAF</span></span>
<span><span class="co">#&gt; 1         0.5                   2                2   1.0000000</span></span>
<span><span class="co">#&gt; 2         0.0                   2                4   0.5000000</span></span>
<span><span class="co">#&gt; 3         1.0                   3                3   1.0000000</span></span>
<span><span class="co">#&gt; 4         1.0                   3                3   1.0000000</span></span>
<span><span class="co">#&gt; 5         1.0                   1                3   0.3333333</span></span>
<span><span class="co">#&gt; 6         1.0                   1                1   1.0000000</span></span>
<span><span class="co">#&gt;   S_1_2_N.occurrences S_1_2_N.coverage S_1_2_N.VAF S_1_2_P.occurrences</span></span>
<span><span class="co">#&gt; 1                   3                5         0.6                   1</span></span>
<span><span class="co">#&gt; 2                   3                3         1.0                   4</span></span>
<span><span class="co">#&gt; 3                   5                5         1.0                   6</span></span>
<span><span class="co">#&gt; 4                   2                2         1.0                   3</span></span>
<span><span class="co">#&gt; 5                   2                4         0.5                   0</span></span>
<span><span class="co">#&gt; 6                   0                0         NaN                   2</span></span>
<span><span class="co">#&gt;   S_1_2_P.coverage S_1_2_P.VAF S_2_1_N.occurrences S_2_1_N.coverage S_2_1_N.VAF</span></span>
<span><span class="co">#&gt; 1                1           1                   2                3   0.6666667</span></span>
<span><span class="co">#&gt; 2                4           1                   2                2   1.0000000</span></span>
<span><span class="co">#&gt; 3                6           1                   3                3   1.0000000</span></span>
<span><span class="co">#&gt; 4                3           1                   2                2   1.0000000</span></span>
<span><span class="co">#&gt; 5                2           0                   0                1   0.0000000</span></span>
<span><span class="co">#&gt; 6                2           1                   5                5   1.0000000</span></span>
<span><span class="co">#&gt;   S_2_1_P.occurrences S_2_1_P.coverage S_2_1_P.VAF S_2_2_N.occurrences</span></span>
<span><span class="co">#&gt; 1                   1                3   0.3333333                   2</span></span>
<span><span class="co">#&gt; 2                   3                3   1.0000000                   0</span></span>
<span><span class="co">#&gt; 3                   1                1   1.0000000                   1</span></span>
<span><span class="co">#&gt; 4                   2                2   1.0000000                   1</span></span>
<span><span class="co">#&gt; 5                   0                2   0.0000000                   0</span></span>
<span><span class="co">#&gt; 6                   8                8   1.0000000                   1</span></span>
<span><span class="co">#&gt;   S_2_2_N.coverage S_2_2_N.VAF S_2_2_P.occurrences S_2_2_P.coverage S_2_2_P.VAF</span></span>
<span><span class="co">#&gt; 1                2           1                   1                2   0.5000000</span></span>
<span><span class="co">#&gt; 2                1           0                   2                3   0.6666667</span></span>
<span><span class="co">#&gt; 3                1           1                   4                4   1.0000000</span></span>
<span><span class="co">#&gt; 4                1           1                   1                1   1.0000000</span></span>
<span><span class="co">#&gt; 5                0         NaN                   0                0         NaN</span></span>
<span><span class="co">#&gt; 6                1           1                   3                3   1.0000000</span></span>
<span><span class="co">#&gt;   normal_sample.occurrences normal_sample.coverage normal_sample.VAF</span></span>
<span><span class="co">#&gt; 1                         1                      1         1.0000000</span></span>
<span><span class="co">#&gt; 2                         0                      0               NaN</span></span>
<span><span class="co">#&gt; 3                         1                      1         1.0000000</span></span>
<span><span class="co">#&gt; 4                         1                      1         1.0000000</span></span>
<span><span class="co">#&gt; 5                         2                      3         0.6666667</span></span>
<span><span class="co">#&gt; 6                         3                      3         1.0000000</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="saving-the-simulated-reads">Saving the Simulated Reads<a class="anchor" aria-label="anchor" href="#saving-the-simulated-reads"></a>
</h3>
<p>rRACES/RACE can also save the simulated reads in the SAM format (see
<span class="citation">(Li et al. 2009)</span>). By setting the optional
parameter <code>write_SAM</code> to <code>TRUE</code>, the function
<code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> creates the directory
<code>rRACES_SAM</code> and saves the SAM files in it. Each file is
named after one of the reference genome chromosomes and contains the
simulated reads. The reads are split into read groups corresponding to
the collected samples.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>, write_SAM <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [█---------------------------------------] 0% [00m:02s] Processing chr. 22 [██--------------------------------------] 4% [00m:03s] Processing chr. 22 [████------------------------------------] 8% [00m:04s] Processing chr. 22 [█████-----------------------------------] 12% [00m:05s] Processing chr. 22 [███████---------------------------------] 16% [00m:06s] Processing chr. 22 [█████████-------------------------------] 21% [00m:07s] Processing chr. 22 [███████████-----------------------------] 25% [00m:08s] Processing chr. 22 [████████████----------------------------] 29% [00m:09s] Processing chr. 22 [██████████████--------------------------] 33% [00m:10s] Processing chr. 22 [███████████████-------------------------] 37% [00m:11s] Processing chr. 22 [█████████████████-----------------------] 42% [00m:12s] Processing chr. 22 [███████████████████---------------------] 46% [00m:13s] Processing chr. 22 [█████████████████████-------------------] 50% [00m:14s] Processing chr. 22 [██████████████████████------------------] 54% [00m:15s] Processing chr. 22 [███████████████████████-----------------] 57% [00m:16s] Processing chr. 22 [█████████████████████████---------------] 61% [00m:17s] Processing chr. 22 [██████████████████████████--------------] 64% [00m:18s] Processing chr. 22 [████████████████████████████------------] 68% [00m:19s] Processing chr. 22 [█████████████████████████████-----------] 71% [00m:20s] Processing chr. 22 [███████████████████████████████---------] 75% [00m:21s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:22s] Processing chr. 22 [█████████████████████████████████-------] 82% [00m:23s] Processing chr. 22 [███████████████████████████████████-----] 86% [00m:24s] Processing chr. 22 [████████████████████████████████████----] 89% [00m:25s] Processing chr. 22 [██████████████████████████████████████--] 93% [00m:26s] Processing chr. 22 [███████████████████████████████████████-] 96% [00m:27s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:28s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:29s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:31s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:32s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:33s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:34s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:35s] Read simulated</span></span>
<span></span>
<span><span class="va">sam_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"rRACES_SAM/"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ex_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rRACES_SAM/"</span>, <span class="va">sam_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">line</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">ex_file</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">line</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; @HD  VN:1.6  SO:unknown </span></span>
<span><span class="co">#&gt; @SQ  SN:22   LN:51304566 AN:chromosome22,chr22,chromosome_22,chr_22 </span></span>
<span><span class="co">#&gt; @RG  ID:S_1_1    SM:S_1_1    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:S_1_2    SM:S_1_2    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:S_2_1    SM:S_2_1    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:S_2_2    SM:S_2_2    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:normal_sample    SM:normal_sample    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; r00000000000 0   22  37812620    33  65M1X84M    *   0   0   GAAGTGGGAGGGAAGGGCTGGGTGGAGGCACTGGGAGGAAGCAAGCGCTGGTGCAGCAGGGTCTCTGGTCTGTCCACAGGGGCAGGCAGGATAAGTGCAGACCCGCCCACTGTGCCTCACTCCCCAGCCCTCCACCACCAAGAACCCCCA  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  RG:Z:S_1_1  NM:i:1 </span></span>
<span><span class="co">#&gt; r00000000001 16  22  48495855    33  77M1X72M    *   0   0   GATCTGGGATCCTCCAGGCCTGGGGTAATGAAAGCCCTTCATTGATGGCCAAACCTTCCCTCCACCAGGGCAATATGTCCCACACATCACTGTCCCTTCCTTGTCACCTTCCTCCCTGGAGCTGACCCAGTTATGGGCTTGAAAATATTA  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  RG:Z:S_1_1  NM:i:1 </span></span>
<span><span class="co">#&gt; r00000000002 0   22  23063023    33  150M    *   0   0   TAACGACGACACAAGAATAAAGGAAGTAGATTTGCATGAAGAGACTTCCCTTCCTATGATAAGAGAGGCCTGGAGGTTCCTCCTTAGCTGTGGGCTCAGAAGCAGAGTTCTGGGGTGTCTCCACCATGGCCTGGACCCCTCTCTGGCTCA  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  RG:Z:S_1_1  NM:i:0</span></span></code></pre></div>
<p>Each SAM file contains the reads produced by simulating the
sequencing of <em>all</em> the samples. The command-line tools <a href="http://www.htslib.org" class="external-link"><code>samtools</code></a> can be used to
split the reads by sample.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="ex">foo@bar</span> % samtools split  <span class="at">-f</span> <span class="st">"%*_%\!.sam"</span> rRACES_SAM/chr_22.sam</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="ex">foo@bar</span> % ls chr_22_<span class="pp">*</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="ex">chr_22_S_1_1.sam</span>        chr_22_S_1_2.sam        chr_22_S_2_1.sam        chr_22_S_2_2.sam</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a><span class="ex">chr_22_normal_sample.sam</span></span></code></pre></div>
<p>The resulting files are named after the samples, each containing the
reads of only one sample.</p>
<p>Please refer to the <a href="http://www.htslib.org/doc/samtools-split.html" class="external-link"><code>samtools split</code>
manual</a> for more details.</p>
<p>The <code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> parameter <code>output_dir</code> set
the SAM output directory. See the following section for usage examples
of this parameter.</p>
</div>
<div class="section level3">
<h3 id="sample-purity">Sample Purity<a class="anchor" aria-label="anchor" href="#sample-purity"></a>
</h3>
<p>The <em>sample purity</em> represents the concentration of tumour
cells in a sample. More formally, it is the ratio between the number of
tumour cells in the sample and that of all the sample cells, either
tumour or normal.</p>
<p>rRACES can simulate different sample purity in sequencing
simulations.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># simulate the sequencing of a sample in which 70% of the cells are</span></span>
<span><span class="co"># tumor cells</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>, purity <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                            write_SAM <span class="op">=</span> <span class="cn">TRUE</span>, output_dir <span class="op">=</span> <span class="st">"SAM_0.7"</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [█---------------------------------------] 0% [00m:02s] Processing chr. 22 [███-------------------------------------] 5% [00m:03s] Processing chr. 22 [████------------------------------------] 9% [00m:04s] Processing chr. 22 [██████----------------------------------] 13% [00m:05s] Processing chr. 22 [███████---------------------------------] 17% [00m:06s] Processing chr. 22 [█████████-------------------------------] 21% [00m:07s] Processing chr. 22 [███████████-----------------------------] 26% [00m:08s] Processing chr. 22 [█████████████---------------------------] 30% [00m:09s] Processing chr. 22 [██████████████--------------------------] 34% [00m:10s] Processing chr. 22 [████████████████------------------------] 38% [00m:11s] Processing chr. 22 [█████████████████-----------------------] 42% [00m:12s] Processing chr. 22 [███████████████████---------------------] 46% [00m:13s] Processing chr. 22 [█████████████████████-------------------] 50% [00m:14s] Processing chr. 22 [██████████████████████------------------] 54% [00m:15s] Processing chr. 22 [███████████████████████-----------------] 57% [00m:16s] Processing chr. 22 [████████████████████████----------------] 58% [00m:17s] Processing chr. 22 [█████████████████████████---------------] 62% [00m:18s] Processing chr. 22 [███████████████████████████-------------] 66% [00m:19s] Processing chr. 22 [████████████████████████████------------] 69% [00m:20s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:21s] Processing chr. 22 [███████████████████████████████---------] 77% [00m:22s] Processing chr. 22 [█████████████████████████████████-------] 80% [00m:23s] Processing chr. 22 [██████████████████████████████████------] 84% [00m:24s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:25s] Processing chr. 22 [█████████████████████████████████████---] 91% [00m:26s] Processing chr. 22 [███████████████████████████████████████-] 95% [00m:27s] Processing chr. 22 [████████████████████████████████████████] 98% [00m:28s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:29s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:31s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:32s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:33s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:34s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:36s] Read simulated</span></span>
<span></span>
<span><span class="va">sam_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"SAM_0.7/"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ex_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"SAM_0.7/"</span>, <span class="va">sam_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">line</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">ex_file</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">line</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; @HD  VN:1.6  SO:unknown </span></span>
<span><span class="co">#&gt; @SQ  SN:22   LN:51304566 AN:chromosome22,chr22,chromosome_22,chr_22 </span></span>
<span><span class="co">#&gt; @RG  ID:S_1_1    SM:S_1_1    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:S_1_2    SM:S_1_2    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:S_2_1    SM:S_2_1    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:S_2_2    SM:S_2_2    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; @RG  ID:normal_sample    SM:normal_sample    PL:ILLUMINA </span></span>
<span><span class="co">#&gt; r00000000000 0   22  37812620    33  65M1X84M    *   0   0   GAAGTGGGAGGGAAGGGCTGGGTGGAGGCACTGGGAGGAAGCAAGCGCTGGTGCAGCAGGGTCTCTGGTCTGTCCACAGGGGCAGGCAGGATAAGTGCAGACCCGCCCACTGTGCCTCACTCCCCAGCCCTCCACCACCAAGAACCCCCA  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  RG:Z:S_1_1  NM:i:1 </span></span>
<span><span class="co">#&gt; r00000000001 16  22  48495855    33  77M1X72M    *   0   0   GATCTGGGATCCTCCAGGCCTGGGGTAATGAAAGCCCTTCATTGATGGCCAAACCTTCCCTCCACCAGGGCAATATGTCCCACACATCACTGTCCCTTCCTTGTCACCTTCCTCCCTGGAGCTGACCCAGTTATGGGCTTGAAAATATTA  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  RG:Z:S_1_1  NM:i:1 </span></span>
<span><span class="co">#&gt; r00000000002 0   22  23063023    33  150M    *   0   0   TAACGACGACACAAGAATAAAGGAAGTAGATTTGCATGAAGAGACTTCCCTTCCTATGATAAGAGAGGCCTGGAGGTTCCTCCTTAGCTGTGGGCTCAGAAGCAGAGTTCTGGGGTGTCTCCACCATGGCCTGGACCCCTCTCTGGCTCA  !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!  RG:Z:S_1_1  NM:i:0</span></span></code></pre></div>
<p>The default value of the parameter <code>purity</code> is 1, i.e.,
<code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> assumes all the cells in the sample to be
neoplastic by default.</p>
</div>
<div class="section level3">
<h3 id="normal-cell-sequencing">Normal Cell Sequencing<a class="anchor" aria-label="anchor" href="#normal-cell-sequencing"></a>
</h3>
<p>The function <code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> adds a normal cell sample,
named <code>normal_sample</code> to the samples collected during the
simulation. In this way, rRACES automatically gathers information about
wild-type cells, and it also simulates the sequencing of normal cells
when the SAM writing is enabled.</p>
<p>This default behaviour can be turned off by setting the
<code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> parameter <code>with_normal_sample</code> to
<code>FALSE</code>.</p>
<p>The function <code><a href="../reference/simulate_normal_seq.html">simulate_normal_seq()</a></code> simulates the normal
cell sample sequencing <em>avoiding</em> tumor sequencing.</p>
</div>
<div class="section level3">
<h3 id="simulating-sequencing-errors">Simulating Sequencing Errors<a class="anchor" aria-label="anchor" href="#simulating-sequencing-errors"></a>
</h3>
<p>To simulate sequencing errors, users need to specify a sequencer to
the function <code><a href="../reference/simulate_seq.html">simulate_seq()</a></code>.</p>
<p>At the moment, there are only two classes implementing sequencers:
the <code>ErrorlessIlluminaSequencer</code> class and the
<code>BasicIlluminaSequencer</code> classes.</p>
<div class="section level4">
<h4 id="the-errorlessilluminasequencer-class">The <code>ErrorlessIlluminaSequencer</code> Class<a class="anchor" aria-label="anchor" href="#the-errorlessilluminasequencer-class"></a>
</h4>
<p>This class models a perfect Illumina sequencer. No sequencing errors
are produced, and all the bases have the maximum quality.</p>
<p>The following code simulates the sequencing by using this kind of
sequencer.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build an error-less Illumina sequencer</span></span>
<span><span class="va">no_error_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">ErrorlessIlluminaSequencer</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># let us simulate a 2.5x sequencing of the four sample</span></span>
<span><span class="co"># on the error-less sequencer</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, sequencer <span class="op">=</span> <span class="va">no_error_seq</span>,</span>
<span>                            coverage <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="the-basicilluminasequencer-class">The <code>BasicIlluminaSequencer</code> Class<a class="anchor" aria-label="anchor" href="#the-basicilluminasequencer-class"></a>
</h4>
<p>This class sets a sequencing error rate independent of the base in
which the error occurs, from the position of the base in the read and
from the position of the read on the genome.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build a basic Illumina sequencer model in which errors occur</span></span>
<span><span class="co"># at rate 4e-3 per base</span></span>
<span><span class="va">basic_seq</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">BasicIlluminaSequencer</span>, <span class="fl">4e-3</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># let us simulate a 2.5x sequencing of the four sample</span></span>
<span><span class="co"># on the error-less sequencer</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, sequencer <span class="op">=</span> <span class="va">basic_seq</span>,</span>
<span>                            coverage <span class="op">=</span> <span class="fl">2.5</span>, write_SAM <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="chromosome-sequencing">Chromosome Sequencing<a class="anchor" aria-label="anchor" href="#chromosome-sequencing"></a>
</h3>
<p>The function <code><a href="../reference/simulate_seq.html">simulate_seq()</a></code> allows users to simulate the
sequencing of a selection of the reference chromosomes by using the
parameter <code>chromosomes</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># let us simulate a 2.5x sequencing of the chromosomes 22 and</span></span>
<span><span class="co"># X of the four sample</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, chromosomes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"22"</span>, <span class="st">"X"</span><span class="op">)</span>,</span>
<span>                            coverage <span class="op">=</span> <span class="fl">2.5</span>, write_SAM <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [██--------------------------------------] 4% [00m:02s] Processing chr. 22 [████------------------------------------] 8% [00m:03s] Processing chr. 22 [█████-----------------------------------] 12% [00m:04s] Processing chr. 22 [███████---------------------------------] 16% [00m:05s] Processing chr. 22 [█████████-------------------------------] 21% [00m:06s] Processing chr. 22 [███████████-----------------------------] 25% [00m:07s] Processing chr. 22 [████████████----------------------------] 28% [00m:08s] Processing chr. 22 [█████████████---------------------------] 32% [00m:09s] Processing chr. 22 [███████████████-------------------------] 35% [00m:10s] Processing chr. 22 [████████████████------------------------] 39% [00m:11s] Processing chr. 22 [██████████████████----------------------] 43% [00m:12s] Processing chr. 22 [███████████████████---------------------] 47% [00m:13s] Processing chr. 22 [█████████████████████-------------------] 51% [00m:14s] Processing chr. 22 [███████████████████████-----------------] 55% [00m:15s] Processing chr. 22 [████████████████████████----------------] 58% [00m:16s] Processing chr. 22 [█████████████████████████---------------] 62% [00m:17s] Processing chr. 22 [███████████████████████████-------------] 66% [00m:18s] Processing chr. 22 [█████████████████████████████-----------] 70% [00m:19s] Processing chr. 22 [██████████████████████████████----------] 73% [00m:20s] Processing chr. 22 [███████████████████████████████---------] 77% [00m:21s] Processing chr. 22 [█████████████████████████████████-------] 81% [00m:22s] Processing chr. 22 [██████████████████████████████████------] 84% [00m:23s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:24s] Processing chr. 22 [█████████████████████████████████████---] 92% [00m:25s] Processing chr. 22 [███████████████████████████████████████-] 95% [00m:26s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:27s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:28s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:29s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:31s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:32s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:33s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:36s] Read simulated</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="updating-the-sam-output-directory">Updating the SAM Output Directory<a class="anchor" aria-label="anchor" href="#updating-the-sam-output-directory"></a>
</h3>
<p>Users may want to simulate sequencing in multiple steps, for
instance, by splitting it by chromosome or reaching the aimed coverage
with different simulations. However, rRACES prevents successive writing
in the same directory by default. The Boolean parameter
<code>update_SAM</code> allows multiple writing in the same
directory.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># the default SAM directory already exists</span></span>
<span><span class="va">sam_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"rRACES_SAM/"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># since the default save directory already exists, any call to</span></span>
<span><span class="co"># `simulate_seq` throws an error</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span></span>
<span>  <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>, write_SAM <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span>  <span class="op">}</span>,</span>
<span>  error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">e</span><span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; &lt;std::domain_error in eval(expr, envir, enclos): "rRACES_SAM" already exists&gt;</span></span>
<span></span>
<span><span class="co"># setting `update_SAM` to TRUE enables successive writing in</span></span>
<span><span class="co"># the output directory</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>,</span>
<span>                            write_SAM <span class="op">=</span> <span class="cn">TRUE</span>, update_SAM <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [█---------------------------------------] 0% [00m:00s] Found 22 [█---------------------------------------] 0% [00m:00s] Reading 22 [█---------------------------------------] 0% [00m:01s] Reading 22 [█---------------------------------------] 0% [00m:01s] Processing chr. 22 [█---------------------------------------] 1% [00m:02s] Processing chr. 22 [███-------------------------------------] 5% [00m:03s] Processing chr. 22 [████------------------------------------] 8% [00m:04s] Processing chr. 22 [█████-----------------------------------] 12% [00m:05s] Processing chr. 22 [███████---------------------------------] 16% [00m:06s] Processing chr. 22 [█████████-------------------------------] 20% [00m:07s] Processing chr. 22 [██████████------------------------------] 23% [00m:08s] Processing chr. 22 [███████████-----------------------------] 27% [00m:09s] Processing chr. 22 [█████████████---------------------------] 31% [00m:10s] Processing chr. 22 [███████████████-------------------------] 35% [00m:11s] Processing chr. 22 [████████████████------------------------] 39% [00m:12s] Processing chr. 22 [██████████████████----------------------] 44% [00m:13s] Processing chr. 22 [████████████████████--------------------] 48% [00m:14s] Processing chr. 22 [█████████████████████-------------------] 52% [00m:15s] Processing chr. 22 [███████████████████████-----------------] 56% [00m:16s] Processing chr. 22 [████████████████████████----------------] 58% [00m:17s] Processing chr. 22 [██████████████████████████--------------] 63% [00m:18s] Processing chr. 22 [███████████████████████████-------------] 67% [00m:19s] Processing chr. 22 [█████████████████████████████-----------] 70% [00m:20s] Processing chr. 22 [██████████████████████████████----------] 74% [00m:21s] Processing chr. 22 [████████████████████████████████--------] 78% [00m:22s] Processing chr. 22 [█████████████████████████████████-------] 81% [00m:23s] Processing chr. 22 [███████████████████████████████████-----] 85% [00m:24s] Processing chr. 22 [████████████████████████████████████----] 88% [00m:25s] Processing chr. 22 [█████████████████████████████████████---] 92% [00m:26s] Processing chr. 22 [███████████████████████████████████████-] 96% [00m:27s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:28s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:29s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:30s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:31s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:32s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:33s] Processing chr. 22 [████████████████████████████████████████] 99% [00m:34s] Processing chr. 22 [████████████████████████████████████████] 100% [00m:35s] Read simulated</span></span></code></pre></div>
<p>When the output directory already contains a SAM file associated with
a chromosome, the alignments on that chromosome are saved in a file
whose name is the first available, having the format
<code>"chr_{chromosome name}_{natural number}.sam"</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"rRACES_SAM/"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "chr_22_0.sam" "chr_22.sam"</span></span></code></pre></div>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-10.1093/bioinformatics/btp352" class="csl-entry">
Li, Heng, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils
Homer, Gabor Marth, Goncalo Abecasis, Richard Durbin, and 1000 Genome
Project Data Processing Subgroup. 2009. <span>“<span class="nocase">The
Sequence Alignment/Map format and SAMtools</span>.”</span>
<em>Bioinformatics</em> 25 (16): 2078–79. <a href="https://doi.org/10.1093/bioinformatics/btp352" class="external-link">https://doi.org/10.1093/bioinformatics/btp352</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
