<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rRACES">
<title>Sequencing Simulation • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Sequencing Simulation">
<meta property="og:description" content="rRACES">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.5</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/growth_models.html">Growth Models</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Sequencing Simulation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/sequencing.Rmd" class="external-link"><code>vignettes/sequencing.Rmd</code></a></small>
      <div class="d-none name"><code>sequencing.Rmd</code></div>
    </div>

    
    
<p>Once a phylogenetic forest has been computed (see
<code><a href="../articles/mutations.html">vignette("mutations")</a></code>) for the details), rRACES/RACES can
simulate the sequencing of the samples in the forest and return the
observed data.</p>
<div class="section level3">
<h3 id="sequencing-simulation">Sequencing Simulation<a class="anchor" aria-label="anchor" href="#sequencing-simulation"></a>
</h3>
<p>Let us consider the phylogenetic forest produced in
<code><a href="../articles/mutations.html">vignette("mutations")</a></code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rRACES</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylo_forest</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/load_phylogenetic_forest.html">load_phylogenetic_forest</a></span><span class="op">(</span><span class="st">"phylo_forest.sff"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylo_forest</span></span>
<span><span class="co">#&gt; PhylogeneticForest</span></span>
<span><span class="co">#&gt;   # of trees: 1</span></span>
<span><span class="co">#&gt;   # of nodes: 4773</span></span>
<span><span class="co">#&gt;   # of leaves: 953</span></span>
<span><span class="co">#&gt;   samples: {"S_1_1", "S_1_2", "S_2_1", "S_2_2"}</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;   # of emerged SNVs: 70051</span></span>
<span><span class="co">#&gt;   # of emerged CNAs: 2</span></span></code></pre></div>
<p>The loaded phylogenetic forest models the cell evolution of 4
different samples: <code>S_1_1</code>, <code>S_1_2</code>,
<code>S_2_1</code>, and <code>S_2_2</code>.</p>
<p>We can simulate the sequencing of these samples with coverage of 2.5X
by as it follows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># let us simulate a 2.5x sequencing of the four sample</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># let us load the dplyr library to filter the `simulate_seq` output</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'dplyr'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     filter, lag</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     intersect, setdiff, setequal, union</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">#&gt;   chromosome  chr_pos ref alt causes  classes S_1_1.occurrences S_1_1.coverage</span></span>
<span><span class="co">#&gt; 1         22 10510210   N   C      A   driver                 1              4</span></span>
<span><span class="co">#&gt; 2         22 16051493   G   A   &lt;NA&gt; germinal                 1              1</span></span>
<span><span class="co">#&gt; 3         22 16052080   G   A   &lt;NA&gt; germinal                 1              1</span></span>
<span><span class="co">#&gt; 4         22 16053659   A   C   &lt;NA&gt; germinal                 3              3</span></span>
<span><span class="co">#&gt; 5         22 16053863   G   A   &lt;NA&gt; germinal                 1              1</span></span>
<span><span class="co">#&gt; 6         22 16054740   A   G   &lt;NA&gt; germinal                 3              3</span></span>
<span><span class="co">#&gt;   S_1_1.VAF S_1_2.occurrences S_1_2.coverage S_1_2.VAF S_2_1.occurrences</span></span>
<span><span class="co">#&gt; 1      0.25                 2              2 1.0000000                 1</span></span>
<span><span class="co">#&gt; 2      1.00                 2              6 0.3333333                 3</span></span>
<span><span class="co">#&gt; 3      1.00                 3              3 1.0000000                 3</span></span>
<span><span class="co">#&gt; 4      1.00                 2              2 1.0000000                 4</span></span>
<span><span class="co">#&gt; 5      1.00                 0              2 0.0000000                 2</span></span>
<span><span class="co">#&gt; 6      1.00                 1              1 1.0000000                 3</span></span>
<span><span class="co">#&gt;   S_2_1.coverage S_2_1.VAF S_2_2.occurrences S_2_2.coverage S_2_2.VAF</span></span>
<span><span class="co">#&gt; 1              3 0.3333333                 1              2       0.5</span></span>
<span><span class="co">#&gt; 2              6 0.5000000                 2              4       0.5</span></span>
<span><span class="co">#&gt; 3              4 0.7500000                 3              5       0.6</span></span>
<span><span class="co">#&gt; 4              4 1.0000000                 3              3       1.0</span></span>
<span><span class="co">#&gt; 5              5 0.4000000                 0              1       0.0</span></span>
<span><span class="co">#&gt; 6              3 1.0000000                 6              6       1.0</span></span></code></pre></div>
<p>The function <code>simulate_seq</code> returns a data frame whose
rows represent the mutations observed in the simulated reads (currently,
only SNVs are supported).</p>
<p>The number of columns in the data frame depends on the number of
samples in the phygenetic forest. The first 4 columns describe the
mutation and report the chromosome and the position in the chromosome of
the mutation (columns “<code>chromosome</code>” and
“<code>chr_pos</code>”, respectively), the genetic context in which the
mutation occurs (column “<code>context</code>”), and the new base
(column “<code>alt_base</code>”). Then, there are 3 columns for each of
the samples: they contain the number of simulated reads affected by the
mutation, the sequencing coverage of the mutation locus, and the ratio
between these two last columns (columns
“<code>&lt;sample name&gt; occurrences</code>”,
“<code>&lt;sample name&gt; coverage</code>”, and
“<code>&lt;sample name&gt; VAF</code>”, respectively).</p>
</div>
<div class="section level3">
<h3 id="sample-purity-and-normal-cell-sequencing">Sample Purity and Normal Cell Sequencing<a class="anchor" aria-label="anchor" href="#sample-purity-and-normal-cell-sequencing"></a>
</h3>
<p>The <em>sample purity</em> represents the concentration of tumor
cells in a sample. More formally, it is the ratio between the number of
tumor cells in the sample and that of all the sample cells, being either
tumor or normal cells.</p>
<p>rRACES can simulate different sample purity in sequencing
simulations.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># simulate the sequencing of a sample in which 70% of the cells are</span></span>
<span><span class="co"># tumor cells</span></span>
<span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>, purity <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span></span></code></pre></div>
<p>The default value of the parameter <code>purity</code> is 1, i.e.,
<code>simulate_seq</code> assumes all the cells in the sample to be
neoplastic by default.</p>
<p>The sequencing of a sample exclusively containing wild-type cells can
be simulated by setting the parameter <code>purity</code> to 0.</p>
</div>
<div class="section level3">
<h3 id="cell-partition-by-epigenetic-state-epi-facs">Cell Partition by Epigenetic State (Epi-FACS)<a class="anchor" aria-label="anchor" href="#cell-partition-by-epigenetic-state-epi-facs"></a>
</h3>
<p>The sequencing simulation supports cell partition by epigenetic state
(epi-FACS) by using the <code>simualate_seq</code> parameter
<code>epi_FACS</code>. At the beginning of the sequencing simulation
process, each sample in the phylogenetic forest is splitted in two
distinct samples: <code>&lt;sample name&gt;_P</code>, that contains all
the cells in the original sample having epigenetic state <code>+</code>,
and <code>&lt;sample name&gt;_N</code>, that consists in the cells
having epigenetic state <code>-</code>. The results of the simulation
reports the data for the splitted samples.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>, epi_FACS <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seq_results</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="va">head</span></span>
<span><span class="co">#&gt;   chromosome  chr_pos ref alt causes  classes S_1_1_N.occurrences</span></span>
<span><span class="co">#&gt; 1         22 10510210   N   C      A   driver                   2</span></span>
<span><span class="co">#&gt; 2         22 16051493   G   A   &lt;NA&gt; germinal                   1</span></span>
<span><span class="co">#&gt; 3         22 16052080   G   A   &lt;NA&gt; germinal                   0</span></span>
<span><span class="co">#&gt; 4         22 16053659   A   C   &lt;NA&gt; germinal                   3</span></span>
<span><span class="co">#&gt; 5         22 16053863   G   A   &lt;NA&gt; germinal                   0</span></span>
<span><span class="co">#&gt; 6         22 16054740   A   G   &lt;NA&gt; germinal                   4</span></span>
<span><span class="co">#&gt;   S_1_1_N.coverage S_1_1_N.VAF S_1_1_P.occurrences S_1_1_P.coverage S_1_1_P.VAF</span></span>
<span><span class="co">#&gt; 1                2           1                   1                1         1.0</span></span>
<span><span class="co">#&gt; 2                1           1                   3                6         0.5</span></span>
<span><span class="co">#&gt; 3                1           0                   1                1         1.0</span></span>
<span><span class="co">#&gt; 4                3           1                   2                2         1.0</span></span>
<span><span class="co">#&gt; 5                0         NaN                   0                2         0.0</span></span>
<span><span class="co">#&gt; 6                4           1                   2                2         1.0</span></span>
<span><span class="co">#&gt;   S_1_2_N.occurrences S_1_2_N.coverage S_1_2_N.VAF S_1_2_P.occurrences</span></span>
<span><span class="co">#&gt; 1                   1                3   0.3333333                   0</span></span>
<span><span class="co">#&gt; 2                   1                2   0.5000000                   1</span></span>
<span><span class="co">#&gt; 3                   2                2   1.0000000                   3</span></span>
<span><span class="co">#&gt; 4                   3                3   1.0000000                   2</span></span>
<span><span class="co">#&gt; 5                   3                6   0.5000000                   0</span></span>
<span><span class="co">#&gt; 6                   1                1   1.0000000                   4</span></span>
<span><span class="co">#&gt;   S_1_2_P.coverage S_1_2_P.VAF S_2_1_N.occurrences S_2_1_N.coverage S_2_1_N.VAF</span></span>
<span><span class="co">#&gt; 1                0         NaN                   1                1         1.0</span></span>
<span><span class="co">#&gt; 2                3   0.3333333                   2                5         0.4</span></span>
<span><span class="co">#&gt; 3                5   0.6000000                   2                2         1.0</span></span>
<span><span class="co">#&gt; 4                2   1.0000000                   2                2         1.0</span></span>
<span><span class="co">#&gt; 5                3   0.0000000                   0                2         0.0</span></span>
<span><span class="co">#&gt; 6                4   1.0000000                   8                8         1.0</span></span>
<span><span class="co">#&gt;   S_2_1_P.occurrences S_2_1_P.coverage S_2_1_P.VAF S_2_2_N.occurrences</span></span>
<span><span class="co">#&gt; 1                   2                4         0.5                   0</span></span>
<span><span class="co">#&gt; 2                   0                1         0.0                   2</span></span>
<span><span class="co">#&gt; 3                   0                0         NaN                   0</span></span>
<span><span class="co">#&gt; 4                   1                1         1.0                   2</span></span>
<span><span class="co">#&gt; 5                   0                0         NaN                   3</span></span>
<span><span class="co">#&gt; 6                   3                3         1.0                   2</span></span>
<span><span class="co">#&gt;   S_2_2_N.coverage S_2_2_N.VAF S_2_2_P.occurrences S_2_2_P.coverage S_2_2_P.VAF</span></span>
<span><span class="co">#&gt; 1                2           0                   1                1         1.0</span></span>
<span><span class="co">#&gt; 2                2           1                   0                2         0.0</span></span>
<span><span class="co">#&gt; 3                0         NaN                   1                1         1.0</span></span>
<span><span class="co">#&gt; 4                2           1                   3                3         1.0</span></span>
<span><span class="co">#&gt; 5                3           1                   1                2         0.5</span></span>
<span><span class="co">#&gt; 6                2           1                   3                3         1.0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="saving-the-simulated-reads">Saving the Simulated Reads<a class="anchor" aria-label="anchor" href="#saving-the-simulated-reads"></a>
</h3>
<p>rRACES/RACE can also save the simulated reads in the SAM format (see
<span class="citation">(Li et al. 2009)</span>). By setting the optional
parameter <code>write_SAM</code> to <code>TRUE</code>, the function
<code>simulate_seq</code> creates the directory <code>rRACES_SAM</code>
and saves the SAM files in it. Each file is named after one the
reference genome chromosomes and it contains the reads simulated for
it.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seq_results</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/simulate_seq.html">simulate_seq</a></span><span class="op">(</span><span class="va">phylo_forest</span>, coverage <span class="op">=</span> <span class="fl">2.5</span>, write_SAM <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">SAM_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"rRACES_SAM/"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ex_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rRACES_SAM/"</span>, <span class="va">SAM_files</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, sep<span class="op">=</span><span class="st">""</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">line</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">ex_file</span>, n <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">line</span>,<span class="st">"\n"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt; @HD  VN:1.6  SO:unknown </span></span>
<span><span class="co">#&gt; @SQ  SN:22   LN:51304566 AN:chromosome22,chr22,chromosome_22,chr_22 </span></span>
<span><span class="co">#&gt; @RG  ID:S_1_1 </span></span>
<span><span class="co">#&gt; @RG  ID:S_1_2 </span></span>
<span><span class="co">#&gt; @RG  ID:S_2_1 </span></span>
<span><span class="co">#&gt; @RG  ID:S_2_2 </span></span>
<span><span class="co">#&gt; r00000000000 0   22  37812620    33  150M    *   0   0   GAAGTGGGAGGGAAGGGCTGGGTGGAGGCACTGGGAGGAAGCAAGCGCTGGTGCAGCAGGGTCTCTGGTCTGTCCACAGGGGCAGGCAGGATAAGTGCAGACCCGCCCACTGTGCCTCACTCCCCAGCCCTCCACCACCAAGAACCCCCA  *   RG:Z:S_1_1 </span></span>
<span><span class="co">#&gt; r00000000001 16  22  50800045    33  150M    *   0   0   CTTGCCTTCAGCTATTTTTCCTTCAACAGTCTCAGCCCAGAGTATGCCCTGCTTGTCTTCCTTTGACCCAGGACCCAGAGTATGCCCTGCTCGTCTGCCTTTGACTCTTTGTTCTGGTACTTGTTTTCCCAGGATTATATTATTGACAAC  *   RG:Z:S_1_1 </span></span>
<span><span class="co">#&gt; r00000000002 16  22  25315049    33  150M    *   0   0   TCTGTCACCCAGGCTGGAGTGCAGTGGTGCAATCTCGGCTTACTGCAGCCTCCACTTCCTGGGTTCAAGCAATTCTCTGCCTCAGCCTCTTGAGTAGCTGGGATTACAGGCACCTGCCACCATGCCTGACTAATTTTTTTTGTATTTTTA  *   RG:Z:S_1_1 </span></span>
<span><span class="co">#&gt; r00000000003 16  22  17642772    33  150M    *   0   0   AAAAGAAAAAAAAAAAACCTGGAGTCTAATTCTAATTCCCCTCCCATTGAGTGTGGGCTGGACTTCACTGACTTTTTAATAAATAACATGGGACAGAAGTGAGTGTAGTTGATTTCCAAGGCTAGGTCATAAAAGATACTGTGACTTCCA  *   RG:Z:S_1_1</span></span></code></pre></div>
<p>Each SAM file contains the reads produced by simulating the
sequencing of <em>all</em> the samples. The command line tools
<code>samtools</code> can be used to split the reads by sample (see <a href="http://www.htslib.org" class="external-link">here</a>).</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="ex">foo@bar</span> % samtools split  <span class="at">-f</span> <span class="st">"%*_%\!.sam"</span> rRACES_SAM/chr_22.sam</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="ex">foo@bar</span> % ls chr_22_<span class="pp">*</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="ex">chr_22_S_1_1.sam</span>        chr_22_S_1_2.sam        chr_22_S_2_1.sam        chr_22_S_2_2.sam</span></code></pre></div>
<p>The resulting files are named after the phylogenetic forest samples
and each of them contains the reads of only one sample.</p>
<p>Please, refer to the <a href="http://www.htslib.org/doc/samtools-split.html" class="external-link"><code>samtools split</code>
manual</a> for more details.</p>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-10.1093/bioinformatics/btp352" class="csl-entry">
Li, Heng, Bob Handsaker, Alec Wysoker, Tim Fennell, Jue Ruan, Nils
Homer, Gabor Marth, Goncalo Abecasis, Richard Durbin, and 1000 Genome
Project Data Processing Subgroup. 2009. <span>“<span class="nocase">The
Sequence Alignment/Map format and SAMtools</span>.”</span>
<em>Bioinformatics</em> 25 (16): 2078–79. <a href="https://doi.org/10.1093/bioinformatics/btp352" class="external-link">https://doi.org/10.1093/bioinformatics/btp352</a>.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Giulio Caravagna, Alberto Casagrande.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
