<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="rRACES">
<title>Tissue simulation • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tissue simulation">
<meta property="og:description" content="rRACES">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.7.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/growth_models.html">Growth Models</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
    <a class="dropdown-item" href="../articles/SCOUT.html">SCOUT</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Tissue simulation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/tissue_simulation.Rmd" class="external-link"><code>vignettes/tissue_simulation.Rmd</code></a></small>
      <div class="d-none name"><code>tissue_simulation.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rRACES</span><span class="op">)</span></span></code></pre></div>
<p>To simulate a tissue the following steps are required:</p>
<ol style="list-style-type: decimal">
<li><p>creation of a tissue;</p></li>
<li><p>introduction of cells in the tissue;</p></li>
<li><p>actual simulation.</p></li>
</ol>
<p>The simulation is managed by an object of the S4 class
<code>Simulation</code>, wich allows to program the tissue evolution
over time, adding new cells as far as the simulation progresses. The
state of the simulation and tissue can be visualised using
<code>ggplot</code>-powered plots.</p>
<div class="section level2">
<h2 id="tissue-specification">Tissue specification<a class="anchor" aria-label="anchor" href="#tissue-specification"></a>
</h2>
<p>To perform a simulation a new object of class <code>Simulation</code>
must be created.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Default constructor</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">Simulation</span><span class="op">)</span></span></code></pre></div>
<p>This automatically builds a 1000x1000-cells tissue - which can host
<span class="math inline">\(1\)</span> million cells - and sets the name
of the simulation to be
<code>races_&lt;date&gt;_&lt;hour&gt;</code>.</p>
<p>A simulation custom name can be specified as it follows.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Call your simulation "Test"</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">Simulation</span>, <span class="st">"Test"</span><span class="op">)</span></span></code></pre></div>
<p>In order to save the simulation progresses in the disk and recover it
in the future, use the optional Boolean parameter
<code>save_snapshots</code>. By setting <code>save_snapshots</code> to
<code>TRUE</code>, the simulation progresses will be saved in a
directory whose name is the name of the simulation.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The progresses of the simulation will saved in the "Test" folder.</span></span>
<span><span class="co"># If the directory "Test" already exists an exception is raised.</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">Simulation</span>, <span class="st">"Test"</span>, save_snapshots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Class <code>Simulation</code> exports a
<code>Simulation$show()</code> method to get information on the current
object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span></span>
<span><span class="co">#&gt; ── <span style="background-color: #BBBB00;"> rRACES </span> <span style="color: #BBBBBB; background-color: #BB0000;"> D </span> <span style="color: #BBBBBB; background-color: #BB0000;"> S </span> <span style="color: #BBBBBB; background-color: #BB0000;"> M </span> Test ──────────────────── <span style="color: #BB0000;">▣ </span>A tissue [1000x1000]<span style="color: #BB0000;">  ⏱ </span>0 ──</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> The simulation has no samples yet!</span></span></code></pre></div>
<p>The <code>sim</code> object exposes also methods to get information
about the simulation and control it.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the simulation directory, i.e., "Test"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_name</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Test"</span></span>
<span></span>
<span><span class="co"># Get the tissue size, i.e., c(1000,1000)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_size</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1000 1000</span></span></code></pre></div>
<div class="section level3">
<h3 id="custom-tissue">Custom tissue<a class="anchor" aria-label="anchor" href="#custom-tissue"></a>
</h3>
<p>We can change the tissue size and label it with a name that might
appear in some plots.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the tissue name and size to different values</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">update_tissue</span><span class="op">(</span><span class="st">"Liver"</span>, <span class="fl">1200</span>, <span class="fl">900</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the tissue name, i.e., "Liver"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_name</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Liver"</span></span>
<span></span>
<span><span class="co"># Get the tissue size, i.e., c(1200,900)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_size</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1200  900</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="species-specification">Species specification<a class="anchor" aria-label="anchor" href="#species-specification"></a>
</h2>
<p>In order to simulate the evolution of some species we need to add
them to <code>sim</code>. This process defines the evolutionary
parameters of the species.</p>
<p>A <em>mutant</em> is a set of cells having the same (potentially
unknown) driver mutations. Cells in the same mutant can have different
liveness rates due to different epigenetic states.</p>
<p>A <em>species</em> is a mutant with an optional epigenetic state. At
this point in the simulation, the mutant is just a name (<code>A</code>,
<code>B</code>, ..) that, at a later stage could be linked to mutations
of interest. The epigenome is a binary feature of a species that is
represented by epistates <code>+</code>/<code>-</code> (positive and
negative status). This is an abstraction, and could represent an
active/inactive state linked to a promoter methylation or, more broadly,
a phenotype. The evolution of mutants is non-reversible (no-back
mutations model), while the evolution among epistates is potentially
reversible.</p>
<p>For example, if we define two mutants <code>A</code> and
<code>B</code> with their epistates <code>+</code>/<code>-</code>, we
have 4 distinct species: <code>A+</code> and <code>A-</code>, as well as
<code>B+</code> and <code>B-</code>.</p>
<p>Hybrid models can be obtained, e.g., a mutant <code>A</code> (with no
epistates), together with <code>B+</code> and <code>B-</code>.</p>
<div class="section level3">
<h3 id="evolutionary-parameters">Evolutionary parameters<a class="anchor" aria-label="anchor" href="#evolutionary-parameters"></a>
</h3>
<p>We use a notation common in linear birth-death processes. If a
species <code>A</code> has no epistates then its stochastic behaviour is
defined by the state-change rates</p>
<p><span class="math display">\[
\begin{align}
\text{(growth)}\quad A &amp; \rightarrow_{\lambda} 2 A \\
\text{(death)}\quad A &amp; \rightarrow_{\delta} \emptyset
\end{align}
\]</span></p>
<p>where:</p>
<ul>
<li>
<span class="math inline">\(\lambda&gt;0\)</span> is a growth rate
for cells that duplicate;</li>
<li>
<span class="math inline">\(\delta&gt;0\)</span> is a death rate for
cells that duplicate.</li>
</ul>
<p>Instead, if the species has epistate <code>+</code> (denoted <span class="math inline">\(A_\oplus\)</span>) and <code>-</code> (denoted
<span class="math inline">\(A_\ominus\)</span>), then its stochastic
behaviour is defined by the state-change rates</p>
<p><span class="math display">\[
\begin{align}
\text{(growth +)}\quad A_\oplus &amp; \rightarrow_{\lambda_\oplus} 2
A_\oplus \\
\text{(death +)}\quad A_\oplus &amp; \rightarrow_{\delta_\oplus}
\emptyset \\
\text{(growth -)}\quad A_\ominus &amp; \rightarrow_{\lambda_\ominus} 2
A_\ominus \\
\text{(death -)}\quad A_\ominus &amp; \rightarrow_{\delta_\ominus}
\emptyset \\
\text{(switch +-)}\quad A_\oplus &amp; \rightarrow_{\epsilon_{+-}}
A_\oplus + A_\ominus \\
\text{(switch -+)}\quad A_\ominus &amp; \rightarrow_{\epsilon_{-+}}
A_\ominus + A_\oplus
\end{align}
\]</span></p>
<p>where the rates <span class="math inline">\(\lambda\)</span> and
<span class="math inline">\(\delta\)</span> are as above, and <span class="math inline">\(\epsilon_{+-}\)</span> or <span class="math inline">\(\epsilon_{-+}\)</span> are the ratesat which cells
of a certain epistate duplicate and flip the epigenetic marker of one of
the progeny.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.01</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.08</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># updated object (counts refer to number of cells of each species)</span></span>
<span><span class="va">sim</span></span>
<span><span class="co">#&gt; ── <span style="background-color: #BBBB00;"> rRACES </span> <span style="color: #BBBBBB; background-color: #BB0000;"> D </span> <span style="color: #BBBBBB; background-color: #BB0000;"> S </span> <span style="color: #BBBBBB; background-color: #BB0000;"> M </span> Test ──────────────────────── <span style="color: #BB0000;">▣ </span>Liver [1200x900]<span style="color: #BB0000;">  ⏱ </span>0 ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Species: <span style="color: #00BB00;">2</span>, <span style="color: #00BB00;">with</span> epigenetics</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt;    species   λ      δ    ε    counts   % </span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt;         A-  0.08  0.01  0.01    0     NaN</span></span>
<span><span class="co">#&gt;         A+  0.20  0.10  0.01    0     NaN</span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Firings: 0 total</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> The simulation has no samples yet!</span></span>
<span></span>
<span><span class="co"># Get the simulation species and their rates</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_species</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate growth_rate death_rate switch_rate</span></span>
<span><span class="co">#&gt; 1      A        -        0.08       0.01        0.01</span></span>
<span><span class="co">#&gt; 2      A        +        0.20       0.10        0.01</span></span></code></pre></div>
<p>A mutant without epistates could be added as</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Not run</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"A"</span>, growth_rates <span class="op">=</span> <span class="fl">0.2</span>, death_rates <span class="op">=</span>  <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p>To be able to simulate the model, an initial cell needs to be
displaced in the tissue.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We add one cell of species A+ (mutant A in epistate +) in</span></span>
<span><span class="co"># position (500, 500).</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">place_cell</span><span class="op">(</span><span class="st">"A+"</span>, <span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualisations">Visualisations<a class="anchor" aria-label="anchor" href="#visualisations"></a>
</h3>
<p>We can query the current state of the simulation, and extract the
position of each cell in the tissue.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Counts per species</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts</span></span>
<span><span class="co">#&gt; 1      A        -      0</span></span>
<span><span class="co">#&gt; 2      A        +      1</span></span>
<span></span>
<span><span class="co"># Cells position (one so far)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1       0      A        +        500        500</span></span></code></pre></div>
<p>These information can be plot. Note that the tissue visualisation
uses hexagonal bins to avoid rastering delays when the simulation uses
thousands of cells.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Piechar for counts</span></span>
<span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-12-1.png" width="288"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Spatial distribution for the whole tissue (hexagonal bins)</span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-12-2.png" width="288"></p>
<blockquote>
<p><strong>Note:</strong> since the plots are done with
<code>ggplot</code> they can be assembled and customised.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="species-evolution">Species Evolution<a class="anchor" aria-label="anchor" href="#species-evolution"></a>
</h2>
<p>There are 4 ways to let the simulation evolve:</p>
<ul>
<li>advancing until the number of cells in a species reaches a given
threshold <span class="math inline">\(\theta &gt; 0\)</span>;</li>
<li>advancing until a new time <span class="math inline">\(t&gt;0\)</span> is reached;</li>
<li>advancing until a desired number of firings (of one particular
event) has occurred;</li>
<li>advancing until a formula is not satisfied by the simulation
status.</li>
</ul>
<div class="section level3">
<h3 id="size-based-simulation">Size-Based Simulation<a class="anchor" aria-label="anchor" href="#size-based-simulation"></a>
</h3>
<p>We can run the simulation up to when we have <span class="math inline">\(\theta &gt; 0\)</span>; cells of species
<code>A+</code></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># Counts per species is now 0</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">get_counts</span>()</span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a><span class="co">#&gt;   mutant epistate counts</span></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co">#&gt; 1      A        -      0</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="co">#&gt; 2      A        +      1</span></span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_size</span>(<span class="st">"A+"</span>, <span class="dv">500</span>)</span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Saving snapshot</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" tabindex="-1"></a><span class="co"># Counts per species now reports 500 for A+</span></span>
<span id="cb14-12"><a href="#cb14-12" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">get_counts</span>()</span>
<span id="cb14-13"><a href="#cb14-13" tabindex="-1"></a><span class="co">#&gt;   mutant epistate counts</span></span>
<span id="cb14-14"><a href="#cb14-14" tabindex="-1"></a><span class="co">#&gt; 1      A        -    123</span></span>
<span id="cb14-15"><a href="#cb14-15" tabindex="-1"></a><span class="co">#&gt; 2      A        +    500</span></span>
<span id="cb14-16"><a href="#cb14-16" tabindex="-1"></a></span>
<span id="cb14-17"><a href="#cb14-17" tabindex="-1"></a><span class="fu">plot_tissue</span>(sim)</span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-13-1.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="firing-based-simulation">Firing-Based Simulation<a class="anchor" aria-label="anchor" href="#firing-based-simulation"></a>
</h3>
<p>The number of times each event has fired is accessible</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the number of fired event per species</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_firings</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    event mutant epistate fired</span></span>
<span><span class="co">#&gt; 1  death      A        -     8</span></span>
<span><span class="co">#&gt; 2 growth      A        -    87</span></span>
<span><span class="co">#&gt; 3 switch      A        -     8</span></span>
<span><span class="co">#&gt; 4  death      A        +   366</span></span>
<span><span class="co">#&gt; 5 growth      A        +   909</span></span>
<span><span class="co">#&gt; 6 switch      A        +    52</span></span></code></pre></div>
<p>A small number of cell deaths have occurred in species
<code>A-</code> up to this point, so we can simulate the system until
there are 100 of them.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_event</span>(<span class="st">"death"</span>, <span class="st">"A-"</span>, <span class="dv">100</span>)</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Saving snapshot</span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a><span class="co"># The row "death", for "A" "-" now reports 100</span></span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">get_firings</span>()</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a><span class="co">#&gt;    event mutant epistate fired</span></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="co">#&gt; 1  death      A        -   100</span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a><span class="co">#&gt; 2 growth      A        -   893</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a><span class="co">#&gt; 3 switch      A        -    98</span></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a><span class="co">#&gt; 4  death      A        +  3456</span></span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a><span class="co">#&gt; 5 growth      A        +  7354</span></span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a><span class="co">#&gt; 6 switch      A        +   387</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the tissue by using 200 bins</span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-16-1.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="clock-based-simulation">Clock-Based Simulation<a class="anchor" aria-label="anchor" href="#clock-based-simulation"></a>
</h3>
<p>It is also possible to take the current simulation clock as
reference, and simulate further.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># Get the simulation clock</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">get_clock</span>()</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a><span class="co">#&gt; [1] 64.19203</span></span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="co"># Run the simulation for other 15 time units</span></span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_time</span>(sim<span class="sc">$</span><span class="fu">get_clock</span>() <span class="sc">+</span> <span class="dv">15</span>)</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Saving snapshot</span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a></span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a><span class="co"># Get again the simulation clock</span></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">get_clock</span>()</span>
<span id="cb18-12"><a href="#cb18-12" tabindex="-1"></a><span class="co">#&gt; [1] 79.19229</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-18-1.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="formula-based-simulation-advanced">Formula-Based Simulation (Advanced)<a class="anchor" aria-label="anchor" href="#formula-based-simulation-advanced"></a>
</h3>
<p>rRACES implements a first order unquantified logic having variables
representing the cardinality of the species, the number of events fired
in a species (being duplications, deaths, or switches), and the
simulation time. These variables and reals values are summed by
<code>+</code>, subtracted by <code>-</code>, and multiplied by
<code>*</code> and form expressions. The expressions are then compared
with the standard semantics by <code>&gt;</code>, <code>&gt;=</code>,
<code>==</code>, <code>!=</code>, <code>&lt;=</code>, and
<code>&lt;</code> to form relations. A formula in this language is
either a relation, the conjunction of two formulas (<code>&amp;</code>),
or the disjunction of two formulas (<code>|</code>).</p>
<p>Any formula in above language expresses a condition on the simulation
status and it can be used as the parameter of the method
<code>Simulation$run_until()</code> to let the simulation evolve until
the condition does not hold.</p>
<div class="section level4">
<h4 id="variables">Variables<a class="anchor" aria-label="anchor" href="#variables"></a>
</h4>
<p>The variables represent one among the following quantities:</p>
<ul>
<li>the cardinality of a species;</li>
<li>the number fired event among deaths, duplications and switches in a
species;</li>
<li>the elapse simulation time.</li>
</ul>
<p>All above variables can be built by using the method
<code>Simulation$var()</code>. When the parameter is the string
<code>"Time"</code>, the elapsed simulation time variable is
returned.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build a simulation and add two species to it</span></span>
<span><span class="va">sim2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">Simulation</span><span class="op">)</span></span>
<span><span class="va">sim2</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>                epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.01</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,</span>
<span>                growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.08</span><span class="op">)</span>,</span>
<span>                death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># get the variable representing the simulation time</span></span>
<span><span class="va">v_time</span> <span class="op">&lt;-</span> <span class="va">sim2</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span><span class="st">"Time"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">v_time</span></span>
<span><span class="co">#&gt; Time</span></span></code></pre></div>
<p>When the parameter is the name of a species, a variable representing
the cardinality of the species is built.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the variable representing the cardinality of A+ in sim2</span></span>
<span><span class="va">va_p</span> <span class="op">&lt;-</span> <span class="va">sim2</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span><span class="st">"A+"</span><span class="op">)</span></span>
<span><span class="va">va_p</span></span>
<span><span class="co">#&gt; |A+|</span></span>
<span></span>
<span><span class="co"># get the variable representing the cardinality of A- in sim2</span></span>
<span><span class="va">va_m</span> <span class="op">&lt;-</span> <span class="va">sim2</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span><span class="st">"A-"</span><span class="op">)</span></span>
<span><span class="va">va_m</span></span>
<span><span class="co">#&gt; |A-|</span></span></code></pre></div>
<p>Finally, when the parameter is the name of a species followed by a
<code>.</code> and the name of an event among <code>deaths</code>,
<code>duplications</code>, or <code>switches</code>,
<code>Simulation$var()</code> returns the variable associated with the
number of the corresponding event in the specified species.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># get the variable representing the number of epigenetic</span></span>
<span><span class="co"># switches from A+</span></span>
<span><span class="va">va_ps</span> <span class="op">&lt;-</span> <span class="va">sim2</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span><span class="st">"A+.switches"</span><span class="op">)</span></span>
<span><span class="va">va_ps</span></span>
<span><span class="co">#&gt; |A+.switches|</span></span>
<span></span>
<span><span class="co"># get the variable representing the number of duplications in A+</span></span>
<span><span class="va">sim2</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span><span class="st">"A+.duplications"</span><span class="op">)</span></span>
<span><span class="co">#&gt; |A+.duplications|</span></span>
<span></span>
<span><span class="co"># get the variable representing the number of deaths in A+</span></span>
<span><span class="va">sim2</span><span class="op">$</span><span class="fu">var</span><span class="op">(</span><span class="st">"A+.deaths"</span><span class="op">)</span></span>
<span><span class="co">#&gt; |A+.deaths|</span></span></code></pre></div>
<div class="section level5">
<h5 id="expressions-and-formulas">Expressions and Formulas<a class="anchor" aria-label="anchor" href="#expressions-and-formulas"></a>
</h5>
<p>An expression is one of the following object:</p>
<ul>
<li>a variable, e.g., <code>sim2$var("A+")</code>;</li>
<li>a numeric value, e.g., <code>3.4</code>;</li>
<li>the sum of two expressions, e.g.,
<code>sim2$var("A+") + 3.4</code>;</li>
<li>the subtraction of two expressions, e.g.,
<code>sim2$var("A+") - 3.4</code>;</li>
<li>the multiplication of two expressions, e.g.,
<code>sim2$var("A+") * 3.4</code>.</li>
</ul>
<p>Two expression can be related by <code>&lt;=</code>,
<code>&lt;</code>, <code>==</code>, <code>!=</code>, <code>&gt;</code>
and <code>&gt;=</code>.</p>
<p>A formula is:</p>
<ul>
<li>a relation among two expressions, e.g.,
<code>sim2$var("A+")&gt;=2</code>;</li>
<li>the conjunction of two formulas, e.g.,
<code>sim2$var("A+")&gt;=2 &amp; sim2$var("A+")&lt;=500</code>;</li>
<li>the disjunction of two formulas, e.g.,
<code>sim2$var("A+")&gt;=2 | sim2$var("A+")&lt;=500</code>.</li>
</ul>
<p>The method <code>Simulation$run_until()</code> takes as the parameter
a formula and lets the simulation evolve until the formula does not
hold.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co"># build a condition stating that the cardinality of A+ doubles</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a><span class="co"># that of A-</span></span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a>c1 <span class="ot">&lt;-</span> va_p <span class="sc">&gt;=</span> <span class="dv">2</span> <span class="sc">*</span> va_m</span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>c1</span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="co">#&gt; |A+|&gt;=2*|A-|</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a><span class="co"># build a condition that holds when there are more than</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a><span class="co"># 100000 live cells of mutant A</span></span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a>c2 <span class="ot">&lt;-</span> va_p <span class="sc">+</span> va_m <span class="sc">&gt;</span> <span class="fl">1e5</span></span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a>c2</span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a><span class="co">#&gt; |A+|+|A-|&gt;100000</span></span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" tabindex="-1"></a><span class="co"># build a condition that holds when less than 4000</span></span>
<span id="cb23-14"><a href="#cb23-14" tabindex="-1"></a><span class="co"># epigenetic switches from the species A+ have occured</span></span>
<span id="cb23-15"><a href="#cb23-15" tabindex="-1"></a>c3 <span class="ot">&lt;-</span> va_ps <span class="sc">&lt;</span> <span class="dv">4000</span></span>
<span id="cb23-16"><a href="#cb23-16" tabindex="-1"></a>c3</span>
<span id="cb23-17"><a href="#cb23-17" tabindex="-1"></a><span class="co">#&gt; 4000&gt;|A+.switches|</span></span>
<span id="cb23-18"><a href="#cb23-18" tabindex="-1"></a></span>
<span id="cb23-19"><a href="#cb23-19" tabindex="-1"></a><span class="co"># build a condition that holds when 40 time unit have been</span></span>
<span id="cb23-20"><a href="#cb23-20" tabindex="-1"></a><span class="co"># simulated at least</span></span>
<span id="cb23-21"><a href="#cb23-21" tabindex="-1"></a>c4 <span class="ot">&lt;-</span> v_time <span class="sc">&gt;=</span> <span class="dv">40</span></span>
<span id="cb23-22"><a href="#cb23-22" tabindex="-1"></a>c4</span>
<span id="cb23-23"><a href="#cb23-23" tabindex="-1"></a><span class="co">#&gt; Time&gt;=40</span></span>
<span id="cb23-24"><a href="#cb23-24" tabindex="-1"></a></span>
<span id="cb23-25"><a href="#cb23-25" tabindex="-1"></a><span class="co"># build a condition that holds when c4 and at least one</span></span>
<span id="cb23-26"><a href="#cb23-26" tabindex="-1"></a><span class="co"># among c1, c2, and c3 hold</span></span>
<span id="cb23-27"><a href="#cb23-27" tabindex="-1"></a>c5 <span class="ot">&lt;-</span> c4 <span class="sc">&amp;</span> (c1 <span class="sc">|</span> c2 <span class="sc">|</span> c3)</span>
<span id="cb23-28"><a href="#cb23-28" tabindex="-1"></a>c5</span>
<span id="cb23-29"><a href="#cb23-29" tabindex="-1"></a><span class="co">#&gt; Time&gt;=40 and (|A+|&gt;=2*|A-| or |A+|+|A-|&gt;100000 or 4000&gt;|A+.switches|)</span></span>
<span id="cb23-30"><a href="#cb23-30" tabindex="-1"></a></span>
<span id="cb23-31"><a href="#cb23-31" tabindex="-1"></a><span class="co"># place the initial cell</span></span>
<span id="cb23-32"><a href="#cb23-32" tabindex="-1"></a>sim2<span class="sc">$</span><span class="fu">place_cell</span>(<span class="st">"A+"</span>, <span class="dv">500</span>, <span class="dv">500</span>)</span>
<span id="cb23-33"><a href="#cb23-33" tabindex="-1"></a></span>
<span id="cb23-34"><a href="#cb23-34" tabindex="-1"></a><span class="co"># run the simulation while c5 does not hold</span></span>
<span id="cb23-35"><a href="#cb23-35" tabindex="-1"></a>sim2<span class="sc">$</span><span class="fu">run_until</span>(c5)</span>
<span id="cb23-36"><a href="#cb23-36" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-37"><a href="#cb23-37" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Saving snapshot</span>
<span id="cb23-38"><a href="#cb23-38" tabindex="-1"></a></span>
<span id="cb23-39"><a href="#cb23-39" tabindex="-1"></a>sim2</span>
<span id="cb23-40"><a href="#cb23-40" tabindex="-1"></a><span class="co">#&gt; ──  [43m rRACES  [49m  [37m [42m D  [49m [39m  [37m [41m S  [49m [39m  [37m [41m M  [49m [39m races_20240430-183806 ──  [31m▣  [39mA tissue [1000x1000] [31m  ⏱  [39m40 ──</span></span>
<span id="cb23-41"><a href="#cb23-41" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-42"><a href="#cb23-42" tabindex="-1"></a><span class="co">#&gt; ── Species:  [32m [32m2 [32m [39m,  [32m [32m [32mwith [32m [32m [39m epigenetics</span></span>
<span id="cb23-43"><a href="#cb23-43" tabindex="-1"></a><span class="co">#&gt;    </span></span>
<span id="cb23-44"><a href="#cb23-44" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ========</span></span>
<span id="cb23-45"><a href="#cb23-45" tabindex="-1"></a><span class="co">#&gt;    species   λ      δ    ε    counts     %    </span></span>
<span id="cb23-46"><a href="#cb23-46" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ========</span></span>
<span id="cb23-47"><a href="#cb23-47" tabindex="-1"></a><span class="co">#&gt;         A-  0.08  0.01  0.01    55    16.51652</span></span>
<span id="cb23-48"><a href="#cb23-48" tabindex="-1"></a><span class="co">#&gt;         A+  0.20  0.10  0.01   278    83.48348</span></span>
<span id="cb23-49"><a href="#cb23-49" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ========</span></span>
<span id="cb23-50"><a href="#cb23-50" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-51"><a href="#cb23-51" tabindex="-1"></a><span class="co">#&gt; ── Firings: 714 total</span></span>
<span id="cb23-52"><a href="#cb23-52" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb23-53"><a href="#cb23-53" tabindex="-1"></a><span class="co">#&gt;  Species [A-]:   3 (deaths),  42 (duplications) and   7 (switches)</span></span>
<span id="cb23-54"><a href="#cb23-54" tabindex="-1"></a><span class="co">#&gt;  Species [A+]: 173 (deaths), 466 (duplications) and  23 (switches)</span></span>
<span id="cb23-55"><a href="#cb23-55" tabindex="-1"></a><span class="co">#&gt;  [31m✖ [39m The simulation has no samples yet!</span></span>
<span id="cb23-56"><a href="#cb23-56" tabindex="-1"></a>sim2<span class="sc">$</span><span class="fu">get_clock</span>()</span>
<span id="cb23-57"><a href="#cb23-57" tabindex="-1"></a><span class="co">#&gt; [1] 40.00333</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="getting-cells-advanced">Getting Cells (Advanced)<a class="anchor" aria-label="anchor" href="#getting-cells-advanced"></a>
</h5>
<p>At this point, if we query the simulation we will find more cells (we
use <code>dplyr</code> to process query results). For convenience, the
getters accept parameters to subset the tissue.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load dplyr to use %&gt;%</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue at current simulation time</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   73236      A        +        415        494</span></span>
<span><span class="co">#&gt; 2   76326      A        +        416        481</span></span>
<span><span class="co">#&gt; 3   78132      A        +        416        493</span></span>
<span><span class="co">#&gt; 4   68611      A        +        416        494</span></span>
<span><span class="co">#&gt; 5   73135      A        +        416        501</span></span>
<span><span class="co">#&gt; 6   73136      A        +        416        502</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue rectangular sample having</span></span>
<span><span class="co"># [500,500] and [505,505] as lower and upper corners, respectively</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">505</span>, <span class="fl">505</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   71933      A        +        500        500</span></span>
<span><span class="co">#&gt; 2   60610      A        +        500        501</span></span>
<span><span class="co">#&gt; 3   73408      A        +        500        502</span></span>
<span><span class="co">#&gt; 4   64157      A        +        500        503</span></span>
<span><span class="co">#&gt; 5   59244      A        -        500        504</span></span>
<span><span class="co">#&gt; 6   52866      A        -        500        505</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue having epigenetic state "-"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   54513      A        -        416        506</span></span>
<span><span class="co">#&gt; 2   59158      A        -        416        510</span></span>
<span><span class="co">#&gt; 3   59157      A        -        416        512</span></span>
<span><span class="co">#&gt; 4   57052      A        -        417        481</span></span>
<span><span class="co">#&gt; 5   70530      A        -        417        485</span></span>
<span><span class="co">#&gt; 6   40445      A        -        417        490</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue having epigenetic state "-" and,</span></span>
<span><span class="co"># at the same time, belonging to rectangular sample bounded by</span></span>
<span><span class="co"># [500,500] and [505,505] as lower and upper corners, respectively</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">505</span>, <span class="fl">505</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   59244      A        -        500        504</span></span>
<span><span class="co">#&gt; 2   52866      A        -        500        505</span></span>
<span><span class="co">#&gt; 3   41096      A        -        502        501</span></span>
<span><span class="co">#&gt; 4   46077      A        -        502        503</span></span>
<span><span class="co">#&gt; 5   51553      A        -        502        504</span></span>
<span><span class="co">#&gt; 6   42089      A        -        503        505</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="evolving-new-species">Evolving new species<a class="anchor" aria-label="anchor" href="#evolving-new-species"></a>
</h3>
<p>rRACES can select cells from the tissue, randomly for every mutant,
or in a constrained tissue area.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Stochastic sampling from the whole tissue: it can return A+ or A-</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   59145      A        +        523        427</span></span>
<span></span>
<span><span class="co"># Calling it again may result in a different cell</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   59135      A        +        459        478</span></span>
<span></span>
<span><span class="co"># Constrain sampling in the tissue rectangular selection [500,550]x[350,450]</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span>,  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">350</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">550</span>, <span class="fl">450</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   66820      A        +        527        444</span></span></code></pre></div>
<p>This feature can be used to program the generation of new species,
mimicking new mutants that generate subclonal expansions.</p>
<p>Imagine we want to add a new mutant <code>B</code> with epistates –
and therefore new species <code>B+</code> and <code>B-</code> - as
descending from <code>A,</code>we need to:</p>
<ul>
<li><p>locate one cell of mutant <code>A</code> in the tissue, which is
where we will inject the new mutant;</p></li>
<li><p>add the specifics of mutant <code>B</code> (via
<code>Simulation$add_mutant()</code>, as we did for
<code>A</code>);</p></li>
<li><p>implement the change of the cell of mutant <code>A</code> to a
cell of mutant <code>B</code>.</p></li>
</ul>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We locate a random cell</span></span>
<span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="va">cell</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   38877      A        -        471        478</span></span>
<span></span>
<span><span class="co"># Add mutant</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.05</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.3</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.05</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then we inject the cell, and simulate a little bit.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" tabindex="-1"></a><span class="co"># Mutant injection</span></span>
<span id="cb27-2"><a href="#cb27-2" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">mutate_progeny</span>(cell, <span class="st">"B"</span>)</span>
<span id="cb27-3"><a href="#cb27-3" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" tabindex="-1"></a><span class="co"># Generated event A -&gt; B either in + or - epistates is now recorded</span></span>
<span id="cb27-5"><a href="#cb27-5" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">get_counts</span>()</span>
<span id="cb27-6"><a href="#cb27-6" tabindex="-1"></a><span class="co">#&gt;   mutant epistate counts</span></span>
<span id="cb27-7"><a href="#cb27-7" tabindex="-1"></a><span class="co">#&gt; 1      A        -   5095</span></span>
<span id="cb27-8"><a href="#cb27-8" tabindex="-1"></a><span class="co">#&gt; 2      A        +  17315</span></span>
<span id="cb27-9"><a href="#cb27-9" tabindex="-1"></a><span class="co">#&gt; 3      B        -      1</span></span>
<span id="cb27-10"><a href="#cb27-10" tabindex="-1"></a><span class="co">#&gt; 4      B        +      0</span></span>
<span id="cb27-11"><a href="#cb27-11" tabindex="-1"></a></span>
<span id="cb27-12"><a href="#cb27-12" tabindex="-1"></a><span class="co"># New evolution</span></span>
<span id="cb27-13"><a href="#cb27-13" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_size</span>(<span class="st">"B+"</span>, <span class="dv">600</span>)</span>
<span id="cb27-14"><a href="#cb27-14" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb27-15"><a href="#cb27-15" tabindex="-1"></a> [█<span class="sc">---------------------------------------</span>] <span class="dv">2</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">80105</span></span>
<span id="cb27-16"><a href="#cb27-16" tabindex="-1"></a> [██████<span class="sc">----------------------------------</span>] <span class="dv">14</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">127671</span></span>
<span id="cb27-17"><a href="#cb27-17" tabindex="-1"></a> [████████████<span class="sc">----------------------------</span>] <span class="dv">28</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">01</span>s] Cells<span class="sc">:</span> <span class="dv">163611</span></span>
<span id="cb27-18"><a href="#cb27-18" tabindex="-1"></a> [████████████████████<span class="sc">--------------------</span>] <span class="dv">49</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">02</span>s] Cells<span class="sc">:</span> <span class="dv">194850</span></span>
<span id="cb27-19"><a href="#cb27-19" tabindex="-1"></a> [██████████████████████████████<span class="sc">----------</span>] <span class="dv">73</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">03</span>s] Cells<span class="sc">:</span> <span class="dv">221281</span></span>
<span id="cb27-20"><a href="#cb27-20" tabindex="-1"></a> [███████████████████████████████████████<span class="sc">-</span>] <span class="dv">96</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">04</span>s] Cells<span class="sc">:</span> <span class="dv">246277</span></span>
<span id="cb27-21"><a href="#cb27-21" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">05</span>s] Saving snapshot</span></code></pre></div>
<p>At this point, we can inspect in more details the tissue. It can help
to facet on the species to clearly appreciate the spatial diffusion of
the populations.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-27-1.png" width="288"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-27-2.png" width="288"></p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Facet on species via ggplot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">species</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-28-1.png" width="576"></p>
<p>If, at this point in the simulation, we generate a new mutant
<code>C</code> from <code>A</code> in the rectangle <span class="math inline">\([450,500]\times [550, 600]\)</span>.</p>
<div class="sourceCode" id="cb31"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" tabindex="-1"></a><span class="co"># Define evolutionary parameters</span></span>
<span id="cb31-2"><a href="#cb31-2" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">add_mutant</span>(<span class="at">name =</span> <span class="st">"C"</span>,</span>
<span id="cb31-3"><a href="#cb31-3" tabindex="-1"></a>               <span class="at">epigenetic_rates =</span> <span class="fu">c</span>(<span class="st">"+-"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"-+"</span> <span class="ot">=</span> <span class="fl">0.1</span>),</span>
<span id="cb31-4"><a href="#cb31-4" tabindex="-1"></a>               <span class="at">growth_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.4</span>),</span>
<span id="cb31-5"><a href="#cb31-5" tabindex="-1"></a>               <span class="at">death_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.01</span>))</span>
<span id="cb31-6"><a href="#cb31-6" tabindex="-1"></a></span>
<span id="cb31-7"><a href="#cb31-7" tabindex="-1"></a><span class="co"># Choose and mutate</span></span>
<span id="cb31-8"><a href="#cb31-8" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">mutate_progeny</span>(sim<span class="sc">$</span><span class="fu">choose_cell_in</span>(<span class="st">"A"</span>,  <span class="fu">c</span>(<span class="dv">450</span>, <span class="dv">550</span>), <span class="fu">c</span>(<span class="dv">500</span>, <span class="dv">600</span>)), <span class="st">"C"</span>)</span>
<span id="cb31-9"><a href="#cb31-9" tabindex="-1"></a></span>
<span id="cb31-10"><a href="#cb31-10" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_time</span>(sim<span class="sc">$</span><span class="fu">get_clock</span>() <span class="sc">+</span> <span class="dv">7</span>)</span>
<span id="cb31-11"><a href="#cb31-11" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb31-12"><a href="#cb31-12" tabindex="-1"></a> [██████████████████████████████████████<span class="sc">--</span>] <span class="dv">94</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">274385</span></span>
<span id="cb31-13"><a href="#cb31-13" tabindex="-1"></a> [███████████████████████████████████████<span class="sc">-</span>] <span class="dv">95</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">299003</span></span>
<span id="cb31-14"><a href="#cb31-14" tabindex="-1"></a> [███████████████████████████████████████<span class="sc">-</span>] <span class="dv">95</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">01</span>s] Cells<span class="sc">:</span> <span class="dv">321540</span></span>
<span id="cb31-15"><a href="#cb31-15" tabindex="-1"></a> [███████████████████████████████████████<span class="sc">-</span>] <span class="dv">96</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">02</span>s] Cells<span class="sc">:</span> <span class="dv">342854</span></span>
<span id="cb31-16"><a href="#cb31-16" tabindex="-1"></a> [███████████████████████████████████████<span class="sc">-</span>] <span class="dv">96</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">03</span>s] Cells<span class="sc">:</span> <span class="dv">362743</span></span>
<span id="cb31-17"><a href="#cb31-17" tabindex="-1"></a> [███████████████████████████████████████<span class="sc">-</span>] <span class="dv">97</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">04</span>s] Cells<span class="sc">:</span> <span class="dv">382111</span></span>
<span id="cb31-18"><a href="#cb31-18" tabindex="-1"></a> [███████████████████████████████████████<span class="sc">-</span>] <span class="dv">97</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">05</span>s] Cells<span class="sc">:</span> <span class="dv">400059</span></span>
<span id="cb31-19"><a href="#cb31-19" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">98</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">06</span>s] Cells<span class="sc">:</span> <span class="dv">417457</span></span>
<span id="cb31-20"><a href="#cb31-20" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">98</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">07</span>s] Cells<span class="sc">:</span> <span class="dv">434629</span></span>
<span id="cb31-21"><a href="#cb31-21" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">98</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">08</span>s] Cells<span class="sc">:</span> <span class="dv">451596</span></span>
<span id="cb31-22"><a href="#cb31-22" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">99</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">09</span>s] Cells<span class="sc">:</span> <span class="dv">468167</span></span>
<span id="cb31-23"><a href="#cb31-23" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">99</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">10</span>s] Cells<span class="sc">:</span> <span class="dv">483814</span></span>
<span id="cb31-24"><a href="#cb31-24" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">99</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">11</span>s] Cells<span class="sc">:</span> <span class="dv">498935</span></span>
<span id="cb31-25"><a href="#cb31-25" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">99</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">12</span>s] Cells<span class="sc">:</span> <span class="dv">514746</span></span>
<span id="cb31-26"><a href="#cb31-26" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">13</span>s] Saving snapshot</span></code></pre></div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-30-1.png" width="288"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-30-2.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="other-operations">Other Operations<a class="anchor" aria-label="anchor" href="#other-operations"></a>
</h3>
<div class="section level5">
<h5 id="injection-of-cells-over-a-tissue">Injection of Cells over a Tissue<a class="anchor" aria-label="anchor" href="#injection-of-cells-over-a-tissue"></a>
</h5>
<p>On the tissue, we can inject multiple cells manually; all injected
cells can be retrieved.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Now it will return just the initial cell</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_added_cells</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate position_x position_y time</span></span>
<span><span class="co">#&gt; 1      A        +        500        500    0</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="avoiding-drift">Avoiding Drift<a class="anchor" aria-label="anchor" href="#avoiding-drift"></a>
</h5>
<p>Any species that has a non-zero death rate can become extinct
stochastically by <em>drift</em>.</p>
<p>Drift makes it difficult to simulate an be confident of what species
are in the model. To facilitate the user, RACES can avoid drift by
setting a <em>death activation level.</em> This value is the minimum
number of cells that enables cell death in a species: a cell of species
<span class="math inline">\(S\)</span> can die if and only if <span class="math inline">\(S\)</span> has reached the death activation level
at least once during the simulation.</p>
<p>This threshold holds for all the species and it is set to 1 by
default.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="va">death_activation_level</span></span>
<span><span class="co">#&gt; [1] 50</span></span>
<span></span>
<span><span class="co"># Change death activation level</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="va">death_activation_level</span> <span class="op">&lt;-</span> <span class="fl">50</span></span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="time-series-data">Time Series Data<a class="anchor" aria-label="anchor" href="#time-series-data"></a>
</h2>
<p>It is convenient sometimes to plot a time series of a simulation,
reporting species or firing counts over time. Since rRACES is
programmable, it is immediate to make a for-loop algorithm and collect
the simulation data over time.</p>
<div class="section level3">
<h3 id="default-history-based-data">Default History-Based Data<a class="anchor" aria-label="anchor" href="#default-history-based-data"></a>
</h3>
<p>However, this is not required because at the end of any
<code>run_to_*</code> methods, RACES stores the data about the number of
species cells, and that of event firings. These data can be
extracted.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The firings</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_firing_history</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    event mutant epistate fired     time</span></span>
<span><span class="co">#&gt; 1  death      A        -     8 44.46369</span></span>
<span><span class="co">#&gt; 2 growth      A        -    87 44.46369</span></span>
<span><span class="co">#&gt; 3 switch      A        -     8 44.46369</span></span>
<span><span class="co">#&gt; 4  death      A        +   366 44.46369</span></span>
<span><span class="co">#&gt; 5 growth      A        +   909 44.46369</span></span>
<span><span class="co">#&gt; 6 switch      A        +    52 44.46369</span></span>
<span></span>
<span><span class="co"># For example, total number of the deaths on `B+` at the end of the</span></span>
<span><span class="co"># previous calls of the `run_to_*` methods</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_firing_history</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">event</span> <span class="op">==</span> <span class="st">"death"</span>, <span class="va">mutant</span> <span class="op">==</span> <span class="st">"B"</span>, <span class="va">epistate</span> <span class="op">==</span> <span class="st">"-"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   event mutant epistate fired      time</span></span>
<span><span class="co">#&gt; 1 death      B        -     0  44.46369</span></span>
<span><span class="co">#&gt; 2 death      B        -     0  64.19203</span></span>
<span><span class="co">#&gt; 3 death      B        -     0  79.19229</span></span>
<span><span class="co">#&gt; 4 death      B        -   224 102.41148</span></span>
<span><span class="co">#&gt; 5 death      B        -  1401 109.41148</span></span>
<span></span>
<span><span class="co"># The counts</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_count_history</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate count     time</span></span>
<span><span class="co">#&gt; 1      A        -   123 44.46369</span></span>
<span><span class="co">#&gt; 2      A        +   500 44.46369</span></span>
<span><span class="co">#&gt; 3      B        -     0 44.46369</span></span>
<span><span class="co">#&gt; 4      B        +     0 44.46369</span></span>
<span><span class="co">#&gt; 5      C        -     0 44.46369</span></span>
<span><span class="co">#&gt; 6      C        +     0 44.46369</span></span></code></pre></div>
<p>The time-series can be plot using <code><a href="../reference/plot_timeseries.html">plot_timeseries()</a></code></p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Time-series plot</span></span>
<span><span class="fu"><a href="../reference/plot_timeseries.html">plot_timeseries</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-34-1.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="custom-time-series">Custom Time-Series<a class="anchor" aria-label="anchor" href="#custom-time-series"></a>
</h3>
<p>If the default time-series is not enough coarse-grained, one can set
<code>Simulation$history_delta</code> to increase the sampling rate of
the state (by default, <code>Simulation$history_delta</code> is set to
<span class="math inline">\(0\)</span>).</p>
<p>We show this by re-simulating a tumour with two submutants.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="co"># Example time-series on a new simulation, with coarse-grained time-series</span></span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a>sim <span class="ot">&lt;-</span> <span class="fu">new</span>(Simulation, <span class="st">"Finer Time Series"</span>)</span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a></span>
<span id="cb38-4"><a href="#cb38-4" tabindex="-1"></a>sim<span class="sc">$</span>history_delta <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb38-5"><a href="#cb38-5" tabindex="-1"></a>sim<span class="sc">$</span>death_activation_level <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb38-6"><a href="#cb38-6" tabindex="-1"></a></span>
<span id="cb38-7"><a href="#cb38-7" tabindex="-1"></a><span class="co"># A</span></span>
<span id="cb38-8"><a href="#cb38-8" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">add_mutant</span>(<span class="at">name =</span> <span class="st">"A"</span>,</span>
<span id="cb38-9"><a href="#cb38-9" tabindex="-1"></a>               <span class="at">epigenetic_rates =</span> <span class="fu">c</span>(<span class="st">"+-"</span> <span class="ot">=</span> <span class="fl">0.01</span>, <span class="st">"-+"</span> <span class="ot">=</span> <span class="fl">0.01</span>),</span>
<span id="cb38-10"><a href="#cb38-10" tabindex="-1"></a>               <span class="at">growth_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.08</span>),</span>
<span id="cb38-11"><a href="#cb38-11" tabindex="-1"></a>               <span class="at">death_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.01</span>))</span>
<span id="cb38-12"><a href="#cb38-12" tabindex="-1"></a></span>
<span id="cb38-13"><a href="#cb38-13" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">place_cell</span>(<span class="st">"A+"</span>, <span class="dv">500</span>, <span class="dv">500</span>)</span>
<span id="cb38-14"><a href="#cb38-14" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_size</span>(<span class="st">"A+"</span>, <span class="dv">400</span>)</span>
<span id="cb38-15"><a href="#cb38-15" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb38-16"><a href="#cb38-16" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Saving snapshot</span>
<span id="cb38-17"><a href="#cb38-17" tabindex="-1"></a></span>
<span id="cb38-18"><a href="#cb38-18" tabindex="-1"></a><span class="co"># B (linear inside A)</span></span>
<span id="cb38-19"><a href="#cb38-19" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">add_mutant</span>(<span class="at">name =</span> <span class="st">"B"</span>,</span>
<span id="cb38-20"><a href="#cb38-20" tabindex="-1"></a>               <span class="at">epigenetic_rates =</span> <span class="fu">c</span>(<span class="st">"+-"</span> <span class="ot">=</span> <span class="fl">0.05</span>, <span class="st">"-+"</span> <span class="ot">=</span> <span class="fl">0.05</span>),</span>
<span id="cb38-21"><a href="#cb38-21" tabindex="-1"></a>               <span class="at">growth_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.3</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.3</span>),</span>
<span id="cb38-22"><a href="#cb38-22" tabindex="-1"></a>               <span class="at">death_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.05</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.1</span>))</span>
<span id="cb38-23"><a href="#cb38-23" tabindex="-1"></a></span>
<span id="cb38-24"><a href="#cb38-24" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">mutate_progeny</span>(sim<span class="sc">$</span><span class="fu">choose_cell_in</span>(<span class="st">"A"</span>), <span class="st">"B"</span>)</span>
<span id="cb38-25"><a href="#cb38-25" tabindex="-1"></a></span>
<span id="cb38-26"><a href="#cb38-26" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_size</span>(<span class="st">"B-"</span>, <span class="dv">300</span>)</span>
<span id="cb38-27"><a href="#cb38-27" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb38-28"><a href="#cb38-28" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Saving snapshot</span>
<span id="cb38-29"><a href="#cb38-29" tabindex="-1"></a></span>
<span id="cb38-30"><a href="#cb38-30" tabindex="-1"></a><span class="co"># C (linear inside B)</span></span>
<span id="cb38-31"><a href="#cb38-31" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">add_mutant</span>(<span class="at">name =</span> <span class="st">"C"</span>,</span>
<span id="cb38-32"><a href="#cb38-32" tabindex="-1"></a>               <span class="at">epigenetic_rates =</span> <span class="fu">c</span>(<span class="st">"+-"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"-+"</span> <span class="ot">=</span> <span class="fl">0.1</span>),</span>
<span id="cb38-33"><a href="#cb38-33" tabindex="-1"></a>               <span class="at">growth_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.4</span>),</span>
<span id="cb38-34"><a href="#cb38-34" tabindex="-1"></a>               <span class="at">death_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.01</span>))</span>
<span id="cb38-35"><a href="#cb38-35" tabindex="-1"></a></span>
<span id="cb38-36"><a href="#cb38-36" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">mutate_progeny</span>(sim<span class="sc">$</span><span class="fu">choose_cell_in</span>(<span class="st">"B"</span>), <span class="st">"C"</span>)</span>
<span id="cb38-37"><a href="#cb38-37" tabindex="-1"></a></span>
<span id="cb38-38"><a href="#cb38-38" tabindex="-1"></a><span class="co"># D (linear inside A, so branching with C) - same parameters of C</span></span>
<span id="cb38-39"><a href="#cb38-39" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">add_mutant</span>(<span class="at">name =</span> <span class="st">"D"</span>,</span>
<span id="cb38-40"><a href="#cb38-40" tabindex="-1"></a>               <span class="at">epigenetic_rates =</span> <span class="fu">c</span>(<span class="st">"+-"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"-+"</span> <span class="ot">=</span> <span class="fl">0.1</span>),</span>
<span id="cb38-41"><a href="#cb38-41" tabindex="-1"></a>               <span class="at">growth_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.4</span>),</span>
<span id="cb38-42"><a href="#cb38-42" tabindex="-1"></a>               <span class="at">death_rates =</span> <span class="fu">c</span>(<span class="st">"+"</span> <span class="ot">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="ot">=</span> <span class="fl">0.01</span>))</span>
<span id="cb38-43"><a href="#cb38-43" tabindex="-1"></a></span>
<span id="cb38-44"><a href="#cb38-44" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">mutate_progeny</span>(sim<span class="sc">$</span><span class="fu">choose_cell_in</span>(<span class="st">"A"</span>), <span class="st">"D"</span>)</span>
<span id="cb38-45"><a href="#cb38-45" tabindex="-1"></a></span>
<span id="cb38-46"><a href="#cb38-46" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_size</span>(<span class="st">"D+"</span>, <span class="dv">1000</span>)</span>
<span id="cb38-47"><a href="#cb38-47" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb38-48"><a href="#cb38-48" tabindex="-1"></a> [███<span class="sc">-------------------------------------</span>] <span class="dv">6</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">78004</span></span>
<span id="cb38-49"><a href="#cb38-49" tabindex="-1"></a> [██████<span class="sc">----------------------------------</span>] <span class="dv">14</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">125850</span></span>
<span id="cb38-50"><a href="#cb38-50" tabindex="-1"></a> [█████████<span class="sc">-------------------------------</span>] <span class="dv">20</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">01</span>s] Cells<span class="sc">:</span> <span class="dv">165905</span></span>
<span id="cb38-51"><a href="#cb38-51" tabindex="-1"></a> [██████████<span class="sc">------------------------------</span>] <span class="dv">24</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">02</span>s] Cells<span class="sc">:</span> <span class="dv">200461</span></span>
<span id="cb38-52"><a href="#cb38-52" tabindex="-1"></a> [██████████████<span class="sc">--------------------------</span>] <span class="dv">33</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">03</span>s] Cells<span class="sc">:</span> <span class="dv">232947</span></span>
<span id="cb38-53"><a href="#cb38-53" tabindex="-1"></a> [████████████████<span class="sc">------------------------</span>] <span class="dv">38</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">04</span>s] Cells<span class="sc">:</span> <span class="dv">263033</span></span>
<span id="cb38-54"><a href="#cb38-54" tabindex="-1"></a> [██████████████████<span class="sc">----------------------</span>] <span class="dv">43</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">05</span>s] Cells<span class="sc">:</span> <span class="dv">290062</span></span>
<span id="cb38-55"><a href="#cb38-55" tabindex="-1"></a> [████████████████████<span class="sc">--------------------</span>] <span class="dv">48</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">06</span>s] Cells<span class="sc">:</span> <span class="dv">315524</span></span>
<span id="cb38-56"><a href="#cb38-56" tabindex="-1"></a> [██████████████████████<span class="sc">------------------</span>] <span class="dv">52</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">07</span>s] Cells<span class="sc">:</span> <span class="dv">339390</span></span>
<span id="cb38-57"><a href="#cb38-57" tabindex="-1"></a> [████████████████████████<span class="sc">----------------</span>] <span class="dv">58</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">08</span>s] Cells<span class="sc">:</span> <span class="dv">362479</span></span>
<span id="cb38-58"><a href="#cb38-58" tabindex="-1"></a> [███████████████████████████<span class="sc">-------------</span>] <span class="dv">65</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">09</span>s] Cells<span class="sc">:</span> <span class="dv">383534</span></span>
<span id="cb38-59"><a href="#cb38-59" tabindex="-1"></a> [█████████████████████████████<span class="sc">-----------</span>] <span class="dv">70</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">10</span>s] Cells<span class="sc">:</span> <span class="dv">404599</span></span>
<span id="cb38-60"><a href="#cb38-60" tabindex="-1"></a> [██████████████████████████████<span class="sc">----------</span>] <span class="dv">74</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">11</span>s] Cells<span class="sc">:</span> <span class="dv">425029</span></span>
<span id="cb38-61"><a href="#cb38-61" tabindex="-1"></a> [█████████████████████████████████<span class="sc">-------</span>] <span class="dv">81</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">12</span>s] Cells<span class="sc">:</span> <span class="dv">443988</span></span>
<span id="cb38-62"><a href="#cb38-62" tabindex="-1"></a> [███████████████████████████████████<span class="sc">-----</span>] <span class="dv">85</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">13</span>s] Cells<span class="sc">:</span> <span class="dv">462484</span></span>
<span id="cb38-63"><a href="#cb38-63" tabindex="-1"></a> [█████████████████████████████████████<span class="sc">---</span>] <span class="dv">90</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">14</span>s] Cells<span class="sc">:</span> <span class="dv">480756</span></span>
<span id="cb38-64"><a href="#cb38-64" tabindex="-1"></a> [██████████████████████████████████████<span class="sc">--</span>] <span class="dv">94</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">15</span>s] Cells<span class="sc">:</span> <span class="dv">498394</span></span>
<span id="cb38-65"><a href="#cb38-65" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">99</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">16</span>s] Cells<span class="sc">:</span> <span class="dv">516057</span></span>
<span id="cb38-66"><a href="#cb38-66" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">17</span>s] Saving snapshot</span></code></pre></div>
<p>The time-series can be plot using <code><a href="../reference/plot_timeseries.html">plot_timeseries()</a></code>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Time-series plot</span></span>
<span><span class="fu"><a href="../reference/plot_timeseries.html">plot_timeseries</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-36-1.png" width="384"></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Logscale helps seeing the different effective growth rates</span></span>
<span><span class="fu"><a href="../reference/plot_timeseries.html">plot_timeseries</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in ggplot2::scale_y_log10(): <span style="color: #00BB00;">log-10</span> transformation introduced infinite values.</span></span>
<span><span class="co">#&gt; <span style="color: #00BB00;">log-10</span> transformation introduced infinite values.</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-36-2.png" width="384"></p>
</div>
<div class="section level3">
<h3 id="muller-plot">Muller Plot<a class="anchor" aria-label="anchor" href="#muller-plot"></a>
</h3>
<p>We can also get a Muller plot of the evolution using <a href="https://CRAN.R-project.org/package=ggmuller" class="external-link">ggmuller</a>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># default Muller plot</span></span>
<span><span class="fu"><a href="../reference/plot_muller.html">plot_muller</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-37-1.png" width="576"></p>
<p>In this case every population is annotated as a descendant of the
ancestor mutant. Note however that reversible espistates do not fit a
traditional Muller plot because they violate the no-back mutation
model.</p>
<p>In this case, <em>rRACES</em> will show first the epistate that was
randomly injected in the simulation, and the second will result by
linear. This is not a completely correct perspective of the simulation
time-series; still, it help understand trends.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Custom Mullers</span></span>
<span><span class="va">clock</span> <span class="op">&lt;-</span> <span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_muller.html">plot_muller</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span> <span class="op">+</span> <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="va">clock</span> <span class="op">*</span> <span class="fl">3</span><span class="op">/</span><span class="fl">4</span>, <span class="va">clock</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-38-1.png" width="576"></p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot_muller.html">plot_muller</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="va">clock</span> <span class="op">*</span> <span class="fl">3</span><span class="op">/</span><span class="fl">4</span>, <span class="va">clock</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-38-2.png" width="576"></p>
</div>
</div>
<div class="section level2">
<h2 id="time-varying-evolutionary-rates">Time-Varying Evolutionary Rates<a class="anchor" aria-label="anchor" href="#time-varying-evolutionary-rates"></a>
</h2>
<p>You can model the fact that the rates of one species. For instance,
this happens when a population is subject to a targeted treatment.</p>
<p>Considering the example above, where <code>C</code> and
<code>D</code> have the same rates, we increase the death rate of both
<code>C+</code> and <code>C-</code> species, as well as <code>B+</code>
and <code>B-</code>.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" tabindex="-1"></a><span class="co"># Current rates</span></span>
<span id="cb44-2"><a href="#cb44-2" tabindex="-1"></a>sim</span>
<span id="cb44-3"><a href="#cb44-3" tabindex="-1"></a><span class="co">#&gt;    </span></span>
<span id="cb44-4"><a href="#cb44-4" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ==========</span></span>
<span id="cb44-5"><a href="#cb44-5" tabindex="-1"></a><span class="co">#&gt;    species   λ      δ    ε    counts      %     </span></span>
<span id="cb44-6"><a href="#cb44-6" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ==========</span></span>
<span id="cb44-7"><a href="#cb44-7" tabindex="-1"></a><span class="co">#&gt;         A-  0.08  0.01  0.01  14630   2.8219763 </span></span>
<span id="cb44-8"><a href="#cb44-8" tabindex="-1"></a><span class="co">#&gt;         A+  0.20  0.10  0.01  52828   10.1899771</span></span>
<span id="cb44-9"><a href="#cb44-9" tabindex="-1"></a><span class="co">#&gt;         B-  0.30  0.10  0.05  166090  32.0370503</span></span>
<span id="cb44-10"><a href="#cb44-10" tabindex="-1"></a><span class="co">#&gt;         B+  0.30  0.05  0.05  274615  52.9704049</span></span>
<span id="cb44-11"><a href="#cb44-11" tabindex="-1"></a><span class="co">#&gt;         C-  0.40  0.01  0.10   4688   0.9042669 </span></span>
<span id="cb44-12"><a href="#cb44-12" tabindex="-1"></a><span class="co">#&gt;         C+  0.20  0.10  0.10   1615   0.3115169 </span></span>
<span id="cb44-13"><a href="#cb44-13" tabindex="-1"></a><span class="co">#&gt;         D-  0.40  0.01  0.10   2965   0.5719180 </span></span>
<span id="cb44-14"><a href="#cb44-14" tabindex="-1"></a><span class="co">#&gt;         D+  0.20  0.10  0.10   1000   0.1928897 </span></span>
<span id="cb44-15"><a href="#cb44-15" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ==========</span></span>
<span id="cb44-16"><a href="#cb44-16" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb44-17"><a href="#cb44-17" tabindex="-1"></a><span class="co">#&gt;  Species [A-]:   1303 (deaths),  12217 (duplications) and   1337 (switches)</span></span>
<span id="cb44-18"><a href="#cb44-18" tabindex="-1"></a><span class="co">#&gt;  Species [A+]:  50520 (deaths), 107065 (duplications) and   5053 (switches)</span></span>
<span id="cb44-19"><a href="#cb44-19" tabindex="-1"></a><span class="co">#&gt;  Species [B-]:  58335 (deaths), 204870 (duplications) and  29272 (switches)</span></span>
<span id="cb44-20"><a href="#cb44-20" tabindex="-1"></a><span class="co">#&gt;  Species [B+]:  49200 (deaths), 343370 (duplications) and  48827 (switches)</span></span>
<span id="cb44-21"><a href="#cb44-21" tabindex="-1"></a><span class="co">#&gt;  Species [C-]:    108 (deaths),   5510 (duplications) and   1154 (switches)</span></span>
<span id="cb44-22"><a href="#cb44-22" tabindex="-1"></a><span class="co">#&gt;  Species [C+]:    398 (deaths),   1298 (duplications) and    440 (switches)</span></span>
<span id="cb44-23"><a href="#cb44-23" tabindex="-1"></a><span class="co">#&gt;  Species [D-]:     72 (deaths),   3436 (duplications) and    665 (switches)</span></span>
<span id="cb44-24"><a href="#cb44-24" tabindex="-1"></a><span class="co">#&gt;  Species [D+]:    249 (deaths),    849 (duplications) and    266 (switches)</span></span>
<span id="cb44-25"><a href="#cb44-25" tabindex="-1"></a></span>
<span id="cb44-26"><a href="#cb44-26" tabindex="-1"></a><span class="co"># Raise the death rate levels</span></span>
<span id="cb44-27"><a href="#cb44-27" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">update_rates</span>(<span class="st">"B+"</span>, <span class="fu">c</span>(<span class="at">death =</span> <span class="dv">3</span>))</span>
<span id="cb44-28"><a href="#cb44-28" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">update_rates</span>(<span class="st">"B-"</span>, <span class="fu">c</span>(<span class="at">death =</span> <span class="dv">3</span>))</span>
<span id="cb44-29"><a href="#cb44-29" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">update_rates</span>(<span class="st">"C+"</span>, <span class="fu">c</span>(<span class="at">death =</span> <span class="dv">3</span>))</span>
<span id="cb44-30"><a href="#cb44-30" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">update_rates</span>(<span class="st">"C-"</span>, <span class="fu">c</span>(<span class="at">death =</span> <span class="dv">3</span>))</span>
<span id="cb44-31"><a href="#cb44-31" tabindex="-1"></a></span>
<span id="cb44-32"><a href="#cb44-32" tabindex="-1"></a><span class="co"># Now D will become larger</span></span>
<span id="cb44-33"><a href="#cb44-33" tabindex="-1"></a>sim<span class="sc">$</span><span class="fu">run_up_to_size</span>(<span class="st">"D+"</span>, <span class="dv">6000</span>)</span>
<span id="cb44-34"><a href="#cb44-34" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb44-35"><a href="#cb44-35" tabindex="-1"></a> [███████<span class="sc">---------------------------------</span>] <span class="dv">17</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">456293</span></span>
<span id="cb44-36"><a href="#cb44-36" tabindex="-1"></a> [███████<span class="sc">---------------------------------</span>] <span class="dv">17</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">00</span>s] Cells<span class="sc">:</span> <span class="dv">379481</span></span>
<span id="cb44-37"><a href="#cb44-37" tabindex="-1"></a> [████████<span class="sc">--------------------------------</span>] <span class="dv">18</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">01</span>s] Cells<span class="sc">:</span> <span class="dv">286363</span></span>
<span id="cb44-38"><a href="#cb44-38" tabindex="-1"></a> [█████████<span class="sc">-------------------------------</span>] <span class="dv">21</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">02</span>s] Cells<span class="sc">:</span> <span class="dv">170668</span></span>
<span id="cb44-39"><a href="#cb44-39" tabindex="-1"></a> [███████████████████<span class="sc">---------------------</span>] <span class="dv">45</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">03</span>s] Cells<span class="sc">:</span> <span class="dv">99085</span></span>
<span id="cb44-40"><a href="#cb44-40" tabindex="-1"></a> [████████████████████████████████████████] <span class="dv">100</span>% [<span class="dv">00</span>m<span class="sc">:</span><span class="dv">04</span>s] Saving snapshot</span>
<span id="cb44-41"><a href="#cb44-41" tabindex="-1"></a></span>
<span id="cb44-42"><a href="#cb44-42" tabindex="-1"></a><span class="co"># Current state</span></span>
<span id="cb44-43"><a href="#cb44-43" tabindex="-1"></a>sim</span>
<span id="cb44-44"><a href="#cb44-44" tabindex="-1"></a><span class="co">#&gt;    </span></span>
<span id="cb44-45"><a href="#cb44-45" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ==========</span></span>
<span id="cb44-46"><a href="#cb44-46" tabindex="-1"></a><span class="co">#&gt;    species   λ      δ    ε    counts      %     </span></span>
<span id="cb44-47"><a href="#cb44-47" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ==========</span></span>
<span id="cb44-48"><a href="#cb44-48" tabindex="-1"></a><span class="co">#&gt;         A-  0.08  0.01  0.01  23112   17.7318132</span></span>
<span id="cb44-49"><a href="#cb44-49" tabindex="-1"></a><span class="co">#&gt;         A+  0.20  0.10  0.01  82617   63.3847877</span></span>
<span id="cb44-50"><a href="#cb44-50" tabindex="-1"></a><span class="co">#&gt;         B-  0.30  3.00  0.05    5     0.0038361 </span></span>
<span id="cb44-51"><a href="#cb44-51" tabindex="-1"></a><span class="co">#&gt;         B+  0.30  3.00  0.05    3     0.0023016 </span></span>
<span id="cb44-52"><a href="#cb44-52" tabindex="-1"></a><span class="co">#&gt;         C-  0.40  3.00  0.10    0     0.0000000 </span></span>
<span id="cb44-53"><a href="#cb44-53" tabindex="-1"></a><span class="co">#&gt;         C+  0.20  3.00  0.10    0     0.0000000 </span></span>
<span id="cb44-54"><a href="#cb44-54" tabindex="-1"></a><span class="co">#&gt;         D-  0.40  0.01  0.10  18605   14.2739869</span></span>
<span id="cb44-55"><a href="#cb44-55" tabindex="-1"></a><span class="co">#&gt;         D+  0.20  0.10  0.10   6000   4.6032745 </span></span>
<span id="cb44-56"><a href="#cb44-56" tabindex="-1"></a><span class="co">#&gt;    =======  ====  ====  ====  ======  ==========</span></span>
<span id="cb44-57"><a href="#cb44-57" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb44-58"><a href="#cb44-58" tabindex="-1"></a><span class="co">#&gt;  Species [A-]:   2081 (deaths),  19381 (duplications) and   2128 (switches)</span></span>
<span id="cb44-59"><a href="#cb44-59" tabindex="-1"></a><span class="co">#&gt;  Species [A+]:  79227 (deaths), 167657 (duplications) and   7940 (switches)</span></span>
<span id="cb44-60"><a href="#cb44-60" tabindex="-1"></a><span class="co">#&gt;  Species [B-]: 248750 (deaths), 227181 (duplications) and  32501 (switches)</span></span>
<span id="cb44-61"><a href="#cb44-61" tabindex="-1"></a><span class="co">#&gt;  Species [B+]: 357997 (deaths), 379574 (duplications) and  54075 (switches)</span></span>
<span id="cb44-62"><a href="#cb44-62" tabindex="-1"></a><span class="co">#&gt;  Species [C-]:   5551 (deaths),   6398 (duplications) and   1349 (switches)</span></span>
<span id="cb44-63"><a href="#cb44-63" tabindex="-1"></a><span class="co">#&gt;  Species [C+]:   2322 (deaths),   1474 (duplications) and    502 (switches)</span></span>
<span id="cb44-64"><a href="#cb44-64" tabindex="-1"></a><span class="co">#&gt;  Species [D-]:    464 (deaths),  21983 (duplications) and   4325 (switches)</span></span>
<span id="cb44-65"><a href="#cb44-65" tabindex="-1"></a><span class="co">#&gt;  Species [D+]:   1416 (deaths),   4501 (duplications) and   1411 (switches)</span></span>
<span id="cb44-66"><a href="#cb44-66" tabindex="-1"></a></span>
<span id="cb44-67"><a href="#cb44-67" tabindex="-1"></a><span class="co"># This now show the change in rates</span></span>
<span id="cb44-68"><a href="#cb44-68" tabindex="-1"></a>clock <span class="ot">&lt;-</span> sim<span class="sc">$</span><span class="fu">get_clock</span>()</span>
<span id="cb44-69"><a href="#cb44-69" tabindex="-1"></a><span class="fu">plot_muller</span>(sim) <span class="sc">+</span> ggplot2<span class="sc">::</span><span class="fu">xlim</span>(clock <span class="sc">*</span> <span class="dv">3</span><span class="sc">/</span><span class="dv">4</span>, clock)</span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-39-1.png" width="576"></p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="../reference/plot_muller.html">plot_muller</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="va">clock</span> <span class="op">*</span> <span class="fl">3</span><span class="op">/</span><span class="fl">4</span>, <span class="va">clock</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-39-2.png" width="576"></p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Giulio Caravagna, Alberto Casagrande.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
