<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Tissue simulation • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tissue simulation">
<meta name="description" content="Learn how to simulate tumour evolution
">
<meta property="og:description" content="Learn how to simulate tumour evolution
">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/rRACES.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a></li>
    <li><a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a></li>
    <li><a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a></li>
    <li><a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">More articles...</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Tissue simulation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACES/vignettes/tissue_simulation.Rmd" class="external-link"><code>vignettes/tissue_simulation.Rmd</code></a></small>
      <div class="d-none name"><code>tissue_simulation.Rmd</code></div>
    </div>

    
    
<blockquote>
<p><em>Disclaimer:</em> RACES/rRACES internally implements the
probability distributions using the C++11 random number distribution
classes. The standard does not specify their algorithms, and the class
implementations are left free for the compiler. Thus, the simulation
output depends on the compiler used to compile RACES, and because of
that, the results reported in this article may differ from those
obtained by the reader.</p>
</blockquote>
<p>To simulate a tissue the following steps are required:</p>
<ol style="list-style-type: decimal">
<li><p>creation of a tissue;</p></li>
<li><p>introduction of cells in the tissue;</p></li>
<li><p>actual simulation.</p></li>
</ol>
<p>The simulation is managed by an object of the S4 class
<code>SpatialSimulation</code>, wich allows to program the tissue
evolution over time, adding new cells as far as the simulation
progresses. The state of the simulation and tissue can be visualised
using <code>ggplot</code>-powered plots.</p>
<div class="section level2">
<h2 id="tissue-specification">Tissue specification<a class="anchor" aria-label="anchor" href="#tissue-specification"></a>
</h2>
<p>To perform a simulation a new object of class
<code>SpatialSimulation</code> must be created.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://caravagnalab.github.io/rRACES">rRACES</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Default constructor</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialSimulation.html">SpatialSimulation</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This automatically builds a 1000x1000-cells tissue - which can host
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>1</mn><annotation encoding="application/x-tex">1</annotation></semantics></math>
million cells - and sets the name of the simulation to be
<code>races_&lt;date&gt;_&lt;hour&gt;</code>.</p>
<p>A simulation custom name can be specified as it follows.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://caravagnalab.github.io/rRACES">rRACES</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># set the seed of the random number generator</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Call your simulation "Test"</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialSimulation.html">SpatialSimulation</a></span><span class="op">(</span><span class="st">"Test"</span><span class="op">)</span></span></code></pre></div>
<p>The <code><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed()</a></code> call is not mandatory and can be avoided;
however, it is highly suggested to guarantee the simulation
repeatability.</p>
<p>In order to save the simulation progresses in the disk and recover it
in the future, use the optional Boolean parameter
<code>save_snapshots</code>. By setting <code>save_snapshots</code> to
<code>TRUE</code>, the simulation progresses will be saved in a
directory whose name is the name of the simulation.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The progresses of the simulation will saved in the "Test" folder.</span></span>
<span><span class="co"># If the directory "Test" already exists an exception is raised.</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialSimulation.html">SpatialSimulation</a></span><span class="op">(</span><span class="st">"Test"</span>, save_snapshots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Class <code>SpatialSimulation</code> exports a
<code>SpatialSimulation$show()</code> method to get information on the
current object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span></span>
<span><span class="co">#&gt; ── <span style="background-color: #BBBB00;"> rRACES </span> <span style="color: #BBBBBB; background-color: #BB0000;"> D </span> <span style="color: #BBBBBB; background-color: #BB0000;"> S </span> <span style="color: #BBBBBB; background-color: #BB0000;"> M </span> Test ──────────────────────────── <span style="color: #BB0000;">▣ </span> [1000x1000]<span style="color: #BB0000;">  ⏱ </span>0 ──</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> The simulation has no samples yet!</span></span></code></pre></div>
<p>The <code>sim</code> object exposes also methods to get information
about the simulation and control it.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the simulation directory, i.e., "Test"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_name</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Test"</span></span>
<span></span>
<span><span class="co"># Get the tissue size, i.e., c(1000,1000)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_size</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1000 1000</span></span></code></pre></div>
<div class="section level3">
<h3 id="custom-tissue-sizes">Custom Tissue Sizes<a class="anchor" aria-label="anchor" href="#custom-tissue-sizes"></a>
</h3>
<p>The tissue sizes can be specified during simulation creation using
the parameters <code>width</code> and <code>height</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># build a spatial simulation whose tissue has width 1200 and</span></span>
<span><span class="co"># height 900</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/SpatialSimulation.html">SpatialSimulation</a></span><span class="op">(</span><span class="st">"Test"</span>, width<span class="op">=</span><span class="fl">1200</span>, height<span class="op">=</span><span class="fl">900</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the tissue size, i.e., c(1200,900)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_size</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1200  900</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="species-specification">Species specification<a class="anchor" aria-label="anchor" href="#species-specification"></a>
</h2>
<p>In order to simulate the evolution of some species we need to add
them to <code>sim</code>. This process defines the evolutionary
parameters of the species.</p>
<p>A <em>mutant</em> is a set of cells having the same (potentially
unknown) driver mutations. Cells in the same mutant can have different
liveness rates due to different epigenetic states.</p>
<p>A <em>species</em> is a mutant with an optional epigenetic state. At
this point in the simulation, the mutant is just a name (<code>A</code>,
<code>B</code>, ..) that, at a later stage could be linked to mutations
of interest. The epigenome is a binary feature of a species that is
represented by epistates <code>+</code>/<code>-</code> (positive and
negative status). This is an abstraction, and could represent an
active/inactive state linked to a promoter methylation or, more broadly,
a phenotype. The evolution of mutants is non-reversible (no-back
mutations model), while the evolution among epistates is potentially
reversible.</p>
<p>For example, if we define two mutants <code>A</code> and
<code>B</code> with their epistates <code>+</code>/<code>-</code>, we
have 4 distinct species: <code>A+</code> and <code>A-</code>, as well as
<code>B+</code> and <code>B-</code>.</p>
<p>Hybrid models can be obtained, e.g., a mutant <code>A</code> (with no
epistates), together with <code>B+</code> and <code>B-</code>.</p>
<div class="section level3">
<h3 id="evolutionary-parameters">Evolutionary parameters<a class="anchor" aria-label="anchor" href="#evolutionary-parameters"></a>
</h3>
<p>We use a notation common in linear birth-death processes. If a
species <code>A</code> has no epistates then its stochastic behaviour is
defined by the state-change rates</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(growth)</mtext><mspace width="1.0em"></mspace><mi>A</mi></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><mi>λ</mi></msub><mn>2</mn><mi>A</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(death)</mtext><mspace width="1.0em"></mspace><mi>A</mi></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><mi>δ</mi></msub><mi>∅</mi></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
\text{(growth)}\quad A &amp; \rightarrow_{\lambda} 2 A \\
\text{(death)}\quad A &amp; \rightarrow_{\delta} \emptyset
\end{align}
</annotation></semantics></math></p>
<p>where:</p>
<ul>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>λ</mi><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\lambda&gt;0</annotation></semantics></math>
is a growth rate for cells that duplicate;</li>
<li>
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>δ</mi><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\delta&gt;0</annotation></semantics></math>
is a death rate for cells that duplicate.</li>
</ul>
<p>Instead, if the species has epistate <code>+</code> (denoted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>A</mi><mo>⊕</mo></msub><annotation encoding="application/x-tex">A_\oplus</annotation></semantics></math>)
and <code>-</code> (denoted
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>A</mi><mo>⊖</mo></msub><annotation encoding="application/x-tex">A_\ominus</annotation></semantics></math>),
then its stochastic behaviour is defined by the state-change rates</p>
<p><math display="block" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mtable><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(growth +)</mtext><mspace width="1.0em"></mspace><msub><mi>A</mi><mo>⊕</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><msub><mi>λ</mi><mo>⊕</mo></msub></msub><mn>2</mn><msub><mi>A</mi><mo>⊕</mo></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(death +)</mtext><mspace width="1.0em"></mspace><msub><mi>A</mi><mo>⊕</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><msub><mi>δ</mi><mo>⊕</mo></msub></msub><mi>∅</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(growth -)</mtext><mspace width="1.0em"></mspace><msub><mi>A</mi><mo>⊖</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><msub><mi>λ</mi><mo>⊖</mo></msub></msub><mn>2</mn><msub><mi>A</mi><mo>⊖</mo></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(death -)</mtext><mspace width="1.0em"></mspace><msub><mi>A</mi><mo>⊖</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><msub><mi>δ</mi><mo>⊖</mo></msub></msub><mi>∅</mi></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(switch +-)</mtext><mspace width="1.0em"></mspace><msub><mi>A</mi><mo>⊕</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><msub><mi>ϵ</mi><mrow><mi>+</mi><mo>−</mo></mrow></msub></msub><msub><mi>A</mi><mo>⊕</mo></msub><mo>+</mo><msub><mi>A</mi><mo>⊖</mo></msub></mtd></mtr><mtr><mtd columnalign="right" style="text-align: right"><mtext mathvariant="normal">(switch -+)</mtext><mspace width="1.0em"></mspace><msub><mi>A</mi><mo>⊖</mo></msub></mtd><mtd columnalign="left" style="text-align: left"><msub><mo>→</mo><msub><mi>ϵ</mi><mrow><mi>−</mi><mo>+</mo></mrow></msub></msub><msub><mi>A</mi><mo>⊖</mo></msub><mo>+</mo><msub><mi>A</mi><mo>⊕</mo></msub></mtd></mtr></mtable><annotation encoding="application/x-tex">
\begin{align}
\text{(growth +)}\quad A_\oplus &amp; \rightarrow_{\lambda_\oplus} 2 A_\oplus \\
\text{(death +)}\quad A_\oplus &amp; \rightarrow_{\delta_\oplus} \emptyset \\
\text{(growth -)}\quad A_\ominus &amp; \rightarrow_{\lambda_\ominus} 2 A_\ominus \\
\text{(death -)}\quad A_\ominus &amp; \rightarrow_{\delta_\ominus} \emptyset \\
\text{(switch +-)}\quad A_\oplus &amp; \rightarrow_{\epsilon_{+-}} A_\oplus + A_\ominus \\
\text{(switch -+)}\quad A_\ominus &amp; \rightarrow_{\epsilon_{-+}} A_\ominus + A_\oplus
\end{align}
</annotation></semantics></math></p>
<p>where the rates
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>λ</mi><annotation encoding="application/x-tex">\lambda</annotation></semantics></math>
and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>δ</mi><annotation encoding="application/x-tex">\delta</annotation></semantics></math>
are as above, and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ϵ</mi><mrow><mi>+</mi><mo>−</mo></mrow></msub><annotation encoding="application/x-tex">\epsilon_{+-}</annotation></semantics></math>
or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><msub><mi>ϵ</mi><mrow><mi>−</mi><mo>+</mo></mrow></msub><annotation encoding="application/x-tex">\epsilon_{-+}</annotation></semantics></math>
are the ratesat which cells of a certain epistate duplicate and flip the
epigenetic marker of one of the progeny.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.01</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.08</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.05</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># updated object (counts refer to number of cells of each species)</span></span>
<span><span class="va">sim</span></span>
<span><span class="co">#&gt; ── <span style="background-color: #BBBB00;"> rRACES </span> <span style="color: #BBBBBB; background-color: #BB0000;"> D </span> <span style="color: #BBBBBB; background-color: #BB0000;"> S </span> <span style="color: #BBBBBB; background-color: #BB0000;"> M </span> Test ───────────────────────────── <span style="color: #BB0000;">▣ </span> [1200x900]<span style="color: #BB0000;">  ⏱ </span>0 ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Species: <span style="color: #00BB00;">2</span>, <span style="color: #00BB00;">with</span> epigenetics</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt;    species   λ      δ    ε    counts   % </span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt;         A-  0.08  0.01  0.01    0     NaN</span></span>
<span><span class="co">#&gt;         A+  0.20  0.05  0.01    0     NaN</span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Firings: 0 total</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> The simulation has no samples yet!</span></span>
<span></span>
<span><span class="co"># Get the simulation species and their rates</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_species</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate growth_rate death_rate switch_rate</span></span>
<span><span class="co">#&gt; 1      A        -        0.08       0.01        0.01</span></span>
<span><span class="co">#&gt; 2      A        +        0.20       0.05        0.01</span></span></code></pre></div>
<p>A mutant without epistates could be added as</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Not run</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"A"</span>, growth_rates <span class="op">=</span> <span class="fl">0.2</span>, death_rates <span class="op">=</span>  <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p>To be able to simulate the model, an initial cell needs to be
displaced in the tissue.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We add one cell of species A+ (mutant A in epistate +) in</span></span>
<span><span class="co"># position (500, 500).</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">place_cell</span><span class="op">(</span><span class="st">"A+"</span>, <span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualisations">Visualisations<a class="anchor" aria-label="anchor" href="#visualisations"></a>
</h3>
<p>We can query the current state of the simulation, and extract the
position of each cell in the tissue.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Counts per species</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts overall</span></span>
<span><span class="co">#&gt; 1      A        -      0       0</span></span>
<span><span class="co">#&gt; 2      A        +      1       1</span></span>
<span></span>
<span><span class="co"># Cells position (one so far)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1       0      A        +        500        500</span></span></code></pre></div>
<p>These information can be plot. Note that the tissue visualisation
uses hexagonal bins to avoid rastering delays when the simulation uses
thousands of cells.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Piechar for counts</span></span>
<span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-12-1.png" width="288"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Spatial distribution for the whole tissue (hexagonal bins)</span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-12-2.png" width="288"></p>
<blockquote>
<p><strong>Note:</strong> since the plots are done with
<code>ggplot</code> they can be assembled and customised.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="species-evolution">Species Evolution<a class="anchor" aria-label="anchor" href="#species-evolution"></a>
</h2>
<p>There are 4 ways to let the simulation evolve:</p>
<ul>
<li>advancing until the number of cells in a species reaches a given
threshold
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>θ</mi><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\theta &gt; 0</annotation></semantics></math>;</li>
<li>advancing until a new time
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>t</mi><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">t&gt;0</annotation></semantics></math>
is reached;</li>
<li>advancing until a desired number of firings (of one particular
event) has occurred;</li>
<li>advancing until a formula is not satisfied by the simulation status
(<em>advanced topic</em>; see <code><a href="../articles/run_until.html">vignette("run_until")</a></code>).</li>
</ul>
<div class="section level3">
<h3 id="size-based-simulation">Size-Based Simulation<a class="anchor" aria-label="anchor" href="#size-based-simulation"></a>
</h3>
<p>We can run the simulation up to when we have
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>θ</mi><mo>&gt;</mo><mn>0</mn></mrow><annotation encoding="application/x-tex">\theta &gt; 0</annotation></semantics></math>;
cells of species <code>A+</code></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Counts per species is now 0</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts overall</span></span>
<span><span class="co">#&gt; 1      A        -      0       0</span></span>
<span><span class="co">#&gt; 2      A        +      1       1</span></span>
<span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_size</span><span class="op">(</span><span class="st">"A+"</span>, <span class="fl">500</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span>
<span></span>
<span><span class="co"># Counts per species now reports 500 for A+</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts overall</span></span>
<span><span class="co">#&gt; 1      A        -     42      70</span></span>
<span><span class="co">#&gt; 2      A        +    500    1648</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-13-1.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="firing-based-simulation">Firing-Based Simulation<a class="anchor" aria-label="anchor" href="#firing-based-simulation"></a>
</h3>
<p>The number of times each event has fired is accessible</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the number of fired event per species</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_firings</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    event mutant epistate fired</span></span>
<span><span class="co">#&gt; 1  death      A        -     8</span></span>
<span><span class="co">#&gt; 2 growth      A        -    17</span></span>
<span><span class="co">#&gt; 3 switch      A        -     3</span></span>
<span><span class="co">#&gt; 4  death      A        +   290</span></span>
<span><span class="co">#&gt; 5 growth      A        +   822</span></span>
<span><span class="co">#&gt; 6 switch      A        +    36</span></span></code></pre></div>
<p>A small number of cell deaths have occurred in species
<code>A-</code> up to this point, so we can simulate the system until
there are 100 of them.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_event</span><span class="op">(</span><span class="st">"death"</span>, <span class="st">"A-"</span>, <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span>
<span></span>
<span><span class="co"># The row "death", for "A" "-" now reports 100</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_firings</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    event mutant epistate fired</span></span>
<span><span class="co">#&gt; 1  death      A        -   100</span></span>
<span><span class="co">#&gt; 2 growth      A        -   419</span></span>
<span><span class="co">#&gt; 3 switch      A        -    51</span></span>
<span><span class="co">#&gt; 4  death      A        +  3042</span></span>
<span><span class="co">#&gt; 5 growth      A        +  6022</span></span>
<span><span class="co">#&gt; 6 switch      A        +   281</span></span></code></pre></div>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the tissue by using 200 bins</span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-16-1.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="clock-based-simulation">Clock-Based Simulation<a class="anchor" aria-label="anchor" href="#clock-based-simulation"></a>
</h3>
<p>It is also possible to take the current simulation clock as
reference, and simulate further.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the simulation clock</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 81.62523</span></span>
<span></span>
<span><span class="co"># Run the simulation for other 15 time units</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_time</span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span>
<span></span>
<span><span class="co"># Get again the simulation clock</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 96.62686</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-18-1.png" width="288"></p>
<div class="section level5">
<h5 id="getting-cells-advanced">Getting Cells (Advanced)<a class="anchor" aria-label="anchor" href="#getting-cells-advanced"></a>
</h5>
<p>At this point, if we query the simulation we will find more cells (we
use <code>dplyr</code> to process query results). For convenience, the
getters accept parameters to subset the tissue.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load dplyr to use %&gt;%</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue at current simulation time</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   21730      A        +        458        497</span></span>
<span><span class="co">#&gt; 2   21697      A        +        458        498</span></span>
<span><span class="co">#&gt; 3   19846      A        +        459        483</span></span>
<span><span class="co">#&gt; 4   18816      A        +        459        489</span></span>
<span><span class="co">#&gt; 5   20217      A        -        459        494</span></span>
<span><span class="co">#&gt; 6   21729      A        +        459        497</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue rectangular sample having</span></span>
<span><span class="co"># [500,500] and [505,505] as lower and upper corners, respectively</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">505</span>, <span class="fl">505</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   12150      A        +        500        500</span></span>
<span><span class="co">#&gt; 2   20546      A        +        500        501</span></span>
<span><span class="co">#&gt; 3   17481      A        +        500        502</span></span>
<span><span class="co">#&gt; 4   17482      A        +        500        503</span></span>
<span><span class="co">#&gt; 5   20091      A        -        500        504</span></span>
<span><span class="co">#&gt; 6   19642      A        +        500        505</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue having epigenetic state "-"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   20217      A        -        459        494</span></span>
<span><span class="co">#&gt; 2   21494      A        -        459        504</span></span>
<span><span class="co">#&gt; 3   20144      A        -        460        484</span></span>
<span><span class="co">#&gt; 4   20218      A        -        460        493</span></span>
<span><span class="co">#&gt; 5   18519      A        -        460        496</span></span>
<span><span class="co">#&gt; 6   21493      A        -        460        504</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue having epigenetic state "-" and,</span></span>
<span><span class="co"># at the same time, belonging to rectangular sample bounded by</span></span>
<span><span class="co"># [500,500] and [505,505] as lower and upper corners, respectively</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">505</span>, <span class="fl">505</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   20091      A        -        500        504</span></span>
<span><span class="co">#&gt; 2    9405      A        -        502        500</span></span>
<span><span class="co">#&gt; 3   21653      A        -        502        504</span></span>
<span><span class="co">#&gt; 4   21654      A        -        503        505</span></span>
<span><span class="co">#&gt; 5    4689      A        -        505        502</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="evolving-new-species">Evolving new species<a class="anchor" aria-label="anchor" href="#evolving-new-species"></a>
</h3>
<p>rRACES can select cells from the tissue, randomly for every mutant,
or in a constrained tissue area.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Stochastic sampling from the whole tissue: it can return A+ or A-</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   21730      A        +        458        497</span></span>
<span></span>
<span><span class="co"># Calling it again may result in a different cell</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   21546      A        +        492        455</span></span>
<span></span>
<span><span class="co"># Constrain sampling in the tissue rectangular selection [500,550]x[350,450]</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span>,  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">350</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">550</span>, <span class="fl">450</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   17516      A        +        494        456</span></span></code></pre></div>
<p>This feature can be used to program the generation of new species,
mimicking new mutants that generate subclonal expansions.</p>
<p>Imagine we want to add a new mutant <code>B</code> with epistates –
and therefore new species <code>B+</code> and <code>B-</code> - as
descending from <code>A,</code>we need to:</p>
<ul>
<li><p>locate one cell of mutant <code>A</code> in the tissue, which is
where we will inject the new mutant;</p></li>
<li><p>add the specifics of mutant <code>B</code> (via
<code>SpatialSimulation$add_mutant()</code>, as we did for
<code>A</code>);</p></li>
<li><p>implement the change of the cell of mutant <code>A</code> to a
cell of mutant <code>B</code>.</p></li>
</ul>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We locate a random cell</span></span>
<span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="va">cell</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   21546      A        +        492        455</span></span>
<span></span>
<span><span class="co"># Add mutant</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.05</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.7</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.05</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then we inject the cell, and simulate a little bit.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Mutant injection</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">mutate_progeny</span><span class="op">(</span><span class="va">cell</span>, <span class="st">"B"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generated event A -&gt; B either in + or - epistates is now recorded</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts overall</span></span>
<span><span class="co">#&gt; 1      A        -   1008    3146</span></span>
<span><span class="co">#&gt; 2      A        +   4162   24669</span></span>
<span><span class="co">#&gt; 3      B        -      0       0</span></span>
<span><span class="co">#&gt; 4      B        +      1       1</span></span>
<span></span>
<span><span class="co"># New evolution</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_size</span><span class="op">(</span><span class="st">"B+"</span>, <span class="fl">600</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span></code></pre></div>
<p>At this point, we can inspect in more details the tissue. It can help
to facet on the species to clearly appreciate the spatial diffusion of
the populations.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-23-1.png" width="288"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-23-2.png" width="288"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Facet on species via ggplot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">species</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
<p>If, at this point in the simulation, we generate a new mutant
<code>C</code> from <code>A</code> in the rectangle
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mrow><mo stretchy="true" form="prefix">[</mo><mn>450</mn><mo>,</mo><mn>500</mn><mo stretchy="true" form="postfix">]</mo></mrow><mo>×</mo><mrow><mo stretchy="true" form="prefix">[</mo><mn>550</mn><mo>,</mo><mn>600</mn><mo stretchy="true" form="postfix">]</mo></mrow></mrow><annotation encoding="application/x-tex">[450,500]\times [550, 600]</annotation></semantics></math>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define evolutionary parameters</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"C"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.1</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Choose and mutate</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">mutate_progeny</span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span>,  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">450</span>, <span class="fl">550</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">600</span><span class="op">)</span><span class="op">)</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_time</span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-26-1.png" width="288"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-26-2.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="other-operations">Other Operations<a class="anchor" aria-label="anchor" href="#other-operations"></a>
</h3>
<div class="section level5">
<h5 id="injection-of-cells-over-a-tissue">Injection of Cells over a Tissue<a class="anchor" aria-label="anchor" href="#injection-of-cells-over-a-tissue"></a>
</h5>
<p>On the tissue, we can inject multiple cells manually; all injected
cells can be retrieved.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Now it will return just the initial cell</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_added_cells</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate position_x position_y time</span></span>
<span><span class="co">#&gt; 1      A        +        500        500    0</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="avoiding-drift">Avoiding Drift<a class="anchor" aria-label="anchor" href="#avoiding-drift"></a>
</h5>
<p>Any species that has a non-zero death rate can become extinct
stochastically by <em>drift</em>.</p>
<p>Drift makes it difficult to simulate an be confident of what species
are in the model. To facilitate the user, RACES can avoid drift by
setting a <em>death activation level.</em> This value is the minimum
number of cells that enables cell death in a species: a cell of species
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>S</mi><annotation encoding="application/x-tex">S</annotation></semantics></math>
can die if and only if
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>S</mi><annotation encoding="application/x-tex">S</annotation></semantics></math>
has reached the death activation level at least once during the
simulation.</p>
<p>This threshold holds for all the species and it is set to 1 by
default.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="va">death_activation_level</span></span>
<span><span class="co">#&gt; [1] 1</span></span>
<span></span>
<span><span class="co"># Change death activation level</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="va">death_activation_level</span> <span class="op">&lt;-</span> <span class="fl">50</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
