<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Learn how to simulate tumour evolution
">
<title>Tissue simulation • rRACES</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Tissue simulation">
<meta property="og:description" content="Learn how to simulate tumour evolution
">
<meta property="og:image" content="https://caravagnalab.github.io/rRACES/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">rRACES</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../articles/rRACES.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/tissue_simulation.html">Tissue simulation</a>
    <a class="dropdown-item" href="../articles/sampling.html">Tissue sampling</a>
    <a class="dropdown-item" href="../articles/mutations.html">Simulating Mutations</a>
    <a class="dropdown-item" href="../articles/sequencing.html">Sequencing Simulation</a>
    <div class="dropdown-divider"></div>
    <a class="dropdown-item" href="../articles/index.html">More articles...</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/caravagnalab/rRACES/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Tissue simulation</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rRACESvignettes/tissue_simulation.Rmd" class="external-link"><code>vignettes/tissue_simulation.Rmd</code></a></small>
      <div class="d-none name"><code>tissue_simulation.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">rRACES</span><span class="op">)</span></span></code></pre></div>
<p>To simulate a tissue the following steps are required:</p>
<ol style="list-style-type: decimal">
<li><p>creation of a tissue;</p></li>
<li><p>introduction of cells in the tissue;</p></li>
<li><p>actual simulation.</p></li>
</ol>
<p>The simulation is managed by an object of the S4 class
<code>Simulation</code>, wich allows to program the tissue evolution
over time, adding new cells as far as the simulation progresses. The
state of the simulation and tissue can be visualised using
<code>ggplot</code>-powered plots.</p>
<div class="section level2">
<h2 id="tissue-specification">Tissue specification<a class="anchor" aria-label="anchor" href="#tissue-specification"></a>
</h2>
<p>To perform a simulation a new object of class <code>Simulation</code>
must be created.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Default constructor</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">Simulation</span><span class="op">)</span></span></code></pre></div>
<p>This automatically builds a 1000x1000-cells tissue - which can host
<span class="math inline">\(1\)</span> million cells - and sets the name
of the simulation to be
<code>races_&lt;date&gt;_&lt;hour&gt;</code>.</p>
<p>A simulation custom name can be specified as it follows.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Call your simulation "Test"</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">Simulation</span>, <span class="st">"Test"</span><span class="op">)</span></span></code></pre></div>
<p>In order to save the simulation progresses in the disk and recover it
in the future, use the optional Boolean parameter
<code>save_snapshots</code>. By setting <code>save_snapshots</code> to
<code>TRUE</code>, the simulation progresses will be saved in a
directory whose name is the name of the simulation.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The progresses of the simulation will saved in the "Test" folder.</span></span>
<span><span class="co"># If the directory "Test" already exists an exception is raised.</span></span>
<span><span class="va">sim</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/new.html" class="external-link">new</a></span><span class="op">(</span><span class="va">Simulation</span>, <span class="st">"Test"</span>, save_snapshots <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Class <code>Simulation</code> exports a
<code>Simulation$show()</code> method to get information on the current
object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span></span>
<span><span class="co">#&gt; ── <span style="background-color: #BBBB00;"> rRACES </span> <span style="color: #BBBBBB; background-color: #BB0000;"> D </span> <span style="color: #BBBBBB; background-color: #BB0000;"> S </span> <span style="color: #BBBBBB; background-color: #BB0000;"> M </span> Test ──────────────────── <span style="color: #BB0000;">▣ </span>A tissue [1000x1000]<span style="color: #BB0000;">  ⏱ </span>0 ──</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> The simulation has no samples yet!</span></span></code></pre></div>
<p>The <code>sim</code> object exposes also methods to get information
about the simulation and control it.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the simulation directory, i.e., "Test"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_name</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Test"</span></span>
<span></span>
<span><span class="co"># Get the tissue size, i.e., c(1000,1000)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_size</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1000 1000</span></span></code></pre></div>
<div class="section level3">
<h3 id="custom-tissue">Custom tissue<a class="anchor" aria-label="anchor" href="#custom-tissue"></a>
</h3>
<p>We can change the tissue size and label it with a name that might
appear in some plots.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set the tissue name and size to different values</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">update_tissue</span><span class="op">(</span><span class="st">"Liver"</span>, <span class="fl">1200</span>, <span class="fl">900</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the tissue name, i.e., "Liver"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_name</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Liver"</span></span>
<span></span>
<span><span class="co"># Get the tissue size, i.e., c(1200,900)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_tissue_size</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1200  900</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="species-specification">Species specification<a class="anchor" aria-label="anchor" href="#species-specification"></a>
</h2>
<p>In order to simulate the evolution of some species we need to add
them to <code>sim</code>. This process defines the evolutionary
parameters of the species.</p>
<p>A <em>mutant</em> is a set of cells having the same (potentially
unknown) driver mutations. Cells in the same mutant can have different
liveness rates due to different epigenetic states.</p>
<p>A <em>species</em> is a mutant with an optional epigenetic state. At
this point in the simulation, the mutant is just a name (<code>A</code>,
<code>B</code>, ..) that, at a later stage could be linked to mutations
of interest. The epigenome is a binary feature of a species that is
represented by epistates <code>+</code>/<code>-</code> (positive and
negative status). This is an abstraction, and could represent an
active/inactive state linked to a promoter methylation or, more broadly,
a phenotype. The evolution of mutants is non-reversible (no-back
mutations model), while the evolution among epistates is potentially
reversible.</p>
<p>For example, if we define two mutants <code>A</code> and
<code>B</code> with their epistates <code>+</code>/<code>-</code>, we
have 4 distinct species: <code>A+</code> and <code>A-</code>, as well as
<code>B+</code> and <code>B-</code>.</p>
<p>Hybrid models can be obtained, e.g., a mutant <code>A</code> (with no
epistates), together with <code>B+</code> and <code>B-</code>.</p>
<div class="section level3">
<h3 id="evolutionary-parameters">Evolutionary parameters<a class="anchor" aria-label="anchor" href="#evolutionary-parameters"></a>
</h3>
<p>We use a notation common in linear birth-death processes. If a
species <code>A</code> has no epistates then its stochastic behaviour is
defined by the state-change rates</p>
<p><span class="math display">\[
\begin{align}
\text{(growth)}\quad A &amp; \rightarrow_{\lambda} 2 A \\
\text{(death)}\quad A &amp; \rightarrow_{\delta} \emptyset
\end{align}
\]</span></p>
<p>where:</p>
<ul>
<li>
<span class="math inline">\(\lambda&gt;0\)</span> is a growth rate
for cells that duplicate;</li>
<li>
<span class="math inline">\(\delta&gt;0\)</span> is a death rate for
cells that duplicate.</li>
</ul>
<p>Instead, if the species has epistate <code>+</code> (denoted <span class="math inline">\(A_\oplus\)</span>) and <code>-</code> (denoted
<span class="math inline">\(A_\ominus\)</span>), then its stochastic
behaviour is defined by the state-change rates</p>
<p><span class="math display">\[
\begin{align}
\text{(growth +)}\quad A_\oplus &amp; \rightarrow_{\lambda_\oplus} 2
A_\oplus \\
\text{(death +)}\quad A_\oplus &amp; \rightarrow_{\delta_\oplus}
\emptyset \\
\text{(growth -)}\quad A_\ominus &amp; \rightarrow_{\lambda_\ominus} 2
A_\ominus \\
\text{(death -)}\quad A_\ominus &amp; \rightarrow_{\delta_\ominus}
\emptyset \\
\text{(switch +-)}\quad A_\oplus &amp; \rightarrow_{\epsilon_{+-}}
A_\oplus + A_\ominus \\
\text{(switch -+)}\quad A_\ominus &amp; \rightarrow_{\epsilon_{-+}}
A_\ominus + A_\oplus
\end{align}
\]</span></p>
<p>where the rates <span class="math inline">\(\lambda\)</span> and
<span class="math inline">\(\delta\)</span> are as above, and <span class="math inline">\(\epsilon_{+-}\)</span> or <span class="math inline">\(\epsilon_{-+}\)</span> are the ratesat which cells
of a certain epistate duplicate and flip the epigenetic marker of one of
the progeny.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"A"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.01</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.08</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># updated object (counts refer to number of cells of each species)</span></span>
<span><span class="va">sim</span></span>
<span><span class="co">#&gt; ── <span style="background-color: #BBBB00;"> rRACES </span> <span style="color: #BBBBBB; background-color: #BB0000;"> D </span> <span style="color: #BBBBBB; background-color: #BB0000;"> S </span> <span style="color: #BBBBBB; background-color: #BB0000;"> M </span> Test ──────────────────────── <span style="color: #BB0000;">▣ </span>Liver [1200x900]<span style="color: #BB0000;">  ⏱ </span>0 ──</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Species: <span style="color: #00BB00;">2</span>, <span style="color: #00BB00;">with</span> epigenetics</span></span>
<span><span class="co">#&gt;    </span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt;    species   λ      δ    ε    counts   % </span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt;         A-  0.08  0.01  0.01    0     NaN</span></span>
<span><span class="co">#&gt;         A+  0.20  0.10  0.01    0     NaN</span></span>
<span><span class="co">#&gt;    =======  ====  ====  ====  ======  ===</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ── Firings: 0 total</span></span>
<span><span class="co">#&gt; <span style="color: #BB0000;">✖</span> The simulation has no samples yet!</span></span>
<span></span>
<span><span class="co"># Get the simulation species and their rates</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_species</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate growth_rate death_rate switch_rate</span></span>
<span><span class="co">#&gt; 1      A        -        0.08       0.01        0.01</span></span>
<span><span class="co">#&gt; 2      A        +        0.20       0.10        0.01</span></span></code></pre></div>
<p>A mutant without epistates could be added as</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Not run</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"A"</span>, growth_rates <span class="op">=</span> <span class="fl">0.2</span>, death_rates <span class="op">=</span>  <span class="fl">0.1</span><span class="op">)</span></span></code></pre></div>
<p>To be able to simulate the model, an initial cell needs to be
displaced in the tissue.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We add one cell of species A+ (mutant A in epistate +) in</span></span>
<span><span class="co"># position (500, 500).</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">place_cell</span><span class="op">(</span><span class="st">"A+"</span>, <span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualisations">Visualisations<a class="anchor" aria-label="anchor" href="#visualisations"></a>
</h3>
<p>We can query the current state of the simulation, and extract the
position of each cell in the tissue.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Counts per species</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts</span></span>
<span><span class="co">#&gt; 1      A        -      0</span></span>
<span><span class="co">#&gt; 2      A        +      1</span></span>
<span></span>
<span><span class="co"># Cells position (one so far)</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1       0      A        +        500        500</span></span></code></pre></div>
<p>These information can be plot. Note that the tissue visualisation
uses hexagonal bins to avoid rastering delays when the simulation uses
thousands of cells.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Piechar for counts</span></span>
<span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-12-1.png" width="288"></p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Spatial distribution for the whole tissue (hexagonal bins)</span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-12-2.png" width="288"></p>
<blockquote>
<p><strong>Note:</strong> since the plots are done with
<code>ggplot</code> they can be assembled and customised.</p>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="species-evolution">Species Evolution<a class="anchor" aria-label="anchor" href="#species-evolution"></a>
</h2>
<p>There are 4 ways to let the simulation evolve:</p>
<ul>
<li>advancing until the number of cells in a species reaches a given
threshold <span class="math inline">\(\theta &gt; 0\)</span>;</li>
<li>advancing until a new time <span class="math inline">\(t&gt;0\)</span> is reached;</li>
<li>advancing until a desired number of firings (of one particular
event) has occurred;</li>
<li>advancing until a formula is not satisfied by the simulation status
(<em>advanced topic</em>; see <code><a href="../articles/run_until.html">vignette("run_until")</a></code>).</li>
</ul>
<div class="section level3">
<h3 id="size-based-simulation">Size-Based Simulation<a class="anchor" aria-label="anchor" href="#size-based-simulation"></a>
</h3>
<p>We can run the simulation up to when we have <span class="math inline">\(\theta &gt; 0\)</span>; cells of species
<code>A+</code></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Counts per species is now 0</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts</span></span>
<span><span class="co">#&gt; 1      A        -      0</span></span>
<span><span class="co">#&gt; 2      A        +      1</span></span>
<span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_size</span><span class="op">(</span><span class="st">"A+"</span>, <span class="fl">500</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span>
<span></span>
<span><span class="co"># Counts per species now reports 500 for A+</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts</span></span>
<span><span class="co">#&gt; 1      A        -    263</span></span>
<span><span class="co">#&gt; 2      A        +    500</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-13-1.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="firing-based-simulation">Firing-Based Simulation<a class="anchor" aria-label="anchor" href="#firing-based-simulation"></a>
</h3>
<p>The number of times each event has fired is accessible</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the number of fired event per species</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_firings</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    event mutant epistate fired</span></span>
<span><span class="co">#&gt; 1  death      A        -    46</span></span>
<span><span class="co">#&gt; 2 growth      A        -   266</span></span>
<span><span class="co">#&gt; 3 switch      A        -    31</span></span>
<span><span class="co">#&gt; 4  death      A        +   852</span></span>
<span><span class="co">#&gt; 5 growth      A        +  1394</span></span>
<span><span class="co">#&gt; 6 switch      A        +    74</span></span></code></pre></div>
<p>A small number of cell deaths have occurred in species
<code>A-</code> up to this point, so we can simulate the system until
there are 100 of them.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_event</span><span class="op">(</span><span class="st">"death"</span>, <span class="st">"A-"</span>, <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span>
<span></span>
<span><span class="co"># The row "death", for "A" "-" now reports 100</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_firings</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;    event mutant epistate fired</span></span>
<span><span class="co">#&gt; 1  death      A        -   100</span></span>
<span><span class="co">#&gt; 2 growth      A        -   535</span></span>
<span><span class="co">#&gt; 3 switch      A        -    66</span></span>
<span><span class="co">#&gt; 4  death      A        +  1670</span></span>
<span><span class="co">#&gt; 5 growth      A        +  2514</span></span>
<span><span class="co">#&gt; 6 switch      A        +   131</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Plot the tissue by using 200 bins</span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">200</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-16-1.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="clock-based-simulation">Clock-Based Simulation<a class="anchor" aria-label="anchor" href="#clock-based-simulation"></a>
</h3>
<p>It is also possible to take the current simulation clock as
reference, and simulate further.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get the simulation clock</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 84.48365</span></span>
<span></span>
<span><span class="co"># Run the simulation for other 15 time units</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_time</span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span>
<span></span>
<span><span class="co"># Get again the simulation clock</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 99.49105</span></span></code></pre></div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-18-1.png" width="288"></p>
<div class="section level5">
<h5 id="getting-cells-advanced">Getting Cells (Advanced)<a class="anchor" aria-label="anchor" href="#getting-cells-advanced"></a>
</h5>
<p>At this point, if we query the simulation we will find more cells (we
use <code>dplyr</code> to process query results). For convenience, the
getters accept parameters to subset the tissue.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load dplyr to use %&gt;%</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue at current simulation time</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   10962      A        +        470        508</span></span>
<span><span class="co">#&gt; 2   10961      A        -        471        508</span></span>
<span><span class="co">#&gt; 3   10974      A        +        472        491</span></span>
<span><span class="co">#&gt; 4   10973      A        +        472        492</span></span>
<span><span class="co">#&gt; 5   10762      A        +        472        507</span></span>
<span><span class="co">#&gt; 6    9721      A        +        472        508</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue rectangular sample having</span></span>
<span><span class="co"># [500,500] and [505,505] as lower and upper corners, respectively</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">505</span>, <span class="fl">505</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   10034      A        +        500        500</span></span>
<span><span class="co">#&gt; 2    5462      A        +        500        501</span></span>
<span><span class="co">#&gt; 3    6564      A        +        500        502</span></span>
<span><span class="co">#&gt; 4    4686      A        -        500        503</span></span>
<span><span class="co">#&gt; 5    7362      A        -        500        504</span></span>
<span><span class="co">#&gt; 6    7681      A        +        500        505</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue having epigenetic state "-"</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   10961      A        -        471        508</span></span>
<span><span class="co">#&gt; 2    9748      A        -        473        499</span></span>
<span><span class="co">#&gt; 3    9747      A        -        474        500</span></span>
<span><span class="co">#&gt; 4    6093      A        -        475        497</span></span>
<span><span class="co">#&gt; 5   10794      A        -        475        505</span></span>
<span><span class="co">#&gt; 6    7757      A        -        476        503</span></span>
<span></span>
<span><span class="co"># Get the cells in the tissue having epigenetic state "-" and,</span></span>
<span><span class="co"># at the same time, belonging to rectangular sample bounded by</span></span>
<span><span class="co"># [500,500] and [505,505] as lower and upper corners, respectively</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">500</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">505</span>, <span class="fl">505</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"-"</span><span class="op">)</span><span class="op">)</span>  <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1    4686      A        -        500        503</span></span>
<span><span class="co">#&gt; 2    7362      A        -        500        504</span></span>
<span><span class="co">#&gt; 3    6569      A        -        501        500</span></span>
<span><span class="co">#&gt; 4     735      A        -        501        502</span></span>
<span><span class="co">#&gt; 5    2962      A        -        501        503</span></span>
<span><span class="co">#&gt; 6    9111      A        -        501        504</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="evolving-new-species">Evolving new species<a class="anchor" aria-label="anchor" href="#evolving-new-species"></a>
</h3>
<p>rRACES can select cells from the tissue, randomly for every mutant,
or in a constrained tissue area.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Stochastic sampling from the whole tissue: it can return A+ or A-</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1    9706      A        -        500        473</span></span>
<span></span>
<span><span class="co"># Calling it again may result in a different cell</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   10930      A        +        486        474</span></span>
<span></span>
<span><span class="co"># Constrain sampling in the tissue rectangular selection [500,550]x[350,450]</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span>,  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">350</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">550</span>, <span class="fl">450</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   10930      A        +        486        474</span></span></code></pre></div>
<p>This feature can be used to program the generation of new species,
mimicking new mutants that generate subclonal expansions.</p>
<p>Imagine we want to add a new mutant <code>B</code> with epistates –
and therefore new species <code>B+</code> and <code>B-</code> - as
descending from <code>A,</code>we need to:</p>
<ul>
<li><p>locate one cell of mutant <code>A</code> in the tissue, which is
where we will inject the new mutant;</p></li>
<li><p>add the specifics of mutant <code>B</code> (via
<code>Simulation$add_mutant()</code>, as we did for
<code>A</code>);</p></li>
<li><p>implement the change of the cell of mutant <code>A</code> to a
cell of mutant <code>B</code>.</p></li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># We locate a random cell</span></span>
<span><span class="va">cell</span> <span class="op">&lt;-</span> <span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span><span class="op">)</span></span>
<span><span class="va">cell</span></span>
<span><span class="co">#&gt;   cell_id mutant epistate position_x position_y</span></span>
<span><span class="co">#&gt; 1   10588      A        +        481        481</span></span>
<span></span>
<span><span class="co"># Add mutant</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"B"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.05</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.3</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.05</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Then we inject the cell, and simulate a little bit.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Mutant injection</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">mutate_progeny</span><span class="op">(</span><span class="va">cell</span>, <span class="st">"B"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Generated event A -&gt; B either in + or - epistates is now recorded</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_counts</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate counts</span></span>
<span><span class="co">#&gt; 1      A        -    906</span></span>
<span><span class="co">#&gt; 2      A        +   1204</span></span>
<span><span class="co">#&gt; 3      B        -      0</span></span>
<span><span class="co">#&gt; 4      B        +      1</span></span>
<span></span>
<span><span class="co"># New evolution</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_size</span><span class="op">(</span><span class="st">"B+"</span>, <span class="fl">600</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span></code></pre></div>
<p>At this point, we can inspect in more details the tissue. It can help
to facet on the species to clearly appreciate the spatial diffusion of
the populations.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-23-1.png" width="288"></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-23-2.png" width="288"></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Facet on species via ggplot</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">species</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-24-1.png" width="576"></p>
<p>If, at this point in the simulation, we generate a new mutant
<code>C</code> from <code>A</code> in the rectangle <span class="math inline">\([450,500]\times [550, 600]\)</span>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define evolutionary parameters</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">add_mutant</span><span class="op">(</span>name <span class="op">=</span> <span class="st">"C"</span>,</span>
<span>               epigenetic_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+-"</span> <span class="op">=</span> <span class="fl">0.1</span>, <span class="st">"-+"</span> <span class="op">=</span> <span class="fl">0.1</span><span class="op">)</span>,</span>
<span>               growth_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.2</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.4</span><span class="op">)</span>,</span>
<span>               death_rates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"+"</span> <span class="op">=</span> <span class="fl">0.1</span>, <span class="st">"-"</span> <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Choose and mutate</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">mutate_progeny</span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="fu">choose_cell_in</span><span class="op">(</span><span class="st">"A"</span>,  <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">450</span>, <span class="fl">550</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">500</span>, <span class="fl">600</span><span class="op">)</span><span class="op">)</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">run_up_to_time</span><span class="op">(</span><span class="va">sim</span><span class="op">$</span><span class="fu">get_clock</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fl">7</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [████████████████████████████████████████] 100% [00m:00s] Saving snapshot</span></span></code></pre></div>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_state.html">plot_state</a></span><span class="op">(</span><span class="va">sim</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-26-1.png" width="288"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_tissue.html">plot_tissue</a></span><span class="op">(</span><span class="va">sim</span>, num_of_bins <span class="op">=</span> <span class="fl">250</span><span class="op">)</span></span></code></pre></div>
<p><img src="tissue_simulation_files/figure-html/unnamed-chunk-26-2.png" width="288"></p>
</div>
<div class="section level3">
<h3 id="other-operations">Other Operations<a class="anchor" aria-label="anchor" href="#other-operations"></a>
</h3>
<div class="section level5">
<h5 id="injection-of-cells-over-a-tissue">Injection of Cells over a Tissue<a class="anchor" aria-label="anchor" href="#injection-of-cells-over-a-tissue"></a>
</h5>
<p>On the tissue, we can inject multiple cells manually; all injected
cells can be retrieved.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Now it will return just the initial cell</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="fu">get_added_cells</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;   mutant epistate position_x position_y time</span></span>
<span><span class="co">#&gt; 1      A        +        500        500    0</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="avoiding-drift">Avoiding Drift<a class="anchor" aria-label="anchor" href="#avoiding-drift"></a>
</h5>
<p>Any species that has a non-zero death rate can become extinct
stochastically by <em>drift</em>.</p>
<p>Drift makes it difficult to simulate an be confident of what species
are in the model. To facilitate the user, RACES can avoid drift by
setting a <em>death activation level.</em> This value is the minimum
number of cells that enables cell death in a species: a cell of species
<span class="math inline">\(S\)</span> can die if and only if <span class="math inline">\(S\)</span> has reached the death activation level
at least once during the simulation.</p>
<p>This threshold holds for all the species and it is set to 1 by
default.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sim</span><span class="op">$</span><span class="va">death_activation_level</span></span>
<span><span class="co">#&gt; [1] 50</span></span>
<span></span>
<span><span class="co"># Change death activation level</span></span>
<span><span class="va">sim</span><span class="op">$</span><span class="va">death_activation_level</span> <span class="op">&lt;-</span> <span class="fl">50</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Alberto Casagrande, Giulio Caravagna.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
